

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>align &mdash; MDAnalysis v0.6.5 documentation</title>
    <link rel="stylesheet" href="../_static/default.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '0.6.5',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="top" title="MDAnalysis v0.6.5 documentation" href="../index.html" />
    <link rel="up" title="Module code" href="index.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li><a href="../mdanalysis_documentation_index.html">MDAnalysis v0.6.5 documentation</a> &raquo;</li>
          <li><a href="index.html" accesskey="U">Module code</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <h1>Source code for align</h1><div class="highlight"><pre>
<span class="c"># aligning of coordinate frames</span>
<span class="c"># Part of MDAnalysis http://mdanalysis.googlecode.com</span>
<span class="c"># Copyright (c) 2006-2010 Naveen Michaud-Agrawal, Elizabeth J. Denning, Oliver Beckstein</span>
<span class="c"># Released under the GNU Public Licence, v2+</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Coordinate fitting and alignment --- :mod:`MDAnalysis.analysis.align`</span>
<span class="sd">=====================================================================</span>

<span class="sd">:Author: Oliver Beckstein</span>
<span class="sd">:Year: 2010</span>
<span class="sd">:Copyright: GNU Public License v3</span>

<span class="sd">The script shows how to select a group of atoms (such as the C-alphas),</span>
<span class="sd">calculate the RMSD and transformation matrix, and apply the</span>
<span class="sd">transformation to the current frame of a trajectory to obtain the</span>
<span class="sd">rotated structure.</span>

<span class="sd">In this example we use files provided as part of the MDAnalysis test</span>
<span class="sd">suite (in the variables PSF, DCD, and PDB_small)::</span>

<span class="sd">   &gt;&gt;&gt; from MDAnalysis import *</span>
<span class="sd">   &gt;&gt;&gt; from MDAnalysis.analysis.align import *</span>
<span class="sd">   &gt;&gt;&gt; from MDAnalysis.tests.datafiles import PSF, DCD, PDB_small</span>
<span class="sd">   &gt;&gt;&gt; ref = Universe(PSF, PDB_small)   # reference structure 1AKE</span>
<span class="sd">   &gt;&gt;&gt; trj = Universe(PSF, DCD)         # trajectory of change 1AKE-&gt;4AKE</span>
<span class="sd">   &gt;&gt;&gt; rms_fit_trj(trj, ref, filename=&#39;rmsfit.dcd&#39;)</span>


<span class="sd">It is also possible to align two arbitrary structures by providing a</span>
<span class="sd">mapping between atoms based on a sequence alignment. This allows</span>
<span class="sd">fitting of structural homologs or wild type and mutant.</span>

<span class="sd">If a alignment was provided as &quot;sequences.aln&quot; one would first produce</span>
<span class="sd">the appropriate MDAnalysis selections with the :func:`fasta2select`</span>
<span class="sd">function and then feed the resulting dictionary to :func:`rms_fit_trj`::</span>

<span class="sd">   &gt;&gt;&gt; seldict = fasta2select(&#39;sequences.aln&#39;)</span>
<span class="sd">   &gt;&gt;&gt; rms_fit_trj(trj, ref, filename=&#39;rmsfit.dcd&#39;, select=seldict)</span>

<span class="sd">(See the documentation of the functions for this advanced usage.)</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="kn">import</span> <span class="nn">numpy</span>
<span class="kn">import</span> <span class="nn">MDAnalysis</span>
<span class="kn">import</span> <span class="nn">MDAnalysis.core.rms_fitting</span>
<span class="kn">from</span> <span class="nn">MDAnalysis</span> <span class="kn">import</span> <span class="n">SelectionError</span>

<span class="kn">import</span> <span class="nn">os.path</span>
<span class="kn">import</span> <span class="nn">sys</span>

<span class="kn">import</span> <span class="nn">logging</span>
<span class="n">logger</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="s">&#39;MDAnalysis.analysis.align&#39;</span><span class="p">)</span>

<div class="viewcode-block" id="rmsd"><a class="viewcode-back" href="../documentation_pages/analysis_modules.html#align.rmsd">[docs]</a><span class="k">def</span> <span class="nf">rmsd</span><span class="p">(</span><span class="n">a</span><span class="p">,</span><span class="n">b</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Returns RMSD between two coordinate sets a and b.&quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">numpy</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">numpy</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">numpy</span><span class="o">.</span><span class="n">power</span><span class="p">(</span><span class="n">a</span><span class="o">-</span><span class="n">b</span><span class="p">,</span><span class="mi">2</span><span class="p">))</span><span class="o">/</span><span class="n">a</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
</div>
<div class="viewcode-block" id="echo"><a class="viewcode-back" href="../documentation_pages/analysis_modules.html#align.echo">[docs]</a><span class="k">def</span> <span class="nf">echo</span><span class="p">(</span><span class="n">s</span><span class="o">=</span><span class="s">&#39;&#39;</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Simple string output that immediately prints to the console.&quot;&quot;&quot;</span>
    <span class="n">sys</span><span class="o">.</span><span class="n">stderr</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">s</span><span class="p">)</span>
    <span class="n">sys</span><span class="o">.</span><span class="n">stderr</span><span class="o">.</span><span class="n">flush</span><span class="p">()</span>
</div>
<div class="viewcode-block" id="rms_fit_trj"><a class="viewcode-back" href="../documentation_pages/analysis_modules.html#align.rms_fit_trj">[docs]</a><span class="k">def</span> <span class="nf">rms_fit_trj</span><span class="p">(</span><span class="n">traj</span><span class="p">,</span><span class="n">ref</span><span class="p">,</span><span class="n">select</span><span class="o">=</span><span class="s">&#39;backbone&#39;</span><span class="p">,</span><span class="n">filename</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span><span class="n">prefix</span><span class="o">=</span><span class="s">&#39;rmsfit_&#39;</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;RMS-fit trajectory to a reference structure using a selection.</span>

<span class="sd">      rms_fit_trj(traj, ref, &#39;backbone or name CB or name OT*&#39;)</span>

<span class="sd">    :Arguments:</span>
<span class="sd">      *traj*</span>
<span class="sd">         trajectory, :class:`MDAnalysis.Universe` object</span>
<span class="sd">      *ref*</span>
<span class="sd">         reference coordinates; :class:`MDAnalysis.Universe` object</span>
<span class="sd">         (uses the current time step of the object)</span>
<span class="sd">      *filename*</span>
<span class="sd">         file name for the RMS-fitted trajectory or pdb; defaults to the </span>
<span class="sd">         original trajectory filename (from *traj*) with *prefix* prepended</span>
<span class="sd">      *prefix*</span>
<span class="sd">         prefix for autogenerating the new output filename</span>
<span class="sd">      *select*</span>
<span class="sd">         any valid selection string for</span>
<span class="sd">         :meth:`MDAnalysis.AtomGroup.selectAtoms` that produces identical</span>
<span class="sd">         selections in *traj* and *ref* or dictionary {&#39;reference&#39;:sel1,</span>
<span class="sd">         &#39;target&#39;:sel2}.  The :func:`fasta2select` function returns such a</span>
<span class="sd">         dictionary based on a ClustalW_ or STAMP_ sequence alignment.</span>

<span class="sd">    Both reference and trajectory must be :class:`MDAnalysis.Universe`</span>
<span class="sd">    instances. If they contain a trajectory then it is used. The</span>
<span class="sd">    output file format is the same as the input *traj*.</span>

<span class="sd">    .. Note:: This function is flexible but slow. In order to fit a long</span>
<span class="sd">       trajectory it is suggested to only align one frame and then use</span>
<span class="sd">       that frame with a more efficient program (that otherwise could</span>
<span class="sd">       not easily deal with aligning different molecules).</span>

<span class="sd">    .. _ClustalW: http://www.clustal.org/</span>
<span class="sd">    .. _STAMP: http://www.compbio.dundee.ac.uk/manuals/stamp.4.2/</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">frames</span> <span class="o">=</span> <span class="n">traj</span><span class="o">.</span><span class="n">trajectory</span>

    <span class="k">if</span> <span class="n">filename</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">:</span>
        <span class="n">path</span><span class="p">,</span><span class="n">fn</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">frames</span><span class="o">.</span><span class="n">filename</span><span class="p">)</span>
        <span class="n">filename</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">path</span><span class="p">,</span><span class="n">prefix</span><span class="o">+</span><span class="n">fn</span><span class="p">)</span>
    <span class="k">if</span> <span class="nb">type</span><span class="p">(</span><span class="n">select</span><span class="p">)</span> <span class="ow">is</span> <span class="nb">str</span><span class="p">:</span>
        <span class="n">select</span> <span class="o">=</span> <span class="p">{</span><span class="s">&#39;reference&#39;</span><span class="p">:</span><span class="n">select</span><span class="p">,</span><span class="s">&#39;target&#39;</span><span class="p">:</span><span class="n">select</span><span class="p">}</span>

    <span class="c"># TODO: dealing with different formats is not nice; this should be</span>
    <span class="c"># rewritten with the trajectory API in mind...</span>
    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">frames</span><span class="p">,</span><span class="n">MDAnalysis</span><span class="o">.</span><span class="n">coordinates</span><span class="o">.</span><span class="n">DCD</span><span class="o">.</span><span class="n">DCDReader</span><span class="p">):</span>
        <span class="n">writer</span> <span class="o">=</span> <span class="n">MDAnalysis</span><span class="o">.</span><span class="n">coordinates</span><span class="o">.</span><span class="n">DCD</span><span class="o">.</span><span class="n">DCDWriter</span><span class="p">(</span>
            <span class="n">filename</span><span class="p">,</span><span class="n">frames</span><span class="o">.</span><span class="n">numatoms</span><span class="p">,</span>
            <span class="n">frames</span><span class="o">.</span><span class="n">start_timestep</span><span class="p">,</span>
            <span class="n">frames</span><span class="o">.</span><span class="n">skip_timestep</span><span class="p">,</span>
            <span class="n">frames</span><span class="o">.</span><span class="n">delta</span><span class="p">,</span>
            <span class="n">remarks</span><span class="o">=</span><span class="s">&#39;RMS fitted trajectory to ref&#39;</span><span class="p">)</span>
    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">frames</span><span class="p">,</span><span class="n">MDAnalysis</span><span class="o">.</span><span class="n">coordinates</span><span class="o">.</span><span class="n">XTC</span><span class="o">.</span><span class="n">XTCReader</span><span class="p">):</span>
        <span class="n">writer</span> <span class="o">=</span> <span class="n">MDAnalysis</span><span class="o">.</span><span class="n">coordinates</span><span class="o">.</span><span class="n">XTC</span><span class="o">.</span><span class="n">XTCWriter</span><span class="p">(</span>  <span class="c"># untested!</span>
            <span class="n">filename</span><span class="p">,</span><span class="n">frames</span><span class="o">.</span><span class="n">numatoms</span><span class="p">,</span>
            <span class="n">frames</span><span class="o">.</span><span class="n">start_timestep</span><span class="p">,</span>
            <span class="n">frames</span><span class="o">.</span><span class="n">skip_timestep</span><span class="p">,</span>
            <span class="n">frames</span><span class="o">.</span><span class="n">delta</span><span class="p">,</span>
            <span class="n">remarks</span><span class="o">=</span><span class="s">&#39;RMS fitted trajectory to ref&#39;</span><span class="p">)</span>
    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">frames</span><span class="p">,</span><span class="n">MDAnalysis</span><span class="o">.</span><span class="n">coordinates</span><span class="o">.</span><span class="n">TRR</span><span class="o">.</span><span class="n">TRRReader</span><span class="p">):</span>
        <span class="n">writer</span> <span class="o">=</span> <span class="n">MDAnalysis</span><span class="o">.</span><span class="n">coordinates</span><span class="o">.</span><span class="n">TRR</span><span class="o">.</span><span class="n">TRRWriter</span><span class="p">(</span>  <span class="c"># untested!</span>
            <span class="n">filename</span><span class="p">,</span><span class="n">frames</span><span class="o">.</span><span class="n">numatoms</span><span class="p">,</span>
            <span class="n">frames</span><span class="o">.</span><span class="n">start_timestep</span><span class="p">,</span>
            <span class="n">frames</span><span class="o">.</span><span class="n">skip_timestep</span><span class="p">,</span>
            <span class="n">frames</span><span class="o">.</span><span class="n">delta</span><span class="p">,</span>
            <span class="n">remarks</span><span class="o">=</span><span class="s">&#39;RMS fitted trajectory to ref&#39;</span><span class="p">)</span>
    <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">frames</span><span class="p">,</span><span class="n">MDAnalysis</span><span class="o">.</span><span class="n">coordinates</span><span class="o">.</span><span class="n">PDB</span><span class="o">.</span><span class="n">PDBReader</span><span class="p">):</span>
        <span class="n">writer</span> <span class="o">=</span> <span class="n">MDAnalysis</span><span class="o">.</span><span class="n">coordinates</span><span class="o">.</span><span class="n">PDB</span><span class="o">.</span><span class="n">PDBWriter</span><span class="p">(</span> <span class="c"># still working like this?</span>
            <span class="n">filename</span><span class="p">,</span> <span class="n">universe</span><span class="o">=</span><span class="n">traj</span><span class="p">,</span>
            <span class="n">remarks</span><span class="o">=</span><span class="s">&#39;RMS fitted pdb frame to ref&#39;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s">&#39;traj must contain a DCD, XTC, TRR, or a PDB.&#39;</span><span class="p">)</span>

    <span class="n">ref_atoms</span> <span class="o">=</span> <span class="n">ref</span><span class="o">.</span><span class="n">selectAtoms</span><span class="p">(</span><span class="n">select</span><span class="p">[</span><span class="s">&#39;reference&#39;</span><span class="p">])</span>
    <span class="n">traj_atoms</span> <span class="o">=</span> <span class="n">traj</span><span class="o">.</span><span class="n">selectAtoms</span><span class="p">(</span><span class="n">select</span><span class="p">[</span><span class="s">&#39;target&#39;</span><span class="p">])</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">ref_atoms</span><span class="p">)</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="n">traj_atoms</span><span class="p">):</span>
        <span class="k">raise</span> <span class="n">SelectionError</span><span class="p">(</span><span class="s">&quot;Reference and trajectory atom selections do not contain &quot;</span><span class="o">+</span>
                             <span class="s">&quot;the same number of atoms: N_ref=</span><span class="si">%d</span><span class="s">, N_traj=</span><span class="si">%d</span><span class="s">&quot;</span> <span class="o">%</span> \
                             <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">ref_atoms</span><span class="p">),</span> <span class="nb">len</span><span class="p">(</span><span class="n">traj_atoms</span><span class="p">)))</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&quot;RMS-fitting on </span><span class="si">%d</span><span class="s"> atoms.&quot;</span> <span class="o">%</span> <span class="nb">len</span><span class="p">(</span><span class="n">ref_atoms</span><span class="p">))</span>
    <span class="n">mass_mismatches</span> <span class="o">=</span> <span class="p">(</span><span class="n">numpy</span><span class="o">.</span><span class="n">absolute</span><span class="p">(</span><span class="n">ref_atoms</span><span class="o">.</span><span class="n">masses</span><span class="p">()</span> <span class="o">-</span> <span class="n">traj_atoms</span><span class="o">.</span><span class="n">masses</span><span class="p">())</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">numpy</span><span class="o">.</span><span class="n">any</span><span class="p">(</span><span class="n">mass_mismatches</span><span class="p">):</span>
        <span class="c"># diagnostic output:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="s">&quot;Atoms: reference | trajectory&quot;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">ar</span><span class="p">,</span><span class="n">at</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">ref_atoms</span><span class="p">,</span><span class="n">traj_atoms</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">ar</span><span class="o">.</span><span class="n">name</span> <span class="o">!=</span> <span class="n">at</span><span class="o">.</span><span class="n">name</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="s">&quot;</span><span class="si">%4s</span><span class="s"> </span><span class="si">%3d</span><span class="s"> </span><span class="si">%3s</span><span class="s"> </span><span class="si">%3s</span><span class="s"> </span><span class="si">%6.3f</span><span class="s">  |  </span><span class="si">%4s</span><span class="s"> </span><span class="si">%3d</span><span class="s"> </span><span class="si">%3s</span><span class="s"> </span><span class="si">%3s</span><span class="s"> </span><span class="si">%6.3f</span><span class="s">&quot;</span> <span class="o">%</span>  \
                      <span class="p">(</span><span class="n">ar</span><span class="o">.</span><span class="n">segid</span><span class="p">,</span> <span class="n">ar</span><span class="o">.</span><span class="n">resid</span><span class="p">,</span> <span class="n">ar</span><span class="o">.</span><span class="n">resname</span><span class="p">,</span> <span class="n">ar</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="n">ar</span><span class="o">.</span><span class="n">mass</span><span class="p">,</span>
                       <span class="n">at</span><span class="o">.</span><span class="n">segid</span><span class="p">,</span> <span class="n">at</span><span class="o">.</span><span class="n">resid</span><span class="p">,</span> <span class="n">at</span><span class="o">.</span><span class="n">resname</span><span class="p">,</span> <span class="n">at</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="n">at</span><span class="o">.</span><span class="n">mass</span><span class="p">,))</span>
        <span class="n">errmsg</span> <span class="o">=</span> <span class="s">&quot;Inconsistent selections, masses don&#39;t match; mis-matching atoms are shown above.&quot;</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="n">errmsg</span><span class="p">)</span>
        <span class="k">raise</span> <span class="n">SelectionError</span><span class="p">(</span><span class="n">errmsg</span><span class="p">)</span>
    <span class="k">del</span> <span class="n">mass_mismatches</span>
    <span class="n">masses</span> <span class="o">=</span> <span class="n">ref_atoms</span><span class="o">.</span><span class="n">masses</span><span class="p">()</span>

    <span class="c"># reference centre of mass system</span>
    <span class="c"># (compatibility with pre 1.0 numpy: explicitly cast coords to float32)</span>
    <span class="n">ref_com</span> <span class="o">=</span> <span class="n">ref_atoms</span><span class="o">.</span><span class="n">centerOfMass</span><span class="p">()</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="n">numpy</span><span class="o">.</span><span class="n">float32</span><span class="p">)</span>
    <span class="n">ref_coordinates</span> <span class="o">=</span> <span class="n">ref_atoms</span><span class="o">.</span><span class="n">coordinates</span><span class="p">()</span> <span class="o">-</span> <span class="n">ref_com</span>

    <span class="c"># allocate the array for selection atom coords</span>
    <span class="n">traj_coordinates</span> <span class="o">=</span> <span class="n">traj_atoms</span><span class="o">.</span><span class="n">coordinates</span><span class="p">()</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

    <span class="c"># R: rotation matrix that aligns r-r_com, x~-x~com   </span>
    <span class="c">#    (x~: selected coordinates, x: all coordinates)</span>
    <span class="c"># Final transformed traj coordinates: x&#39; = (x-x~_com)*R + ref_com</span>
    <span class="k">for</span> <span class="n">ts</span> <span class="ow">in</span> <span class="n">frames</span><span class="p">:</span>
        <span class="c"># shift coordinates for rotation fitting</span>
        <span class="c"># selection is updated with the time frame</span>
        <span class="n">x_com</span> <span class="o">=</span> <span class="n">traj_atoms</span><span class="o">.</span><span class="n">centerOfMass</span><span class="p">()</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="n">numpy</span><span class="o">.</span><span class="n">float32</span><span class="p">)</span>
        <span class="n">traj_coordinates</span><span class="p">[:]</span> <span class="o">=</span> <span class="n">traj_atoms</span><span class="o">.</span><span class="n">coordinates</span><span class="p">()</span> <span class="o">-</span> <span class="n">x_com</span>
        <span class="n">R</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">matrix</span><span class="p">(</span><span class="n">MDAnalysis</span><span class="o">.</span><span class="n">core</span><span class="o">.</span><span class="n">rms_fitting</span><span class="o">.</span><span class="n">rms_rotation_matrix</span><span class="p">(</span>
                <span class="n">traj_coordinates</span><span class="p">,</span><span class="n">ref_coordinates</span><span class="p">,</span><span class="n">masses</span><span class="p">),</span><span class="n">dtype</span><span class="o">=</span><span class="n">numpy</span><span class="o">.</span><span class="n">float32</span><span class="p">)</span>
        <span class="c"># Transform each atom in the trajectory (use inplace ops to avoid copying arrays)</span>
        <span class="n">ts</span><span class="o">.</span><span class="n">_pos</span>   <span class="o">-=</span> <span class="n">x_com</span>
        <span class="n">ts</span><span class="o">.</span><span class="n">_pos</span><span class="p">[:]</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">_pos</span> <span class="o">*</span> <span class="n">R</span> <span class="c"># R acts to the left &amp; is broadcasted N times.</span>
        <span class="n">ts</span><span class="o">.</span><span class="n">_pos</span>   <span class="o">+=</span> <span class="n">ref_com</span>
        <span class="n">writer</span><span class="o">.</span><span class="n">write_next_timestep</span><span class="p">(</span><span class="n">ts</span><span class="p">)</span>
        <span class="c"># for debugging:</span>
        <span class="c"># rmsd_old = rmsd(ref_atoms.coordinates(),traj_coordinates)</span>
        <span class="c"># rmsd_new = rmsd(ref_atoms.coordinates(),traj_atoms.coordinates())</span>
        <span class="c"># logger.debug(&quot;Fitted frame %5d/%d  [%5.1f%%]  %5.2fA --&gt; %5.2fA  |translation|=%.2fA\r&quot; % \</span>
        <span class="c">#            (ts.frame,frames.numframes,100.0*ts.frame/frames.numframes,</span>
        <span class="c">#             rmsd_old, rmsd_new, rmsd(x_com,ref_com)) )</span>
        <span class="k">if</span> <span class="n">ts</span><span class="o">.</span><span class="n">frame</span> <span class="o">%</span> <span class="mi">10</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">echo</span><span class="p">(</span><span class="s">&quot;Fitted frame </span><span class="si">%5d</span><span class="s">/</span><span class="si">%d</span><span class="s">  [</span><span class="si">%5.1f%%</span><span class="s">]</span><span class="se">\r</span><span class="s">&quot;</span> <span class="o">%</span> \
                <span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">frame</span><span class="p">,</span><span class="n">frames</span><span class="o">.</span><span class="n">numframes</span><span class="p">,</span><span class="mf">100.0</span><span class="o">*</span><span class="n">ts</span><span class="o">.</span><span class="n">frame</span><span class="o">/</span><span class="n">frames</span><span class="o">.</span><span class="n">numframes</span><span class="p">))</span>
    <span class="c"># done</span>
    <span class="n">echo</span><span class="p">(</span><span class="s">&quot;Fitted frame </span><span class="si">%5d</span><span class="s">/</span><span class="si">%d</span><span class="s">  [</span><span class="si">%5.1f%%</span><span class="s">]</span><span class="se">\r\n</span><span class="s">&quot;</span> <span class="o">%</span> \
             <span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">frame</span><span class="p">,</span><span class="n">frames</span><span class="o">.</span><span class="n">numframes</span><span class="p">,</span><span class="mf">100.0</span><span class="o">*</span><span class="n">ts</span><span class="o">.</span><span class="n">frame</span><span class="o">/</span><span class="n">frames</span><span class="o">.</span><span class="n">numframes</span><span class="p">))</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&quot;Wrote </span><span class="si">%d</span><span class="s"> RMS-fitted coordinate frames to file </span><span class="si">%r</span><span class="s">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">frames</span><span class="o">.</span><span class="n">numframes</span><span class="p">,</span> <span class="n">filename</span><span class="p">))</span>
</div>
<div class="viewcode-block" id="fasta2select"><a class="viewcode-back" href="../documentation_pages/analysis_modules.html#align.fasta2select">[docs]</a><span class="k">def</span> <span class="nf">fasta2select</span><span class="p">(</span><span class="n">fastafilename</span><span class="p">,</span><span class="n">is_aligned</span><span class="o">=</span><span class="bp">False</span><span class="p">,</span>
                 <span class="n">ref_resids</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span> <span class="n">target_resids</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                 <span class="n">ref_offset</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span><span class="n">target_offset</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span><span class="n">verbosity</span><span class="o">=</span><span class="mi">3</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Align two sequences from a FASTA file and construct a MDAnalysis</span>
<span class="sd">    selection string of the common atoms.</span>

<span class="sd">      fasta2select(fastafilename) -&gt; selection_dict</span>

<span class="sd">    *fastafilename* contains the two un-aligned sequences in FASTA</span>
<span class="sd">    format. The reference is assumed to be the first sequence, the</span>
<span class="sd">    target the second. ClustalW_ produces a pairwise</span>
<span class="sd">    alignment (which is written to a file with suffix .aln).  The</span>
<span class="sd">    output contains atom selection strings that select the same atoms</span>
<span class="sd">    in the two structures.</span>

<span class="sd">    Unless *ref_offset* and/or *target_offset* are specified, the resids</span>
<span class="sd">    in the structure are assumed to correspond to the positions in the</span>
<span class="sd">    un-aligned sequence, namely the first residue has resid == 1.</span>

<span class="sd">    In more complicated cases (e.g. when the resid numbering in the</span>
<span class="sd">    structure/psf has gaps due to missing parts), simply provide the</span>
<span class="sd">    sequence of resids as they appear in the psf in *ref_resids* or</span>
<span class="sd">    *target_resids*, e.g. ::</span>

<span class="sd">       target_resids = [a.resid for a in trj.selectAtoms(&#39;name CA&#39;)]</span>
<span class="sd"> </span>
<span class="sd">    (This translation table *is* combined with any value for *xxx_offset*!)</span>

<span class="sd">    :Arguments:</span>
<span class="sd">      *fastafilename*    </span>
<span class="sd">         FASTA file with first sequence as reference and</span>
<span class="sd">         second the one to be aligned (ORDER IS IMPORTANT!)</span>
<span class="sd">      *is_aligned*</span>
<span class="sd">         False: run clustalw for sequence alignment; True: use</span>
<span class="sd">         the alignment in the file (e.g. from STAMP)</span>
<span class="sd">      *ref_offset*</span>
<span class="sd">         add this number to the column number in the FASTA file</span>
<span class="sd">         to get the original residue number</span>
<span class="sd">      *target_offset*</span>
<span class="sd">         same for the target</span>
<span class="sd">      *ref_resids*</span>
<span class="sd">         sequence of resids as they appear in the reference structure</span>
<span class="sd">      *target_resids*</span>
<span class="sd">         sequence of resids as they appear in the target</span>
<span class="sd"> </span>
<span class="sd">    :Returns:</span>
<span class="sd">      *select_dict*      </span>
<span class="sd">          dictionary with &#39;reference&#39; and &#39;target&#39; selection string</span>
<span class="sd">          that can be used immediately in :func:`rms_fit_trj` as</span>
<span class="sd">          ``select=select_dict``.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="kn">import</span> <span class="nn">Bio</span>
    <span class="kn">import</span> <span class="nn">numpy</span>
    
    <span class="k">if</span> <span class="n">is_aligned</span><span class="p">:</span>
        <span class="kn">import</span> <span class="nn">Bio.SeqIO</span><span class="o">,</span> <span class="nn">Bio.Alphabet</span>
        <span class="n">protein_gapped</span> <span class="o">=</span> <span class="n">Bio</span><span class="o">.</span><span class="n">Alphabet</span><span class="o">.</span><span class="n">Gapped</span><span class="p">(</span><span class="n">Bio</span><span class="o">.</span><span class="n">Alphabet</span><span class="o">.</span><span class="n">IUPAC</span><span class="o">.</span><span class="n">protein</span><span class="p">)</span>
        <span class="n">fasta</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">fastafilename</span><span class="p">)</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">alignment</span> <span class="o">=</span> <span class="n">Bio</span><span class="o">.</span><span class="n">SeqIO</span><span class="o">.</span><span class="n">to_alignment</span><span class="p">(</span>
                <span class="n">Bio</span><span class="o">.</span><span class="n">SeqIO</span><span class="o">.</span><span class="n">FastaIO</span><span class="o">.</span><span class="n">FastaIterator</span><span class="p">(</span><span class="n">fasta</span><span class="p">,</span><span class="n">alphabet</span><span class="o">=</span><span class="n">protein_gapped</span><span class="p">),</span>
                <span class="n">alphabet</span><span class="o">=</span><span class="n">protein_gapped</span><span class="p">)</span>
        <span class="k">finally</span><span class="p">:</span>
            <span class="n">fasta</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&quot;Using provided alignment, </span><span class="si">%s</span><span class="s">&quot;</span><span class="p">,</span> <span class="n">fastafilename</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="kn">import</span> <span class="nn">Bio.Clustalw</span>
        <span class="kn">import</span> <span class="nn">os.path</span>
        <span class="n">filepath</span><span class="p">,</span><span class="n">ext</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">splitext</span><span class="p">(</span><span class="n">fastafilename</span><span class="p">)</span>
        <span class="n">alnfilename</span> <span class="o">=</span> <span class="n">filepath</span> <span class="o">+</span> <span class="s">&#39;.aln&#39;</span>
        <span class="n">cline</span> <span class="o">=</span> <span class="n">Bio</span><span class="o">.</span><span class="n">Clustalw</span><span class="o">.</span><span class="n">MultipleAlignCL</span><span class="p">(</span><span class="n">fastafilename</span><span class="p">)</span>
        <span class="n">cline</span><span class="o">.</span><span class="n">set_output</span><span class="p">(</span><span class="n">alnfilename</span><span class="p">)</span>
        <span class="n">cline</span><span class="o">.</span><span class="n">set_type</span><span class="p">(</span><span class="s">&#39;protein&#39;</span><span class="p">)</span>
        <span class="n">alignment</span> <span class="o">=</span> <span class="n">Bio</span><span class="o">.</span><span class="n">Clustalw</span><span class="o">.</span><span class="n">do_alignment</span><span class="p">(</span><span class="n">cline</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&quot;Using clustalw sequence alignment, </span><span class="si">%s</span><span class="s">.</span><span class="se">\n</span><span class="s">&quot;</span> <span class="o">%</span> <span class="n">alnfilename</span><span class="p">)</span>

    <span class="n">nseq</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">alignment</span><span class="o">.</span><span class="n">_records</span><span class="p">)</span>    <span class="c"># the stupid class should provide __len__ !</span>
    <span class="k">if</span> <span class="n">nseq</span> <span class="o">!=</span> <span class="mi">2</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s">&quot;Only two sequences in the alignment can be processed.&quot;</span><span class="p">)</span>

    <span class="n">orig_resids</span> <span class="o">=</span> <span class="p">[</span><span class="n">ref_resids</span><span class="p">,</span><span class="n">target_resids</span><span class="p">]</span> <span class="c"># implict assertion that</span>
                                             <span class="c"># we only have to sequenceses in the alignment</span>
    <span class="n">offsets</span> <span class="o">=</span> <span class="p">[</span><span class="n">ref_offset</span><span class="p">,</span><span class="n">target_offset</span><span class="p">]</span>
    <span class="k">for</span> <span class="n">iseq</span><span class="p">,</span><span class="n">a</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">alignment</span><span class="p">):</span>      <span class="c"># need iseq index to change orig_resids</span>
        <span class="k">if</span> <span class="n">orig_resids</span><span class="p">[</span><span class="n">iseq</span><span class="p">]</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">:</span>
            <span class="c"># build default: assume consecutive numbering of all</span>
            <span class="c"># residues in the alignment</span>
            <span class="n">GAP</span> <span class="o">=</span> <span class="n">a</span><span class="o">.</span><span class="n">seq</span><span class="o">.</span><span class="n">alphabet</span><span class="o">.</span><span class="n">gap_char</span>
            <span class="n">length</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">a</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span> <span class="o">-</span> <span class="n">a</span><span class="o">.</span><span class="n">seq</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="n">GAP</span><span class="p">)</span>
            <span class="n">orig_resids</span><span class="p">[</span><span class="n">iseq</span><span class="p">]</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="n">length</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">orig_resids</span><span class="p">[</span><span class="n">iseq</span><span class="p">]</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">asarray</span><span class="p">(</span><span class="n">orig_resids</span><span class="p">[</span><span class="n">iseq</span><span class="p">])</span>
    <span class="c"># add offsets to the sequence &lt;--&gt; resid translation table</span>
    <span class="n">seq2resids</span> <span class="o">=</span> <span class="p">[</span><span class="n">resids</span> <span class="o">+</span> <span class="n">offset</span> <span class="k">for</span> <span class="n">resids</span><span class="p">,</span><span class="n">offset</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">orig_resids</span><span class="p">,</span><span class="n">offsets</span><span class="p">)]</span>
    <span class="k">del</span> <span class="n">orig_resids</span>
    <span class="k">del</span> <span class="n">offsets</span>
	
    <span class="k">def</span> <span class="nf">resid_factory</span><span class="p">(</span><span class="n">alignment</span><span class="p">,</span><span class="n">seq2resids</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Return a function that gives the resid for a position ipos in</span>
<span class="sd">        the nseq&#39;th alignment.</span>
<span class="sd">        </span>
<span class="sd">        resid = resid_factory(alignment,seq2resids)</span>
<span class="sd">        r = resid(nseq,ipos)</span>

<span class="sd">        It is based on a look up table that translates position in the</span>
<span class="sd">        alignment to the residue number in the original</span>
<span class="sd">        sequence/structure.</span>
<span class="sd">        </span>
<span class="sd">        The first index of resid() is the alignmment number, the</span>
<span class="sd">        second the position in the alignment.</span>

<span class="sd">        seq2resids translates the residues in the sequence to resid</span>
<span class="sd">        numbers in the psf. In the simplest case this is a linear map</span>
<span class="sd">        but if whole parts such as loops are ommitted from the protein</span>
<span class="sd">        the seq2resids may have big gaps.  </span>

<span class="sd">        Format: a tuple of two numpy arrays; the first array is for</span>
<span class="sd">        the reference, the second for the target, The index in each</span>
<span class="sd">        array gives the consecutive number of the amino acid in the</span>
<span class="sd">        sequence, the value the resid in the structure/psf.</span>

<span class="sd">        Note: assumes that alignments have same length and are padded if</span>
<span class="sd">        necessary.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c"># could maybe use Bio.PDB.StructureAlignment instead?</span>
        <span class="n">nseq</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">alignment</span><span class="o">.</span><span class="n">_records</span><span class="p">)</span>
        <span class="n">t</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">nseq</span><span class="p">,</span><span class="n">alignment</span><span class="o">.</span><span class="n">get_alignment_length</span><span class="p">()),</span><span class="n">dtype</span><span class="o">=</span><span class="nb">int</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">iseq</span><span class="p">,</span><span class="n">a</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">alignment</span><span class="p">):</span>
            <span class="n">GAP</span> <span class="o">=</span> <span class="n">a</span><span class="o">.</span><span class="n">seq</span><span class="o">.</span><span class="n">alphabet</span><span class="o">.</span><span class="n">gap_char</span>
            <span class="n">t</span><span class="p">[</span><span class="n">iseq</span><span class="p">,:]</span> <span class="o">=</span> <span class="n">seq2resids</span><span class="p">[</span><span class="n">iseq</span><span class="p">][</span><span class="n">numpy</span><span class="o">.</span><span class="n">cumsum</span><span class="p">(</span><span class="n">numpy</span><span class="o">.</span><span class="n">where</span><span class="p">(</span>
                        <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">a</span><span class="o">.</span><span class="n">seq</span><span class="p">))</span><span class="o">==</span><span class="n">GAP</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">))</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span>  
            <span class="c"># -1 because seq2resid is index-1 based (resids start at 1)</span>

        <span class="k">def</span> <span class="nf">resid</span><span class="p">(</span><span class="n">nseq</span><span class="p">,</span><span class="n">ipos</span><span class="p">):</span>
            <span class="k">return</span> <span class="n">t</span><span class="p">[</span><span class="n">nseq</span><span class="p">,</span><span class="n">ipos</span><span class="p">]</span>
        <span class="k">return</span> <span class="n">resid</span>
    <span class="n">resid</span> <span class="o">=</span> <span class="n">resid_factory</span><span class="p">(</span><span class="n">alignment</span><span class="p">,</span><span class="n">seq2resids</span><span class="p">)</span>

    <span class="n">res_list</span> <span class="o">=</span> <span class="p">[]</span>     <span class="c"># collect individual selection string</span>
    <span class="c"># could collect just resid and type (with/without CB) and</span>
    <span class="c"># then post-process and use ranges for continuous stretches, eg</span>
    <span class="c"># ( resid 1:35 and ( backbone or name CB ) ) or ( resid 36 and backbone ) ...</span>

    <span class="n">GAP</span> <span class="o">=</span> <span class="n">alignment</span><span class="o">.</span><span class="n">get_seq_by_num</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">alphabet</span><span class="o">.</span><span class="n">gap_char</span> <span class="c"># should be same for all seqs</span>
    <span class="k">for</span> <span class="n">ipos</span> <span class="ow">in</span> <span class="nb">xrange</span><span class="p">(</span><span class="n">alignment</span><span class="o">.</span><span class="n">get_alignment_length</span><span class="p">()):</span>
        <span class="n">aligned</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">alignment</span><span class="o">.</span><span class="n">get_column</span><span class="p">(</span><span class="n">ipos</span><span class="p">))</span>
        <span class="k">if</span> <span class="n">GAP</span> <span class="ow">in</span> <span class="n">aligned</span><span class="p">:</span>
            <span class="k">continue</span>       <span class="c"># skip residue</span>
        <span class="n">template</span> <span class="o">=</span> <span class="s">&quot;resid </span><span class="si">%i</span><span class="s">&quot;</span>
        <span class="k">if</span> <span class="s">&#39;G&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">aligned</span><span class="p">:</span>
            <span class="c"># can use CB</span>
            <span class="n">template</span> <span class="o">+=</span> <span class="s">&quot; and ( backbone or name CB )&quot;</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">template</span> <span class="o">+=</span> <span class="s">&quot; and backbone&quot;</span>
        <span class="n">template</span> <span class="o">=</span> <span class="s">&quot;( &quot;</span><span class="o">+</span><span class="n">template</span><span class="o">+</span><span class="s">&quot; )&quot;</span>

        <span class="n">res_list</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">template</span> <span class="o">%</span> <span class="n">resid</span><span class="p">(</span><span class="n">iseq</span><span class="p">,</span><span class="n">ipos</span><span class="p">)</span> <span class="k">for</span> <span class="n">iseq</span> <span class="ow">in</span> <span class="nb">xrange</span><span class="p">(</span><span class="n">nseq</span><span class="p">)])</span>

    <span class="n">sel</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">res_list</span><span class="p">)</span><span class="o">.</span><span class="n">transpose</span><span class="p">()</span>

    <span class="n">ref_selection</span> <span class="o">=</span>  <span class="s">&quot; or &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">sel</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
    <span class="n">target_selection</span> <span class="o">=</span>  <span class="s">&quot; or &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">sel</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
    <span class="k">return</span> <span class="p">{</span><span class="s">&#39;reference&#39;</span><span class="p">:</span><span class="n">ref_selection</span><span class="p">,</span> <span class="s">&#39;target&#39;</span><span class="p">:</span><span class="n">target_selection</span><span class="p">}</span>
</pre></div></div>

          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" size="18" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li><a href="../mdanalysis_documentation_index.html">MDAnalysis v0.6.5 documentation</a> &raquo;</li>
          <li><a href="index.html" >Module code</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2010, Naveen Michaud-Agrawal, Elizabeth J. Denning, Oliver Beckstein, Danny Parton, Philip Fowler, Tyler Reddy.
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.0.4.
    </div>
  </body>
</html>