<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>7.2. Fundamental building blocks — MDAnalysis.core.AtomGroup &mdash; MDAnalysis 0.11.0-dev documentation</title>
    
    <link rel="stylesheet" href="../../_static/sphinxdoc.css" type="text/css" />
    <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../../',
        VERSION:     '0.11.0-dev',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="search" type="application/opensearchdescription+xml"
          title="Search within MDAnalysis 0.11.0-dev documentation"
          href="../../_static/opensearch.xml"/>
    <link rel="shortcut icon" href="../../_static/mdanalysis-logo.ico"/>
    <link rel="top" title="MDAnalysis 0.11.0-dev documentation" href="../../index.html" />
    <link rel="up" title="7. Core modules" href="../core_modules.html" />
    <link rel="next" title="7.3. Atom selection Hierarchy — MDAnalysis.core.Selection" href="Selection.html" />
    <link rel="prev" title="7.1. Core functions of MDAnalysis" href="init.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="Selection.html" title="7.3. Atom selection Hierarchy — MDAnalysis.core.Selection"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="init.html" title="7.1. Core functions of MDAnalysis"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../../index.html">MDAnalysis 0.11.0-dev documentation</a> &raquo;</li>
          <li class="nav-item nav-item-1"><a href="../core_modules.html" accesskey="U">7. Core modules</a> &raquo;</li> 
      </ul>
    </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../../index.html">
              <img class="logo" src="../../_static/mdanalysis-logo-200x150.png" alt="Logo"/>
            </a></p>
  <h3><a href="../../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">7.2. Fundamental building blocks &#8212; <code class="docutils literal"><span class="pre">MDAnalysis.core.AtomGroup</span></code></a><ul>
<li><a class="reference internal" href="#class-hierarchy">7.2.1. Class Hierarchy</a></li>
<li><a class="reference internal" href="#manipulating-atoms-residues-and-segments">7.2.2. Manipulating atoms, residues, and segments</a></li>
<li><a class="reference internal" href="#working-with-topologies">7.2.3. Working with Topologies</a></li>
<li><a class="reference internal" href="#combining-objects-system-building">7.2.4. Combining objects: system building</a><ul>
<li><a class="reference internal" href="#merging">7.2.4.1. Merging</a></li>
<li><a class="reference internal" href="#replicating">7.2.4.2. Replicating</a></li>
</ul>
</li>
<li><a class="reference internal" href="#classes-and-functions">7.2.5. Classes and functions</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="init.html"
                        title="previous chapter">7.1. Core functions of MDAnalysis</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="Selection.html"
                        title="next chapter">7.3. Atom selection Hierarchy &#8212; <code class="docutils literal"><span class="pre">MDAnalysis.core.Selection</span></code></a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../../_sources/documentation_pages/core/AtomGroup.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <span class="target" id="module-MDAnalysis.core.AtomGroup"></span><div class="section" id="fundamental-building-blocks-mdanalysis-core-atomgroup">
<h1>7.2. Fundamental building blocks &#8212; <a class="reference internal" href="#module-MDAnalysis.core.AtomGroup" title="MDAnalysis.core.AtomGroup"><code class="xref py py-mod docutils literal"><span class="pre">MDAnalysis.core.AtomGroup</span></code></a><a class="headerlink" href="#fundamental-building-blocks-mdanalysis-core-atomgroup" title="Permalink to this headline">¶</a></h1>
<p>The most important data structure in MDAnalysis is the
<a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup" title="MDAnalysis.core.AtomGroup.AtomGroup"><code class="xref py py-class docutils literal"><span class="pre">AtomGroup</span></code></a>, which contains <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Atom" title="MDAnalysis.core.AtomGroup.Atom"><code class="xref py py-class docutils literal"><span class="pre">Atom</span></code></a> instances.</p>
<p>A <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Universe" title="MDAnalysis.core.AtomGroup.Universe"><code class="xref py py-class docutils literal"><span class="pre">Universe</span></code></a> is the user-visible entry point and collects all
information needed to analyze a structure or a whole trajectory.</p>
<p>Segments and residues are a way to refer to a collection of atoms. By
convention, a <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Residue" title="MDAnalysis.core.AtomGroup.Residue"><code class="xref py py-class docutils literal"><span class="pre">Residue</span></code></a> is a single amino acid, or a water
molecule, ion, or ligand. A <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Segment" title="MDAnalysis.core.AtomGroup.Segment"><code class="xref py py-class docutils literal"><span class="pre">Segment</span></code></a> is a collection of
residues such as a whole protein or a chain in a protein or all the
water in the system.</p>
<div class="section" id="class-hierarchy">
<h2>7.2.1. Class Hierarchy<a class="headerlink" href="#class-hierarchy" title="Permalink to this headline">¶</a></h2>
<p>A <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Universe" title="MDAnalysis.core.AtomGroup.Universe"><code class="xref py py-class docutils literal"><span class="pre">Universe</span></code></a> contains Segments, which contain Residues, which
contain Atoms; all containers are derived from <a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup" title="MDAnalysis.core.AtomGroup.AtomGroup"><code class="xref py py-class docutils literal"><span class="pre">AtomGroup</span></code></a>, and
thus one can always analyze them as a collection of atoms, independent
of the hierarchical level.</p>
<p>Each <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Atom" title="MDAnalysis.core.AtomGroup.Atom"><code class="xref py py-class docutils literal"><span class="pre">Atom</span></code></a> can only belong to a single <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Residue" title="MDAnalysis.core.AtomGroup.Residue"><code class="xref py py-class docutils literal"><span class="pre">Residue</span></code></a>, and a
<a class="reference internal" href="#MDAnalysis.core.AtomGroup.Residue" title="MDAnalysis.core.AtomGroup.Residue"><code class="xref py py-class docutils literal"><span class="pre">Residue</span></code></a> belongs to one specific <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Segment" title="MDAnalysis.core.AtomGroup.Segment"><code class="xref py py-class docutils literal"><span class="pre">Segment</span></code></a>. This
hierarchy can be described as</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">Segment</span> <span class="o">&gt;</span> <span class="n">Residue</span> <span class="o">&gt;</span> <span class="n">Atom</span>
</pre></div>
</div>
<p>Depending on the use case, it can be more convenient to access data
on, for instance, the basis of residues than atoms, or to write out
individual chains (segments) of a protein. MDAnalysis simply provides
access to these groupings and keeps track of where an atom
belongs. Each object provides three attributes (<a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup.atoms" title="MDAnalysis.core.AtomGroup.AtomGroup.atoms"><code class="xref py py-attr docutils literal"><span class="pre">atoms</span></code></a>,
<a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup.residues" title="MDAnalysis.core.AtomGroup.AtomGroup.residues"><code class="xref py py-attr docutils literal"><span class="pre">residues</span></code></a> or <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Atom.residue" title="MDAnalysis.core.AtomGroup.Atom.residue"><code class="xref py py-attr docutils literal"><span class="pre">residue</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup.segments" title="MDAnalysis.core.AtomGroup.AtomGroup.segments"><code class="xref py py-attr docutils literal"><span class="pre">segments</span></code></a> or
<code class="xref py py-attr docutils literal"><span class="pre">segment</span></code>) that give access to the tiers in the hierarchy
that the object belongs to.</p>
</div>
<div class="section" id="manipulating-atoms-residues-and-segments">
<h2>7.2.2. Manipulating atoms, residues, and segments<a class="headerlink" href="#manipulating-atoms-residues-and-segments" title="Permalink to this headline">¶</a></h2>
<p>When working with MDAnalysis it is useful to remember that the fundamental
object is the <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Atom" title="MDAnalysis.core.AtomGroup.Atom"><code class="xref py py-class docutils literal"><span class="pre">Atom</span></code></a>. Each particle in the topology is represented by
exactly one <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Atom" title="MDAnalysis.core.AtomGroup.Atom"><code class="xref py py-class docutils literal"><span class="pre">Atom</span></code></a> instance. One <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Atom" title="MDAnalysis.core.AtomGroup.Atom"><code class="xref py py-class docutils literal"><span class="pre">Atom</span></code></a>, however, can be a member
of multiple <a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup" title="MDAnalysis.core.AtomGroup.AtomGroup"><code class="xref py py-class docutils literal"><span class="pre">AtomGroup</span></code></a> collections, for instance from different
selections even though they all refer to the <em>same</em> <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Atom" title="MDAnalysis.core.AtomGroup.Atom"><code class="xref py py-class docutils literal"><span class="pre">Atom</span></code></a> object. Thus,
changing a property of a specific and <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Atom" title="MDAnalysis.core.AtomGroup.Atom"><code class="xref py py-class docutils literal"><span class="pre">Atom</span></code></a> in one <a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup" title="MDAnalysis.core.AtomGroup.AtomGroup"><code class="xref py py-class docutils literal"><span class="pre">AtomGroup</span></code></a>
changes it &#8220;everywhere&#8221;.</p>
<p>The same is mostly true for <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Residue" title="MDAnalysis.core.AtomGroup.Residue"><code class="xref py py-class docutils literal"><span class="pre">Residue</span></code></a> instances although they are
derived from <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Atom" title="MDAnalysis.core.AtomGroup.Atom"><code class="xref py py-class docutils literal"><span class="pre">Atom</span></code></a> instances: all <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Atom" title="MDAnalysis.core.AtomGroup.Atom"><code class="xref py py-class docutils literal"><span class="pre">Atom</span></code></a> objects with the same
<a class="reference internal" href="#MDAnalysis.core.AtomGroup.Atom.resid" title="MDAnalysis.core.AtomGroup.Atom.resid"><code class="xref py py-attr docutils literal"><span class="pre">Atom.resid</span></code></a> are bundled into a single <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Residue" title="MDAnalysis.core.AtomGroup.Residue"><code class="xref py py-class docutils literal"><span class="pre">Residue</span></code></a> with
<code class="xref py py-class docutils literal"><span class="pre">Residue.id</span></code> = <em>resid</em>. This means that just changing, say, the residue
name with a command such as</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">r</span> <span class="o">=</span> <span class="n">u</span><span class="o">.</span><span class="n">selectAtoms</span><span class="p">(</span><span class="s">&quot;resid 99&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">residues</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span><span class="p">(</span><span class="n">r</span><span class="p">)</span>
<span class="go">&lt;Residue &#39;ALA&#39;, 99&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">r</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="s">&quot;UNK&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span><span class="p">(</span><span class="n">r</span><span class="p">)</span>
<span class="go">&lt;Residue &#39;UNK&#39;, 99&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rnew</span> <span class="o">=</span> <span class="n">u</span><span class="o">.</span><span class="n">selectAtoms</span><span class="p">(</span><span class="s">&quot;resid 99&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">residues</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span><span class="p">(</span><span class="n">rnew</span><span class="p">)</span>
<span class="go">&lt;Residue &#39;UNK&#39;, 99&gt;</span>
</pre></div>
</div>
<p>will typically work as expected. When working with collections such as
<a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup" title="MDAnalysis.core.AtomGroup.AtomGroup"><code class="xref py py-class docutils literal"><span class="pre">AtomGroup</span></code></a> or <a class="reference internal" href="#MDAnalysis.core.AtomGroup.ResidueGroup" title="MDAnalysis.core.AtomGroup.ResidueGroup"><code class="xref py py-class docutils literal"><span class="pre">ResidueGroup</span></code></a> it is generally better to use
provided setter methods such as <a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup.set_resname" title="MDAnalysis.core.AtomGroup.AtomGroup.set_resname"><code class="xref py py-meth docutils literal"><span class="pre">AtomGroup.set_resname()</span></code></a> or
<a class="reference internal" href="#MDAnalysis.core.AtomGroup.ResidueGroup.set_resname" title="MDAnalysis.core.AtomGroup.ResidueGroup.set_resname"><code class="xref py py-meth docutils literal"><span class="pre">ResidueGroup.set_resname()</span></code></a>.</p>
<p>There are two cases when it is very important to use the setters:</p>
<ul class="simple">
<li>changing <em>resid</em>: <a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup.set_resid" title="MDAnalysis.core.AtomGroup.AtomGroup.set_resid"><code class="xref py py-meth docutils literal"><span class="pre">AtomGroup.set_resid()</span></code></a> and <a class="reference internal" href="#MDAnalysis.core.AtomGroup.ResidueGroup.set_resid" title="MDAnalysis.core.AtomGroup.ResidueGroup.set_resid"><code class="xref py py-meth docutils literal"><span class="pre">ResidueGroup.set_resid()</span></code></a></li>
<li>changing <em>segid</em>: <a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup.set_segid" title="MDAnalysis.core.AtomGroup.AtomGroup.set_segid"><code class="xref py py-meth docutils literal"><span class="pre">AtomGroup.set_segid()</span></code></a> and <code class="xref py py-meth docutils literal"><span class="pre">ResidueGroup.set_segid()</span></code></li>
</ul>
<p>Because residues are determined by the <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Atom.resid" title="MDAnalysis.core.AtomGroup.Atom.resid"><code class="xref py py-attr docutils literal"><span class="pre">Atom.resid</span></code></a> and segments by
<a class="reference internal" href="#MDAnalysis.core.AtomGroup.Atom.segid" title="MDAnalysis.core.AtomGroup.Atom.segid"><code class="xref py py-attr docutils literal"><span class="pre">Atom.segid</span></code></a>, the above methods take extra care to rebuild the list of
segments and residues.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last"><a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup.set_resid" title="MDAnalysis.core.AtomGroup.AtomGroup.set_resid"><code class="xref py py-meth docutils literal"><span class="pre">AtomGroup.set_resid()</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.AtomGroup.ResidueGroup.set_resid" title="MDAnalysis.core.AtomGroup.ResidueGroup.set_resid"><code class="xref py py-meth docutils literal"><span class="pre">ResidueGroup.set_resid()</span></code></a>,
<a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup.set_segid" title="MDAnalysis.core.AtomGroup.AtomGroup.set_segid"><code class="xref py py-meth docutils literal"><span class="pre">AtomGroup.set_segid()</span></code></a>, <code class="xref py py-meth docutils literal"><span class="pre">ResidueGroup.set_segid()</span></code> can change the
topology: they can split or merge residues or segments.</p>
</div>
<p>Splitting/merging of residues is probably not very useful because no chemical
rearrangements are carried out. Manipulating segments might be more useful in
order to add additional structure to a <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Universe" title="MDAnalysis.core.AtomGroup.Universe"><code class="xref py py-class docutils literal"><span class="pre">Universe</span></code></a> and provide instant
segment selectors for interactive work:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">u</span><span class="o">.</span><span class="n">selectAtoms</span><span class="p">(</span><span class="s">&quot;protein&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">set_segid</span><span class="p">(</span><span class="s">&quot;protein&quot;</span><span class="p">)</span>
<span class="n">u</span><span class="o">.</span><span class="n">selectAtoms</span><span class="p">(</span><span class="s">&quot;resname POPE or resname POPC&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">set_segid</span><span class="p">(</span><span class="s">&quot;lipids&quot;</span><span class="p">)</span>
<span class="n">u</span><span class="o">.</span><span class="n">selectAtoms</span><span class="p">(</span><span class="s">&quot;resname SOL&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">set_segid</span><span class="p">(</span><span class="s">&quot;water&quot;</span><span class="p">)</span>
<span class="n">u</span><span class="o">.</span><span class="n">selectAtoms</span><span class="p">(</span><span class="s">&quot;resname NA or resname CL&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">set_segid</span><span class="p">(</span><span class="s">&quot;ions&quot;</span><span class="p">)</span>

<span class="n">u</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">numberOfResidues</span><span class="p">()</span>
<span class="n">water_oxygens</span> <span class="o">=</span> <span class="n">u</span><span class="o">.</span><span class="n">water</span><span class="o">.</span><span class="n">OW</span>
</pre></div>
</div>
<p>The setter methods have the additional advantage that they can assign
lists. For instance, many MD codes number residues consecutively starting from
1. However, the original structure might be missing a few atoms at the
N-terminus. Let&#8217;s say that the first residue is really residue 10. In order to
store the canonical residue IDs (&#8220;resnum&#8221;) one could the use</p>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">import</span> <span class="nn">numpy</span> <span class="kn">as</span> <span class="nn">np</span>
<span class="n">protein</span> <span class="o">=</span> <span class="n">u</span><span class="o">.</span><span class="n">selectAtoms</span><span class="p">(</span><span class="s">&quot;protein&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">residues</span>
<span class="n">protein</span><span class="o">.</span><span class="n">set_resnum</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">protein</span><span class="o">.</span><span class="n">resnums</span><span class="p">())</span> <span class="o">+</span> <span class="mi">9</span><span class="p">)</span>
</pre></div>
</div>
<p>One can then use <code class="docutils literal"><span class="pre">protein.select(&quot;resnum</span> <span class="pre">42&quot;)</span></code> to select the residue that has
the canonical residue id 42 (instead of <code class="docutils literal"><span class="pre">resid</span> <span class="pre">33</span></code>).</p>
<p>One can also read the resids directly from  an original PDB file:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">orig</span> <span class="o">=</span> <span class="n">MDAnalysis</span><span class="o">.</span><span class="n">Universe</span><span class="p">(</span><span class="s">&quot;2jln.pdb&quot;</span><span class="p">)</span>
<span class="n">protein</span><span class="o">.</span><span class="n">set_resnum</span><span class="p">(</span><span class="n">orig</span><span class="o">.</span><span class="n">selectAtoms</span><span class="p">(</span><span class="s">&quot;protein&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">resids</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="section" id="working-with-topologies">
<h2>7.2.3. Working with Topologies<a class="headerlink" href="#working-with-topologies" title="Permalink to this headline">¶</a></h2>
<p>If the topology file given to the Universe had bonding information then this
will have been loaded into the Universe as <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Universe.bonds" title="MDAnalysis.core.AtomGroup.Universe.bonds"><code class="xref py py-attr docutils literal"><span class="pre">Universe.bonds</span></code></a>
<a class="reference internal" href="#MDAnalysis.core.AtomGroup.Universe.angles" title="MDAnalysis.core.AtomGroup.Universe.angles"><code class="xref py py-attr docutils literal"><span class="pre">Universe.angles</span></code></a> <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Universe.torsions" title="MDAnalysis.core.AtomGroup.Universe.torsions"><code class="xref py py-attr docutils literal"><span class="pre">Universe.torsions</span></code></a> and <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Universe.impropers" title="MDAnalysis.core.AtomGroup.Universe.impropers"><code class="xref py py-attr docutils literal"><span class="pre">Universe.impropers</span></code></a>.</p>
<p>If your topology file does not have this information, it is still possible
to construct it based on the positions of the atoms and assumed vdw radii
for these atoms.  See <code class="xref py py-meth docutils literal"><span class="pre">MDAnalysis.AtomGroup.guess_bonds()</span></code> and
<a class="reference internal" href="../topology/core.html#MDAnalysis.topology.core.guess_bonds" title="MDAnalysis.topology.core.guess_bonds"><code class="xref py py-func docutils literal"><span class="pre">MDAnalysis.topology.core.guess_bonds()</span></code></a> for details.</p>
<p>This Topology information is stored in <a class="reference internal" href="../topology/core.html#MDAnalysis.topology.core.TopologyGroup" title="MDAnalysis.topology.core.TopologyGroup"><code class="xref py py-class docutils literal"><span class="pre">MDAnalysis.topology.core.TopologyGroup</span></code></a>
objects.  These are designed to be analogous to the AtomGroup container for Atoms.</p>
<p>For examples working with a box of ethanol:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">MDAnalysis</span> <span class="kn">as</span> <span class="nn">mda</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">u</span> <span class="o">=</span> <span class="n">mda</span><span class="o">.</span><span class="n">Universe</span><span class="p">(</span><span class="s">&#39;ethanol.gro&#39;</span><span class="p">,</span> <span class="n">guess_bonds</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">u</span><span class="o">.</span><span class="n">bonds</span>
<span class="go">&lt;TopologyGroup containing 11784 Bonds&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">u</span><span class="o">.</span><span class="n">bonds</span><span class="o">.</span><span class="n">types</span><span class="p">()</span>  <span class="c"># view available types</span>
<span class="go">[(&#39;O&#39;, &#39;H&#39;), (&#39;C&#39;, &#39;O&#39;), (&#39;C&#39;, &#39;H&#39;), (&#39;C&#39;, &#39;C&#39;)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">u</span><span class="o">.</span><span class="n">bonds</span><span class="o">.</span><span class="n">selectBonds</span><span class="p">((</span><span class="s">&#39;C&#39;</span><span class="p">,</span> <span class="s">&#39;O&#39;</span><span class="p">))</span>  <span class="c"># return all C-O bonds from the group</span>
<span class="go">&lt;TopologyGroup containing 1473 Bonds&gt;</span>
</pre></div>
</div>
<p>Bonds are categorised based on the types of atoms.  This is done in a way
so that type (a, b, c) is equivalent to (c, b, a) ie. bonds are reversible.
For example:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">u</span><span class="o">.</span><span class="n">angles</span><span class="o">.</span><span class="n">types</span><span class="p">()</span>
<span class="go">[(&#39;C&#39;, &#39;C&#39;, &#39;H&#39;),</span>
<span class="go"> (&#39;H&#39;, &#39;C&#39;, &#39;H&#39;),</span>
<span class="go"> (&#39;C&#39;, &#39;O&#39;, &#39;H&#39;),</span>
<span class="go"> (&#39;C&#39;, &#39;C&#39;, &#39;O&#39;),</span>
<span class="go"> (&#39;H&#39;, &#39;C&#39;, &#39;O&#39;)]</span>
</pre></div>
</div>
<p>There is only C-C-H bonds and no H-C-C bonds.  Selection however is
aware that sometimes types are reversed:</p>
<div class="highlight-python"><div class="highlight"><pre>u.angles.selectBonds((&#39;H&#39;, &#39;C&#39;, &#39;C&#39;))  # note reversal of type
&gt;&gt;&gt; &lt;TopologyGroup containing 7365 Angles&gt;
</pre></div>
</div>
<p>TopologyGroups can be combined and indexed:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">tg</span> <span class="o">=</span> <span class="n">u</span><span class="o">.</span><span class="n">angles</span><span class="o">.</span><span class="n">selectBonds</span><span class="p">((</span><span class="s">&#39;C&#39;</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">,</span> <span class="s">&#39;O&#39;</span><span class="p">))</span> <span class="o">+</span> <span class="n">u</span><span class="o">.</span><span class="n">angles</span><span class="o">.</span><span class="n">selectBonds</span><span class="p">((</span><span class="s">&#39;C&#39;</span><span class="p">,</span> <span class="s">&#39;O&#39;</span><span class="p">,</span> <span class="s">&#39;H&#39;</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tg</span><span class="o">.</span><span class="n">types</span><span class="p">()</span>
<span class="go">[(&#39;C&#39;, &#39;O&#39;, &#39;H&#39;), (&#39;C&#39;, &#39;C&#39;, &#39;O&#39;)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tg</span><span class="p">[:</span><span class="mi">100</span><span class="p">]</span>
<span class="go">&lt;TopologyGroup containing 100 Angles&gt;</span>
</pre></div>
</div>
<p>Finally, TopologyGroups are linked to some fast Cython calculation methods to
determine bond lengths and angle sizes:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">tg</span><span class="o">.</span><span class="n">angles</span><span class="p">()</span>
<span class="go">array([ 1.88042373,  1.95928987,  1.74770012, ...,  1.79306789,</span>
<span class="go">        1.95522678,  1.88881045])</span>
</pre></div>
</div>
</div>
<div class="section" id="combining-objects-system-building">
<h2>7.2.4. Combining objects: system building<a class="headerlink" href="#combining-objects-system-building" title="Permalink to this headline">¶</a></h2>
<p>It is often convenient to combined multiple groups of atoms into a single
object. If they are contained in a single <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Universe" title="MDAnalysis.core.AtomGroup.Universe"><code class="xref py py-class docutils literal"><span class="pre">Universe</span></code></a> then the methods
described above (especially manipulating the segments) might be
useful. However, if the atoms reside in different universes, the <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Merge" title="MDAnalysis.core.AtomGroup.Merge"><code class="xref py py-func docutils literal"><span class="pre">Merge()</span></code></a>
function can be used.</p>
<div class="section" id="merging">
<h3>7.2.4.1. Merging<a class="headerlink" href="#merging" title="Permalink to this headline">¶</a></h3>
<p>In the following example for <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Merge" title="MDAnalysis.core.AtomGroup.Merge"><code class="xref py py-func docutils literal"><span class="pre">Merge()</span></code></a>, protein, ligand, and solvent were
externally prepared in three different PDB files. They are loaded into separate
<a class="reference internal" href="#MDAnalysis.core.AtomGroup.Universe" title="MDAnalysis.core.AtomGroup.Universe"><code class="xref py py-class docutils literal"><span class="pre">Universe</span></code></a> objects (where they could be further manipulated,
e.g. renumbered, relabeled, rotated, ...) The <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Merge" title="MDAnalysis.core.AtomGroup.Merge"><code class="xref py py-func docutils literal"><span class="pre">Merge()</span></code></a> command is used to
combine all of them together:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">import</span> <span class="nn">MDAnalysis</span>
<span class="n">u1</span> <span class="o">=</span> <span class="n">MDAnalysis</span><span class="o">.</span><span class="n">Universe</span><span class="p">(</span><span class="s">&quot;protein.pdb&quot;</span><span class="p">)</span>
<span class="n">u2</span> <span class="o">=</span> <span class="n">MDAnalysis</span><span class="o">.</span><span class="n">Universe</span><span class="p">(</span><span class="s">&quot;ligand.pdb&quot;</span><span class="p">)</span>
<span class="n">u3</span> <span class="o">=</span> <span class="n">MDAnalysis</span><span class="o">.</span><span class="n">Universe</span><span class="p">(</span><span class="s">&quot;solvent.pdb&quot;</span><span class="p">)</span>
<span class="n">u</span> <span class="o">=</span> <span class="n">MDAnalysis</span><span class="o">.</span><span class="n">Merge</span><span class="p">(</span><span class="n">u1</span><span class="o">.</span><span class="n">selectAtoms</span><span class="p">(</span><span class="s">&quot;protein&quot;</span><span class="p">),</span> <span class="n">u2</span><span class="o">.</span><span class="n">atoms</span><span class="p">,</span> <span class="n">u3</span><span class="o">.</span><span class="n">atoms</span><span class="p">)</span>
<span class="n">u</span><span class="o">.</span><span class="n">atoms</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;system.pdb&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>The complete system is then written out to a new PDB file.</p>
</div>
<div class="section" id="replicating">
<h3>7.2.4.2. Replicating<a class="headerlink" href="#replicating" title="Permalink to this headline">¶</a></h3>
<p>It is also possible to replicate a molecule to build a system with
multiple copies of the same molecule. In the example, we replicate an
AdK molecule and then translate and rotate the second copy:</p>
<div class="highlight-python"><div class="highlight"><pre>import MDAnalysis; from MDAnalysis.tests.datafiles import *
u = MDAnalysis.Universe(PSF, DCD)
p = u.selectAtoms(&quot;protein&quot;)
m = MDAnalysis.Merge(p,p)

# now renumber resids and segids for each copy

# first copy of the protein (need to use atom indices because currently that&#39;s the only reliable property in the
merged universe)
p1 = m.selectAtoms(&quot;bynum 1:3341&quot;)
# second copy
p2 = m.selectAtoms(&quot;bynum 3342:6682&quot;)

p1.set_segid(&quot;A&quot;)
p2.set_segid(&quot;B&quot;)
p2.residues.set_resid(p2.residues.resids() + p1.residues.resids()[-1])  # increment resids for p2 with the last
resid from p1

# you must regenerate the selections after modifying them (see notes in the docs!)
# because the changed resids are not reflected in the selection (due to how residues are referenced internally)
p1 = m.selectAtoms(&quot;segid A&quot;)       # or as before:  m.selectAtoms(&quot;bynum 1:3341&quot;)
p2 = m.selectAtoms(&quot;segid B&quot;)

# rotate and translate
p2.rotateby(180, [0,0,1])
p2.translate([50,0,0])
</pre></div>
</div>
<p>Note that we have to manually set the residue numbers (resids) and
segment identifies because <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Merge" title="MDAnalysis.core.AtomGroup.Merge"><code class="xref py py-func docutils literal"><span class="pre">Merge()</span></code></a> simply concatenates the
existing atoms and only ensures that all data structures are contained
in the new merged universe.</p>
</div>
</div>
<div class="section" id="classes-and-functions">
<h2>7.2.5. Classes and functions<a class="headerlink" href="#classes-and-functions" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="MDAnalysis.core.AtomGroup.Universe">
<em class="property">class </em><code class="descclassname">MDAnalysis.core.AtomGroup.</code><code class="descname">Universe</code><span class="sig-paren">(</span><em>*args</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#Universe"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.Universe" title="Permalink to this definition">¶</a></dt>
<dd><p>The MDAnalysis Universe contains all the information describing the system.</p>
<p>The system always requires a <em>topology</em> file &#8212; in the simplest case just
a list of atoms. This can be a CHARMM/NAMD PSF file or a simple coordinate
file with atom informations such as XYZ, PDB, Gromacs GRO, or CHARMM CRD. See
<a class="reference internal" href="../topology/init.html#supported-topology-formats"><span>Table of Supported Topology Formats</span></a> for what kind of topologies can be read.</p>
<p>A trajectory provides coordinates; the coordinates have to be ordered in
the same way as the list of atoms in the topology. A trajectory can be a
single frame such as a PDB, CRD, or GRO file, or it can be a MD trajectory
(in CHARMM/NAMD/LAMMPS DCD, Gromacs XTC/TRR, or generic XYZ format).  See
<a class="reference internal" href="../coordinates/init.html#id1"><span>Table of supported coordinate formats</span></a> for what can be read as a
&#8220;trajectory&#8221;.</p>
<p>As a special case, when the topology is a XYZ, PDB, GRO or CRD file
then the coordinates are immediately loaded from the &#8220;topology&#8221;
file unless a trajectory is supplied.</p>
<p>Examples for setting up a universe:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">u</span> <span class="o">=</span> <span class="n">Universe</span><span class="p">(</span><span class="n">topology</span><span class="p">,</span> <span class="n">trajectory</span><span class="p">)</span>          <span class="c"># read system from file(s)</span>
<span class="n">u</span> <span class="o">=</span> <span class="n">Universe</span><span class="p">(</span><span class="n">pdbfile</span><span class="p">)</span>                       <span class="c"># read atoms and coordinates from PDB or GRO</span>
<span class="n">u</span> <span class="o">=</span> <span class="n">Universe</span><span class="p">(</span><span class="n">topology</span><span class="p">,</span> <span class="p">[</span><span class="n">traj1</span><span class="p">,</span> <span class="n">traj2</span><span class="p">,</span> <span class="o">...</span><span class="p">])</span> <span class="c"># read from a list of trajectories</span>
<span class="n">u</span> <span class="o">=</span> <span class="n">Universe</span><span class="p">(</span><span class="n">topology</span><span class="p">,</span> <span class="n">traj1</span><span class="p">,</span> <span class="n">traj2</span><span class="p">,</span> <span class="o">...</span><span class="p">)</span>   <span class="c"># read from multiple trajectories</span>
</pre></div>
</div>
<p>Load new data into a universe (replaces old trajectory and does <em>not</em> append):</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">u</span><span class="o">.</span><span class="n">load_new</span><span class="p">(</span><span class="n">trajectory</span><span class="p">)</span>                      <span class="c"># read from a new trajectory file</span>
</pre></div>
</div>
<p>Select atoms, with syntax similar to CHARMM (see
<a class="reference internal" href="#MDAnalysis.core.AtomGroup.Universe.selectAtoms" title="MDAnalysis.core.AtomGroup.Universe.selectAtoms"><code class="xref py py-class docutils literal"><span class="pre">selectAtoms</span></code></a> for details):</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">u</span><span class="o">.</span><span class="n">selectAtoms</span><span class="p">(</span><span class="o">...</span><span class="p">)</span>
</pre></div>
</div>
<p><em>Attributes:</em></p>
<ul class="simple">
<li><a class="reference internal" href="#MDAnalysis.core.AtomGroup.Universe.trajectory" title="MDAnalysis.core.AtomGroup.Universe.trajectory"><code class="xref py py-attr docutils literal"><span class="pre">Universe.trajectory</span></code></a>: currently loaded trajectory reader;
<code class="xref py py-attr docutils literal"><span class="pre">Universe.trajectory.ts</span></code> is the current time step</li>
<li><a class="reference internal" href="#MDAnalysis.core.AtomGroup.Universe.dimensions" title="MDAnalysis.core.AtomGroup.Universe.dimensions"><code class="xref py py-attr docutils literal"><span class="pre">Universe.dimensions</span></code></a>: current system dimensions (simulation unit cell, if
set in the trajectory)</li>
<li><a class="reference internal" href="#MDAnalysis.core.AtomGroup.Universe.bonds" title="MDAnalysis.core.AtomGroup.Universe.bonds"><code class="xref py py-attr docutils literal"><span class="pre">Universe.bonds</span></code></a>: TopologyGroup of bonds in Universe, also
<a class="reference internal" href="#MDAnalysis.core.AtomGroup.Universe.angles" title="MDAnalysis.core.AtomGroup.Universe.angles"><code class="xref py py-attr docutils literal"><span class="pre">Universe.angles</span></code></a>, <code class="xref py py-attr docutils literal"><span class="pre">Universe.dihedrals</span></code>, and <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Universe.impropers" title="MDAnalysis.core.AtomGroup.Universe.impropers"><code class="xref py py-attr docutils literal"><span class="pre">Universe.impropers</span></code></a>
(low level access through <code class="xref py py-attr docutils literal"><span class="pre">Universe._topology</span></code>)</li>
</ul>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">If atom attributes such as element, mass, or charge are not explicitly
provided in the topology file then MDAnalysis tries to guess them (see
<a class="reference internal" href="../topology/tables.html#module-MDAnalysis.topology.tables" title="MDAnalysis.topology.tables"><code class="xref py py-mod docutils literal"><span class="pre">MDAnalysis.topology.tables</span></code></a>). This does not always work and if you
require correct values (e.g. because you want to calculate the center of
mass) then you need to make sure that MDAnalysis gets all the
information needed.</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 0.7.5: </span>Can also read multi-frame PDB files with the
<a class="reference internal" href="../coordinates/PDB.html#MDAnalysis.coordinates.PDB.PrimitivePDBReader" title="MDAnalysis.coordinates.PDB.PrimitivePDBReader"><code class="xref py py-class docutils literal"><span class="pre">PrimitivePDBReader</span></code></a>.</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 0.8: </span>Parse arbitrary number of arguments as a single topology file and a a sequence
of trajectories.</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 0.9.0: </span>Topology information now loaded lazily, but can be forced with
<a class="reference internal" href="#MDAnalysis.core.AtomGroup.Universe.build_topology" title="MDAnalysis.core.AtomGroup.Universe.build_topology"><code class="xref py py-meth docutils literal"><span class="pre">build_topology()</span></code></a>
Changed .bonds attribute to be a <a class="reference internal" href="../topology/core.html#MDAnalysis.topology.core.TopologyGroup" title="MDAnalysis.topology.core.TopologyGroup"><code class="xref py py-class docutils literal"><span class="pre">TopologyGroup</span></code></a>
Added .angles and .torsions attribute as <a class="reference internal" href="../topology/core.html#MDAnalysis.topology.core.TopologyGroup" title="MDAnalysis.topology.core.TopologyGroup"><code class="xref py py-class docutils literal"><span class="pre">TopologyGroup</span></code></a>
Added fragments to Universe cache</p>
</div>
<p>Initialize the central MDAnalysis Universe object.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Arguments:</th><td class="field-body"><dl class="first last docutils">
<dt><em>topologyfile</em></dt>
<dd><p class="first last">A CHARMM/XPLOR PSF topology file, PDB file or Gromacs GRO file; used to define the
list of atoms. If the file includes bond information, partial
charges, atom masses, ... then these data will be available to
MDAnalysis. A &#8220;structure&#8221; file (PSF, PDB or GRO, in the sense of a
topology) is always required.</p>
</dd>
<dt><em>coordinatefile</em></dt>
<dd><p class="first">A trajectory (such as CHARMM DCD, Gromacs XTC/TRR/GRO, XYZ, XYZ.bz2) or a PDB that
will provide coordinates, possibly multiple frames.
If a <strong>list of filenames</strong> is provided then they are sequentially read and appear
as one single trajectory to the Universe. The list can contain different file
formats.</p>
<div class="last deprecated">
<p><span class="versionmodified">Deprecated since version 0.8: </span>Do not use the <em>coordinatefile</em> keyword argument, just provide trajectories as
positional arguments.</p>
</div>
</dd>
<dt><em>permissive</em></dt>
<dd><p class="first last">currently only relevant for PDB files: Set to <code class="docutils literal"><span class="pre">True</span></code> in order to ignore most errors
and read typical MD simulation PDB files; set to <code class="docutils literal"><span class="pre">False</span></code> to read with the Bio.PDB reader,
which can be useful for real Protein Databank PDB files. <code class="docutils literal"><span class="pre">None</span></code>  selects the
MDAnalysis default (which is set in <code class="xref py py-class docutils literal"><span class="pre">MDAnalysis.core.flags</span></code>) [<code class="docutils literal"><span class="pre">None</span></code>]</p>
</dd>
<dt><em>topology_format</em></dt>
<dd><p class="first last">provide the file format of the topology file; <code class="docutils literal"><span class="pre">None</span></code> guesses it from the file
extension [<code class="docutils literal"><span class="pre">None</span></code>]
Can also pass a subclass of <a class="reference internal" href="../topology/base.html#MDAnalysis.topology.base.TopologyReader" title="MDAnalysis.topology.base.TopologyReader"><code class="xref py py-class docutils literal"><span class="pre">MDAnalysis.topology.base.TopologyReader</span></code></a>
to define a custom reader to be used on the topology file.</p>
</dd>
<dt><em>format</em></dt>
<dd><p class="first last">provide the file format of the coordinate or trajectory file;
<code class="docutils literal"><span class="pre">None</span></code> guesses it from the file extension. Note that this
keyword has no effect if a list of file names is supplied because
the &#8220;chained&#8221; reader has to guess the file format for each
individual list member. [<code class="docutils literal"><span class="pre">None</span></code>]
Can also pass a subclass of <a class="reference internal" href="../coordinates/base.html#MDAnalysis.coordinates.base.Reader" title="MDAnalysis.coordinates.base.Reader"><code class="xref py py-class docutils literal"><span class="pre">MDAnalysis.coordinates.base.Reader</span></code></a>
to define a custom reader to be used on the trajectory file.</p>
</dd>
<dt><em>guess_bonds</em></dt>
<dd><p class="first last">Once Universe has been loaded, attempt to guess the connectivity
between atoms.  This will populate the .bonds .angles and
.torsions attributes of the Universe.</p>
</dd>
<dt><em>vdwradii</em></dt>
<dd><p class="first last">For use with <em>guess_bonds</em>. Supply a dict giving a vdwradii for each atom type
which are used in guessing bonds.</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
<p>This routine tries to do the right thing:</p>
<ol class="arabic">
<li><p class="first">If a pdb/gro file is provided instead of a psf and no <em>coordinatefile</em>
then the coordinates are taken from the first file. Thus you can load
a functional universe with</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">u</span> <span class="o">=</span> <span class="n">Universe</span><span class="p">(</span><span class="s">&#39;1ake.pdb&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>If you want to specify the coordinate file format yourself you can
do so using the <em>format</em> keyword:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">u</span> <span class="o">=</span> <span class="n">Universe</span><span class="p">(</span><span class="s">&#39;1ake.ent1&#39;</span><span class="p">,</span> <span class="n">format</span><span class="o">=</span><span class="s">&#39;pdb&#39;</span><span class="p">)</span>
</pre></div>
</div>
</li>
<li><p class="first">If only a topology file without coordinate information is provided
one will have to load coordinates manually using
<a class="reference internal" href="#MDAnalysis.core.AtomGroup.Universe.load_new" title="MDAnalysis.core.AtomGroup.Universe.load_new"><code class="xref py py-meth docutils literal"><span class="pre">Universe.load_new()</span></code></a>. The file format of the topology file
can be explicitly set with the <em>topology_format</em> keyword.</p>
</li>
</ol>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 0.7.4: </span>New <em>topology_format</em> and <em>format</em> parameters to override the file
format detection.</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 0.10.0: </span>Added <code class="docutils literal"><span class="pre">'guess_bonds'</span></code> keyword to cause topology to be guessed on
Universe creation.
Deprecated <code class="docutils literal"><span class="pre">'bonds'</span></code> keyword, use <code class="docutils literal"><span class="pre">'guess_bonds'</span></code> instead.</p>
</div>
<dl class="attribute">
<dt id="MDAnalysis.core.AtomGroup.Universe.angles">
<code class="descname">angles</code><a class="headerlink" href="#MDAnalysis.core.AtomGroup.Universe.angles" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a <a class="reference internal" href="../topology/core.html#MDAnalysis.topology.core.TopologyGroup" title="MDAnalysis.topology.core.TopologyGroup"><code class="xref py py-class docutils literal"><span class="pre">TopologyGroup</span></code></a>
of all angles in the Universe</p>
<div class="versionadded">
<p><span class="versionmodified">New in version 0.9.0.</span></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 0.9.2: </span>Now can return empty TopologyGroup</p>
</div>
</dd></dl>

<dl class="attribute">
<dt id="MDAnalysis.core.AtomGroup.Universe.bonds">
<code class="descname">bonds</code><a class="headerlink" href="#MDAnalysis.core.AtomGroup.Universe.bonds" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a <a class="reference internal" href="../topology/core.html#MDAnalysis.topology.core.TopologyGroup" title="MDAnalysis.topology.core.TopologyGroup"><code class="xref py py-class docutils literal"><span class="pre">TopologyGroup</span></code></a> of all
bonds in the Universe.</p>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 0.9.0: </span>Now a lazily built <a class="reference internal" href="../topology/core.html#MDAnalysis.topology.core.TopologyGroup" title="MDAnalysis.topology.core.TopologyGroup"><code class="xref py py-class docutils literal"><span class="pre">TopologyGroup</span></code></a></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 0.9.2: </span>Now can return empty TopologyGroup</p>
</div>
</dd></dl>

<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.Universe.build_topology">
<code class="descname">build_topology</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#Universe.build_topology"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.Universe.build_topology" title="Permalink to this definition">¶</a></dt>
<dd><p>Bond angle and torsion information is lazily constructed into the
Universe.</p>
<p>This method forces all this information to be loaded.</p>
</dd></dl>

<dl class="attribute">
<dt id="MDAnalysis.core.AtomGroup.Universe.coord">
<code class="descname">coord</code><a class="headerlink" href="#MDAnalysis.core.AtomGroup.Universe.coord" title="Permalink to this definition">¶</a></dt>
<dd><p>Reference to current timestep and coordinates of universe.</p>
<p>The raw trajectory coordinates are <code class="xref py py-attr docutils literal"><span class="pre">Universe.coord._pos</span></code>,
represented as a <code class="xref py py-class docutils literal"><span class="pre">numpy.float32</span></code> array.</p>
<p>Because <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Universe.coord" title="MDAnalysis.core.AtomGroup.Universe.coord"><code class="xref py py-attr docutils literal"><span class="pre">coord</span></code></a> is a reference to a
<a class="reference internal" href="../coordinates/base.html#MDAnalysis.coordinates.base.Timestep" title="MDAnalysis.coordinates.base.Timestep"><code class="xref py py-class docutils literal"><span class="pre">Timestep</span></code></a>, it changes its contents
while one is stepping through the trajectory.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">In order to access the coordinates it is probably better to use the
<a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup.coordinates" title="MDAnalysis.core.AtomGroup.AtomGroup.coordinates"><code class="xref py py-meth docutils literal"><span class="pre">AtomGroup.coordinates()</span></code></a> method; for instance, all coordinates
of the Universe as a numpy array:
<code class="xref py py-meth docutils literal"><span class="pre">Universe.atoms.coordinates()</span></code>.</p>
</div>
</dd></dl>

<dl class="attribute">
<dt id="MDAnalysis.core.AtomGroup.Universe.dimensions">
<code class="descname">dimensions</code><a class="headerlink" href="#MDAnalysis.core.AtomGroup.Universe.dimensions" title="Permalink to this definition">¶</a></dt>
<dd><p>Current dimensions of the unitcell</p>
</dd></dl>

<dl class="attribute">
<dt id="MDAnalysis.core.AtomGroup.Universe.fragments">
<code class="descname">fragments</code><a class="headerlink" href="#MDAnalysis.core.AtomGroup.Universe.fragments" title="Permalink to this definition">¶</a></dt>
<dd><p>Read only tuple of fragments in the Universe</p>
</dd></dl>

<dl class="attribute">
<dt id="MDAnalysis.core.AtomGroup.Universe.impropers">
<code class="descname">impropers</code><a class="headerlink" href="#MDAnalysis.core.AtomGroup.Universe.impropers" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a <a class="reference internal" href="../topology/core.html#MDAnalysis.topology.core.TopologyGroup" title="MDAnalysis.topology.core.TopologyGroup"><code class="xref py py-class docutils literal"><span class="pre">TopologyGroup</span></code></a>
of all improper torsions in the Universe</p>
<div class="versionadded">
<p><span class="versionmodified">New in version 0.9.0.</span></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 0.9.2: </span>Now can return empty TopologyGroup</p>
</div>
</dd></dl>

<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.Universe.load_new">
<code class="descname">load_new</code><span class="sig-paren">(</span><em>filename</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#Universe.load_new"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.Universe.load_new" title="Permalink to this definition">¶</a></dt>
<dd><p>Load coordinates from <em>filename</em>, using the suffix to detect file format.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Arguments:</th><td class="field-body"><dl class="first docutils">
<dt><em>filename</em></dt>
<dd><p class="first last">the coordinate file (single frame or trajectory) <em>or</em> a list of
filenames, which are read one after another.</p>
</dd>
<dt><em>permissive</em></dt>
<dd><p class="first last">currently only relevant for PDB files: Set to <code class="docutils literal"><span class="pre">True</span></code> in order to ignore most errors
and read typical MD simulation PDB files; set to <code class="docutils literal"><span class="pre">False</span></code> to read with the Bio.PDB reader,
which can be useful for real Protein Databank PDB files. <code class="docutils literal"><span class="pre">None</span></code>  selects the
MDAnalysis default (which is set in <code class="xref py py-class docutils literal"><span class="pre">MDAnalysis.core.flags</span></code>) [<code class="docutils literal"><span class="pre">None</span></code>]</p>
</dd>
<dt><em>format</em></dt>
<dd><p class="first last">provide the file format of the coordinate or trajectory file;
<code class="docutils literal"><span class="pre">None</span></code> guesses it from the file extension. Note that this
keyword has no effect if a list of file names is supplied because
the &#8220;chained&#8221; reader has to guess the file format for each
individual list member [<code class="docutils literal"><span class="pre">None</span></code>]
Can also pass a subclass of <a class="reference internal" href="../coordinates/base.html#MDAnalysis.coordinates.base.Reader" title="MDAnalysis.coordinates.base.Reader"><code class="xref py py-class docutils literal"><span class="pre">MDAnalysis.coordinates.base.Reader</span></code></a>
to define a custom reader to be used on the trajectory file.</p>
</dd>
<dt><em>kwargs</em></dt>
<dd><p class="first last">Other kwargs are passed to the trajectory reader (only for advanced use)</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">(filename, trajectory_format) or <code class="docutils literal"><span class="pre">None</span></code> if <em>filename</em> == <code class="docutils literal"><span class="pre">None</span></code></p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Raises:</th><td class="field-body"><p class="first last"><code class="xref py py-exc docutils literal"><span class="pre">TypeError</span></code> if trajectory format can not be
determined or no appropriate trajectory reader found</p>
</td>
</tr>
</tbody>
</table>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 0.8: </span>If a list or sequence that is provided for <em>filename</em>  only contains a single entry
then it is treated as single coordinate file. This has the consequence that it is
not read by the <a class="reference internal" href="../coordinates/base.html#MDAnalysis.coordinates.base.ChainReader" title="MDAnalysis.coordinates.base.ChainReader"><code class="xref py py-class docutils literal"><span class="pre">ChainReader</span></code></a> but directly by
its specialized file format reader, which typically has more features than the
<a class="reference internal" href="../coordinates/base.html#MDAnalysis.coordinates.base.ChainReader" title="MDAnalysis.coordinates.base.ChainReader"><code class="xref py py-class docutils literal"><span class="pre">ChainReader</span></code></a>.</p>
</div>
</dd></dl>

<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.Universe.selectAtoms">
<code class="descname">selectAtoms</code><span class="sig-paren">(</span><em>sel</em>, <em>*othersel</em>, <em>**selgroups</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#Universe.selectAtoms"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.Universe.selectAtoms" title="Permalink to this definition">¶</a></dt>
<dd><p>Select atoms using a CHARMM selection string.</p>
<p>Returns an <a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup" title="MDAnalysis.core.AtomGroup.AtomGroup"><code class="xref py py-class docutils literal"><span class="pre">AtomGroup</span></code></a> with atoms sorted according to their
index in the psf (this is to ensure that there aren&#8217;t any duplicates,
which can happen with complicated selections).</p>
<p>Existing <a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup" title="MDAnalysis.core.AtomGroup.AtomGroup"><code class="xref py py-class docutils literal"><span class="pre">AtomGroup</span></code></a> objects can be passed as named arguments,
which will then be available to the selection parser.</p>
<p>Subselections can be grouped with parentheses.</p>
<dl class="docutils">
<dt>Example::</dt>
<dd><div class="first highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">sel</span> <span class="o">=</span> <span class="n">universe</span><span class="o">.</span><span class="n">selectAtoms</span><span class="p">(</span><span class="s">&quot;segid DMPC and not ( name H* or name O* )&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sel</span>
<span class="go">&lt;AtomGroup with 3420 atoms&gt;</span>
</pre></div>
</div>
<div class="last highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">universe</span><span class="o">.</span><span class="n">selectAtoms</span><span class="p">(</span><span class="s">&quot;around 10 group notHO&quot;</span><span class="p">,</span> <span class="n">notHO</span><span class="o">=</span><span class="n">sel</span><span class="p">)</span>
<span class="go">&lt;AtomGroup with 1250 atoms&gt;</span>
</pre></div>
</div>
</dd>
</dl>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">If exact ordering of atoms is required (for instance, for
<a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup.angle" title="MDAnalysis.core.AtomGroup.AtomGroup.angle"><code class="xref py py-meth docutils literal"><span class="pre">angle()</span></code></a> or <a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup.dihedral" title="MDAnalysis.core.AtomGroup.AtomGroup.dihedral"><code class="xref py py-meth docutils literal"><span class="pre">dihedral()</span></code></a>
calculations) then one supplies selections <em>separately</em> in the
required order. Also, when multiple <a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup" title="MDAnalysis.core.AtomGroup.AtomGroup"><code class="xref py py-class docutils literal"><span class="pre">AtomGroup</span></code></a> instances are
concatenated with the <code class="docutils literal"><span class="pre">+</span></code> operator then the order of <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Atom" title="MDAnalysis.core.AtomGroup.Atom"><code class="xref py py-class docutils literal"><span class="pre">Atom</span></code></a>
instances is preserved and duplicates are not removed.</p>
</div>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="../selections.html#selection-commands-label"><span>Selection Commands</span></a> for further details and examples.</p>
</div>
<p>The selection parser understands the following CASE SENSITIVE <em>keywords</em>:</p>
<p><strong>Simple selections</strong></p>
<blockquote>
<div><dl class="docutils">
<dt>protein, backbone, nucleic, nucleicbackbone</dt>
<dd>selects all atoms that belong to a standard set of residues; a protein
is identfied by a hard-coded set of residue names so it  may not
work for esoteric residues.</dd>
<dt>segid <em>seg-name</em></dt>
<dd>select by segid (as given in the topology), e.g. <code class="docutils literal"><span class="pre">segid</span> <span class="pre">4AKE</span></code> or <code class="docutils literal"><span class="pre">segid</span> <span class="pre">DMPC</span></code></dd>
<dt>resid <em>residue-number-range</em></dt>
<dd>resid can take a single residue number or a range of numbers. A range
consists of two numbers separated by a colon (inclusive) such
as <code class="docutils literal"><span class="pre">resid</span> <span class="pre">1:5</span></code>. A residue number (&#8220;resid&#8221;) is taken directly from the
topology.</dd>
<dt>resnum <em>resnum-number-range</em></dt>
<dd>resnum is the canonical residue number; typically it is set to the residue id
in the original PDB structure.</dd>
<dt>resname <em>residue-name</em></dt>
<dd>select by residue name, e.g. <code class="docutils literal"><span class="pre">resname</span> <span class="pre">LYS</span></code></dd>
<dt>name <em>atom-name</em></dt>
<dd>select by atom name (as given in the topology). Often, this is force
field dependent. Example: <code class="docutils literal"><span class="pre">name</span> <span class="pre">CA</span></code> (for C&amp;alpha; atoms) or <code class="docutils literal"><span class="pre">name</span> <span class="pre">OW</span></code> (for SPC water oxygen)</dd>
<dt>type <em>atom-type</em></dt>
<dd>select by atom type; this is either a string or a number and depends on
the force field; it is read from the topology file (e.g. the CHARMM PSF
file contains numeric atom types). It has non-sensical values when a
PDB or GRO file is used as a topology.</dd>
<dt>atom <em>seg-name</em>  <em>residue-number</em>  <em>atom-name</em></dt>
<dd>a selector for a single atom consisting of segid resid atomname,
e.g. <code class="docutils literal"><span class="pre">DMPC</span> <span class="pre">1</span> <span class="pre">C2</span></code> selects the C2 carbon of the first residue of the DMPC
segment</dd>
<dt>altloc <em>alternative-location</em></dt>
<dd>a selection for atoms where alternative locations are available,
which is often the case with high-resolution crystal structures
e.g. <cite>resid 4 and resname ALA and altloc B</cite> selects only the atoms
of ALA-4 that have an altloc B record.</dd>
</dl>
</div></blockquote>
<p><strong>Boolean</strong></p>
<blockquote>
<div><dl class="docutils">
<dt>not</dt>
<dd>all atoms not in the selection, e.g. <code class="docutils literal"><span class="pre">not</span> <span class="pre">protein</span></code> selects
all atoms that aren&#8217;t part of a protein</dd>
<dt>and, or</dt>
<dd>combine two selections according to the rules of boolean algebra,
e.g. <code class="docutils literal"><span class="pre">protein</span> <span class="pre">and</span> <span class="pre">not</span> <span class="pre">(resname</span> <span class="pre">ALA</span> <span class="pre">or</span> <span class="pre">resname</span> <span class="pre">LYS)</span></code> selects all atoms
that belong to a protein, but are not in a lysine or alanine residue</dd>
</dl>
</div></blockquote>
<p><strong>Geometric</strong></p>
<blockquote>
<div><dl class="docutils">
<dt>around <em>distance</em>  <em>selection</em></dt>
<dd>selects all atoms a certain cutoff away from another selection,
e.g. <code class="docutils literal"><span class="pre">around</span> <span class="pre">3.5</span> <span class="pre">protein</span></code> selects all atoms not belonging to protein
that are within 3.5 Angstroms from the protein</dd>
<dt>point <em>x</em> <em>y</em> <em>z</em>  <em>distance</em></dt>
<dd>selects all atoms within a cutoff of a point in space, make sure
coordinate is separated by spaces, e.g. <code class="docutils literal"><span class="pre">point</span> <span class="pre">5.0</span> <span class="pre">5.0</span> <span class="pre">5.0</span>&nbsp; <span class="pre">3.5</span></code> selects
all atoms within 3.5 Angstroms of the coordinate (5.0, 5.0, 5.0)</dd>
<dt>prop [abs] <em>property</em>  <em>operator</em>  <em>value</em></dt>
<dd>selects atoms based on position, using <em>property</em>  <strong>x</strong>, <strong>y</strong>, or
<strong>z</strong> coordinate. Supports the <strong>abs</strong> keyword (for absolute value) and
the following <em>operators</em>: <strong>&lt;, &gt;, &lt;=, &gt;=, ==, !=</strong>. For example, <code class="docutils literal"><span class="pre">prop</span> <span class="pre">z</span> <span class="pre">&gt;=</span> <span class="pre">5.0</span></code>
selects all atoms with z coordinate greater than 5.0; <code class="docutils literal"><span class="pre">prop</span> <span class="pre">abs</span> <span class="pre">z</span> <span class="pre">&lt;=</span> <span class="pre">5.0</span></code>
selects all atoms within -5.0 &lt;= z &lt;= 5.0.</dd>
</dl>
</div></blockquote>
<p><strong>Connectivity</strong></p>
<blockquote>
<div><dl class="docutils">
<dt>byres <em>selection</em></dt>
<dd>selects all atoms that are in the same segment and residue as
selection, e.g. specify the subselection after the byres keyword</dd>
</dl>
</div></blockquote>
<p><strong>Index</strong></p>
<blockquote>
<div><dl class="docutils">
<dt>bynum <em>index-range</em></dt>
<dd>selects all atoms within a range of (1-based) inclusive indices,
e.g. <code class="docutils literal"><span class="pre">bynum</span> <span class="pre">1</span></code> selects the first atom in the universe; <code class="docutils literal"><span class="pre">bynum</span> <span class="pre">5:10</span></code>
selects atoms 5 through 10 inclusive. All atoms in the
<code class="xref py py-class docutils literal"><span class="pre">MDAnalysis.Universe</span></code> are consecutively numbered, and the index
runs from 1 up to the total number of atoms.</dd>
</dl>
</div></blockquote>
<p><strong>Preexisting selections</strong></p>
<blockquote>
<div><dl class="docutils">
<dt>group <em>group-name</em></dt>
<dd>selects the atoms in the <a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup" title="MDAnalysis.core.AtomGroup.AtomGroup"><code class="xref py py-class docutils literal"><span class="pre">AtomGroup</span></code></a> passed to the function as an
argument named <em>group-name</em>. Only the atoms common to <em>group-name</em> and the
instance <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Universe.selectAtoms" title="MDAnalysis.core.AtomGroup.Universe.selectAtoms"><code class="xref py py-meth docutils literal"><span class="pre">selectAtoms()</span></code></a> was called from will be considered.
<em>group-name</em> will be included in the parsing just by comparison of atom indices.
This means that it is up to the user to make sure they were defined in an
appropriate <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Universe" title="MDAnalysis.core.AtomGroup.Universe"><code class="xref py py-class docutils literal"><span class="pre">Universe</span></code></a>.</dd>
<dt>fullgroup <em>group-name</em></dt>
<dd>just like the <code class="docutils literal"><span class="pre">group</span></code> keyword with the difference that all the atoms of
<em>group-name</em> are included. The resulting selection may therefore have atoms
that were initially absent from the instance <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Universe.selectAtoms" title="MDAnalysis.core.AtomGroup.Universe.selectAtoms"><code class="xref py py-meth docutils literal"><span class="pre">selectAtoms()</span></code></a> was
called from.</dd>
</dl>
</div></blockquote>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 0.7.4: </span>Added <em>resnum</em> selection.</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 0.8.1: </span>Added <em>group</em> and <em>fullgroup</em> selections.</p>
</div>
</dd></dl>

<dl class="attribute">
<dt id="MDAnalysis.core.AtomGroup.Universe.torsions">
<code class="descname">torsions</code><a class="headerlink" href="#MDAnalysis.core.AtomGroup.Universe.torsions" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a <a class="reference internal" href="../topology/core.html#MDAnalysis.topology.core.TopologyGroup" title="MDAnalysis.topology.core.TopologyGroup"><code class="xref py py-class docutils literal"><span class="pre">TopologyGroup</span></code></a>
of all torsions in the Universe</p>
<div class="versionadded">
<p><span class="versionmodified">New in version 0.9.0.</span></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 0.9.2: </span>Now can return empty TopologyGroup</p>
</div>
</dd></dl>

<dl class="attribute">
<dt id="MDAnalysis.core.AtomGroup.Universe.trajectory">
<code class="descname">trajectory</code><a class="headerlink" href="#MDAnalysis.core.AtomGroup.Universe.trajectory" title="Permalink to this definition">¶</a></dt>
<dd><p>Reference to trajectory reader object containing trajectory data.</p>
</dd></dl>

</dd></dl>

<dl class="function">
<dt id="MDAnalysis.core.AtomGroup.Merge">
<code class="descclassname">MDAnalysis.core.AtomGroup.</code><code class="descname">Merge</code><span class="sig-paren">(</span><em>*args</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#Merge"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.Merge" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Universe" title="MDAnalysis.core.AtomGroup.Universe"><code class="xref py py-class docutils literal"><span class="pre">Universe</span></code></a> from two or more <a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup" title="MDAnalysis.core.AtomGroup.AtomGroup"><code class="xref py py-class docutils literal"><span class="pre">AtomGroup</span></code></a> instances.</p>
<p><a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup" title="MDAnalysis.core.AtomGroup.AtomGroup"><code class="xref py py-class docutils literal"><span class="pre">AtomGroup</span></code></a> instances can come from different Universes, or come
directly from a <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Universe.selectAtoms" title="MDAnalysis.core.AtomGroup.Universe.selectAtoms"><code class="xref py py-meth docutils literal"><span class="pre">selectAtoms()</span></code></a> call.</p>
<p>It can also be used with a single <a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup" title="MDAnalysis.core.AtomGroup.AtomGroup"><code class="xref py py-class docutils literal"><span class="pre">AtomGroup</span></code></a> if the user wants to,
for example, re-order the atoms in the Universe.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Arguments:</th><td class="field-body">One or more <a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup" title="MDAnalysis.core.AtomGroup.AtomGroup"><code class="xref py py-class docutils literal"><span class="pre">AtomGroup</span></code></a> instances.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">an instance of <code class="xref py py-class docutils literal"><span class="pre">Universe</span></code></td>
</tr>
<tr class="field-odd field"><th class="field-name">Raises:</th><td class="field-body"><code class="xref py py-exc docutils literal"><span class="pre">ValueError</span></code> for too few arguments or if an AtomGroup is
empty and <code class="xref py py-exc docutils literal"><span class="pre">TypeError</span></code> if arguments are not
<a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup" title="MDAnalysis.core.AtomGroup.AtomGroup"><code class="xref py py-class docutils literal"><span class="pre">AtomGroup</span></code></a> instances.</td>
</tr>
</tbody>
</table>
<p class="rubric">Example</p>
<p>In this example, protein, ligand, and solvent were externally prepared in
three different PDB files. They are loaded into separate <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Universe" title="MDAnalysis.core.AtomGroup.Universe"><code class="xref py py-class docutils literal"><span class="pre">Universe</span></code></a>
objects (where they could be further manipulated, e.g. renumbered,
relabeled, rotated, ...) The <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Merge" title="MDAnalysis.core.AtomGroup.Merge"><code class="xref py py-func docutils literal"><span class="pre">Merge()</span></code></a> command is used to combine all
of them together:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">u1</span> <span class="o">=</span> <span class="n">Universe</span><span class="p">(</span><span class="s">&quot;protein.pdb&quot;</span><span class="p">)</span>
<span class="n">u2</span> <span class="o">=</span> <span class="n">Universe</span><span class="p">(</span><span class="s">&quot;ligand.pdb&quot;</span><span class="p">)</span>
<span class="n">u3</span> <span class="o">=</span> <span class="n">Universe</span><span class="p">(</span><span class="s">&quot;solvent.pdb&quot;</span><span class="p">)</span>
<span class="n">u</span> <span class="o">=</span> <span class="n">Merge</span><span class="p">(</span><span class="n">u1</span><span class="o">.</span><span class="n">selectAtoms</span><span class="p">(</span><span class="s">&quot;protein&quot;</span><span class="p">),</span> <span class="n">u2</span><span class="o">.</span><span class="n">atoms</span><span class="p">,</span> <span class="n">u3</span><span class="o">.</span><span class="n">atoms</span><span class="p">)</span>
<span class="n">u</span><span class="o">.</span><span class="n">atoms</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;system.pdb&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>The complete system is then written out to a new PDB file.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Merging does not create a full trajectory but only a single
structure even if the input consists of one or more trajectories.</p>
</div>
</dd></dl>

<dl class="class">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup">
<em class="property">class </em><code class="descclassname">MDAnalysis.core.AtomGroup.</code><code class="descname">AtomGroup</code><span class="sig-paren">(</span><em>atoms</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#AtomGroup"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup" title="Permalink to this definition">¶</a></dt>
<dd><p>A group of atoms.</p>
<blockquote>
<div>ag = universe.selectAtoms(atom-list)</div></blockquote>
<p>The AtomGroup contains a list of atoms; typically, a AtomGroup is generated
from a selection. It is build from any list-like collection of
<a class="reference internal" href="#MDAnalysis.core.AtomGroup.Atom" title="MDAnalysis.core.AtomGroup.Atom"><code class="xref py py-class docutils literal"><span class="pre">Atom</span></code></a> instances. It is also possible to create an empty AtomGroup
from an empty list.</p>
<p>An AtomGroup can be indexed and sliced like a list:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">ag</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">ag</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
</pre></div>
</div>
<p>will return the first and the last <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Atom" title="MDAnalysis.core.AtomGroup.Atom"><code class="xref py py-class docutils literal"><span class="pre">Atom</span></code></a> in the group
whereas the slice</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">ag</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">6</span><span class="p">:</span><span class="mi">2</span><span class="p">]</span>
</pre></div>
</div>
<p>returns every second element, corresponding to indices 0, 2, and 4.</p>
<p>It also supports &#8220;advanced slicing&#8221; when the argument is a
<a class="reference external" href="http://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.9)"><code class="xref py py-class docutils literal"><span class="pre">numpy.ndarray</span></code></a> or a <a class="reference external" href="http://docs.python.org/library/functions.html#list" title="(in Python v2.7)"><code class="xref py py-class docutils literal"><span class="pre">list</span></code></a>:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">aslice</span> <span class="o">=</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">3</span><span class="p">]</span>
<span class="n">ag</span><span class="p">[</span><span class="n">aslice</span><span class="p">]</span>
</pre></div>
</div>
<p>will return a new <a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup" title="MDAnalysis.core.AtomGroup.AtomGroup"><code class="xref py py-class docutils literal"><span class="pre">AtomGroup</span></code></a> containing (ag[0], ag[3], ag[-1],
ag[10], ag[3]).</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">AtomGroups originating from a selection are sorted and
duplicate elements are removed. This is not true for AtomGroups
produced by slicing. Thus slicing can be used when the order of
atoms is crucial (for instance, in order to define angles or
dihedrals).</p>
</div>
<p>Atoms can also be accessed in a Pythonic fashion by using the atom name as
an attribute. For instance,</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">ag</span><span class="o">.</span><span class="n">CA</span>
</pre></div>
</div>
<p>will provide a <a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup" title="MDAnalysis.core.AtomGroup.AtomGroup"><code class="xref py py-class docutils literal"><span class="pre">AtomGroup</span></code></a> of all CA atoms in the
group. These <em>instant selector</em> attributes are auto-generated for
each atom name encountered in the group.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">The name-attribute instant selector access to atoms is mainly
meant for quick interactive work. Thus it either returns a
single <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Atom" title="MDAnalysis.core.AtomGroup.Atom"><code class="xref py py-class docutils literal"><span class="pre">Atom</span></code></a> if there is only one matching atom, <em>or</em> a
new <a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup" title="MDAnalysis.core.AtomGroup.AtomGroup"><code class="xref py py-class docutils literal"><span class="pre">AtomGroup</span></code></a> for multiple matches.  This makes it
difficult to use the feature consistently in scripts but it is
much better for interactive work.</p>
</div>
<p class="rubric">References for analysis methods</p>
<table class="docutils citation" frame="void" id="dima2004" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label">[Dima2004]</td><td><em>(<a class="fn-backref" href="#id1">1</a>, <a class="fn-backref" href="#id2">2</a>)</em> Dima, R. I., &amp; Thirumalai, D. (2004). Asymmetry in the
shapes of folded and denatured states of proteins. <em>J
Phys Chem B</em>, 108(21),
6564-6570. doi:<a class="reference external" href="http://dx.doi.org/10.1021/jp037128y">10.1021/jp037128y</a></td></tr>
</tbody>
</table>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 0.7.6: </span>An empty AtomGroup can be created and no longer raises a
<a class="reference internal" href="#MDAnalysis.core.AtomGroup.NoDataError" title="MDAnalysis.core.AtomGroup.NoDataError"><code class="xref py py-exc docutils literal"><span class="pre">NoDataError</span></code></a>.</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 0.9.0: </span>The size at which cache is used for atom lookup is now stored as variable
_atomcache_size within the class.
Added fragments manged property. Is a lazily built, cached entry, similar to residues.</p>
</div>
<dl class="attribute">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup._atoms">
<code class="descname">_atoms</code><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup._atoms" title="Permalink to this definition">¶</a></dt>
<dd><p>immutable list of references to the atoms <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Atom" title="MDAnalysis.core.AtomGroup.Atom"><code class="xref py py-class docutils literal"><span class="pre">Atom</span></code></a> in the
group</p>
</dd></dl>

<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup._rebuild_caches">
<code class="descname">_rebuild_caches</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#AtomGroup._rebuild_caches"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup._rebuild_caches" title="Permalink to this definition">¶</a></dt>
<dd><p>Rebuild all AtomGroup caches.</p>
<p>A number of lists and attributes are cached. These caches are lazily
built the first time they are needed. When editing the topology it
might happen that not all caches were synced properly (even though that
this is supposed to happen eventually). In this case the user can
manually force a complete cache rebuild.</p>
<p>Currently the following caches are used:</p>
<ul class="simple">
<li>atoms (for &#8220;in&#8221; lookup); cache is only built for large systems with
&gt; 10,000 atoms</li>
<li>indices (<a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup.indices" title="MDAnalysis.core.AtomGroup.AtomGroup.indices"><code class="xref py py-meth docutils literal"><span class="pre">AtomGroup.indices()</span></code></a>)</li>
<li>masses (<a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup.masses" title="MDAnalysis.core.AtomGroup.AtomGroup.masses"><code class="xref py py-meth docutils literal"><span class="pre">AtomGroup.masses()</span></code></a>)</li>
<li>residues (<a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup.residues" title="MDAnalysis.core.AtomGroup.AtomGroup.residues"><code class="xref py py-attr docutils literal"><span class="pre">AtomGroup.residues</span></code></a>)</li>
<li>segments (<a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup.segments" title="MDAnalysis.core.AtomGroup.AtomGroup.segments"><code class="xref py py-attr docutils literal"><span class="pre">AtomGroup.segments</span></code></a>)</li>
<li>bonds (<a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup.bonds" title="MDAnalysis.core.AtomGroup.AtomGroup.bonds"><code class="xref py py-attr docutils literal"><span class="pre">AtomGroup.bonds</span></code></a>)</li>
<li>angles (<a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup.angles" title="MDAnalysis.core.AtomGroup.AtomGroup.angles"><code class="xref py py-attr docutils literal"><span class="pre">AtomGroup.angles</span></code></a>)</li>
<li>torsions (<a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup.torsions" title="MDAnalysis.core.AtomGroup.AtomGroup.torsions"><code class="xref py py-attr docutils literal"><span class="pre">AtomGroup.torsions</span></code></a>)</li>
<li>improper torsions (<a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup.impropers" title="MDAnalysis.core.AtomGroup.AtomGroup.impropers"><code class="xref py py-attr docutils literal"><span class="pre">AtomGroup.impropers</span></code></a>)</li>
</ul>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup._clear_caches" title="MDAnalysis.core.AtomGroup.AtomGroup._clear_caches"><code class="xref py py-meth docutils literal"><span class="pre">_clear_caches()</span></code></a></p>
</div>
<div class="versionadded">
<p><span class="versionmodified">New in version 0.7.5.</span></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 0.9.0: </span>Added bonds/angles/torsions/impropers to rebuild.
Reworked how things are rebuilt to avoid code duplication.</p>
</div>
</dd></dl>

<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup._clear_caches">
<code class="descname">_clear_caches</code><span class="sig-paren">(</span><em>*args</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#AtomGroup._clear_caches"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup._clear_caches" title="Permalink to this definition">¶</a></dt>
<dd><p>Clear cache for all <em>args</em>.</p>
<p>If no args are provided, all caches are cleared.</p>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup._rebuild_caches" title="MDAnalysis.core.AtomGroup.AtomGroup._rebuild_caches"><code class="xref py py-meth docutils literal"><span class="pre">_rebuild_caches()</span></code></a></p>
</div>
<div class="versionadded">
<p><span class="versionmodified">New in version 0.8.</span></p>
</div>
</dd></dl>

<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup._fill_cache">
<code class="descname">_fill_cache</code><span class="sig-paren">(</span><em>name</em>, <em>value</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#AtomGroup._fill_cache"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup._fill_cache" title="Permalink to this definition">¶</a></dt>
<dd><p>Populate _cache[name] with value.</p>
<div class="versionadded">
<p><span class="versionmodified">New in version 0.8.</span></p>
</div>
</dd></dl>

<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup.align_principalAxis">
<code class="descname">align_principalAxis</code><span class="sig-paren">(</span><em>axis</em>, <em>vector</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#AtomGroup.align_principalAxis"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup.align_principalAxis" title="Permalink to this definition">¶</a></dt>
<dd><p>Align principal axis with index <em>axis</em> with <em>vector</em>.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Arguments:</th><td class="field-body"><dl class="first last docutils">
<dt><em>axis</em></dt>
<dd><p class="first last">Index of the principal axis (0, 1, or 2), as produced by
<a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup.principalAxes" title="MDAnalysis.core.AtomGroup.AtomGroup.principalAxes"><code class="xref py py-meth docutils literal"><span class="pre">principalAxes()</span></code></a>.</p>
</dd>
<dt><em>vector</em></dt>
<dd><p class="first last">A 3D vector such as the z-axis (<code class="docutils literal"><span class="pre">[0,0,1]</span></code>); can be
anything that looks like a list with three entries.</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
<p>To align the long axis of a channel (the first principal axis,
i.e. <em>axis</em> = 0) with the z-axis:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">u</span><span class="o">.</span><span class="n">atoms</span><span class="o">.</span><span class="n">align_principalAxis</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">])</span>
<span class="n">u</span><span class="o">.</span><span class="n">atoms</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;aligned.pdb&quot;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup.angle">
<code class="descname">angle</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#AtomGroup.angle"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup.angle" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the angle in degrees between atoms 0, 1, 2.</p>
<p>Angle between atoms 0 and 2 with apex at 1:</p>
<div class="highlight-python"><div class="highlight"><pre>   2
  /
 /
1------0
</pre></div>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Only makes sense for a <a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup" title="MDAnalysis.core.AtomGroup.AtomGroup"><code class="xref py py-class docutils literal"><span class="pre">AtomGroup</span></code></a> with exactly 3
<a class="reference internal" href="#MDAnalysis.core.AtomGroup.Atom" title="MDAnalysis.core.AtomGroup.Atom"><code class="xref py py-class docutils literal"><span class="pre">Atom</span></code></a>; anything else will raise a <code class="xref py py-exc docutils literal"><span class="pre">ValueError</span></code>.</p>
</div>
<div class="versionadded">
<p><span class="versionmodified">New in version 0.7.3.</span></p>
</div>
</dd></dl>

<dl class="attribute">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup.angles">
<code class="descname">angles</code><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup.angles" title="Permalink to this definition">¶</a></dt>
<dd><p>All the angles in this AtomGroup</p>
<p>Note that these angles might extend out of the AtomGroup, to select
only angles which are entirely contained by the AtomGroup use
u.angles.atomgroup_intersection(ag, strict=True)</p>
<div class="versionadded">
<p><span class="versionmodified">New in version 0.9.0.</span></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 0.10.0: </span>Now always returns a (possibly empty) TopologyGroup</p>
</div>
</dd></dl>

<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup.asphericity">
<code class="descname">asphericity</code><span class="sig-paren">(</span><em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#AtomGroup.asphericity"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup.asphericity" title="Permalink to this definition">¶</a></dt>
<dd><p>Asphericity.</p>
<p>See <a class="reference internal" href="#dima2004" id="id1">[Dima2004]</a> for background information.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Keywords:</th><td class="field-body"><dl class="first last docutils">
<dt><em>pbc</em></dt>
<dd><p class="first last"><code class="docutils literal"><span class="pre">True</span></code>: Move all atoms within primary unit cell before calculation [<code class="docutils literal"><span class="pre">False</span></code>]</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">The <code class="xref py py-class docutils literal"><span class="pre">MDAnalysis.core.flags</span></code> flag <em>use_pbc</em> when set to <code class="docutils literal"><span class="pre">True</span></code> allows the <em>pbc</em>
flag to be used by default.</p>
</div>
<div class="versionadded">
<p><span class="versionmodified">New in version 0.7.7.</span></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 0.8: </span>Added <em>pbc</em> keyword</p>
</div>
</dd></dl>

<dl class="attribute">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup.atoms">
<code class="descname">atoms</code><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup.atoms" title="Permalink to this definition">¶</a></dt>
<dd><p><a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup" title="MDAnalysis.core.AtomGroup.AtomGroup"><code class="xref py py-class docutils literal"><span class="pre">AtomGroup</span></code></a> of all atoms in this group.</p>
<p>If this is a <a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup" title="MDAnalysis.core.AtomGroup.AtomGroup"><code class="xref py py-class docutils literal"><span class="pre">AtomGroup</span></code></a> then it returns itself. Otherwise, it
will return a new <a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup" title="MDAnalysis.core.AtomGroup.AtomGroup"><code class="xref py py-class docutils literal"><span class="pre">AtomGroup</span></code></a> based on all <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Atom" title="MDAnalysis.core.AtomGroup.Atom"><code class="xref py py-class docutils literal"><span class="pre">Atom</span></code></a>
instances contained.</p>
<p>Apply <cite>:func:`list</cite> to <a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup.atoms" title="MDAnalysis.core.AtomGroup.AtomGroup.atoms"><code class="xref py py-attr docutils literal"><span class="pre">atoms</span></code></a> or use <a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup._atoms" title="MDAnalysis.core.AtomGroup.AtomGroup._atoms"><code class="xref py py-attr docutils literal"><span class="pre">_atoms</span></code></a> if you
really only need a list of individual <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Atom" title="MDAnalysis.core.AtomGroup.Atom"><code class="xref py py-class docutils literal"><span class="pre">Atom</span></code></a> instances.</p>
</dd></dl>

<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup.bbox">
<code class="descname">bbox</code><span class="sig-paren">(</span><em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#AtomGroup.bbox"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup.bbox" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the bounding box of the selection.</p>
<p>The lengths A,B,C of the orthorhombic enclosing box are</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">L</span> <span class="o">=</span> <span class="n">AtomGroup</span><span class="o">.</span><span class="n">bbox</span><span class="p">()</span>
<span class="n">A</span><span class="p">,</span><span class="n">B</span><span class="p">,</span><span class="n">C</span> <span class="o">=</span> <span class="n">L</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">-</span> <span class="n">L</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
</pre></div>
</div>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Keywords:</th><td class="field-body"><dl class="first last docutils">
<dt><em>pbc</em></dt>
<dd><p class="first last"><code class="docutils literal"><span class="pre">True</span></code>: Move all atoms within the primary unit cell before calculation [<code class="docutils literal"><span class="pre">False</span></code>]</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">The <code class="xref py py-class docutils literal"><span class="pre">MDAnalysis.core.flags</span></code> flag <em>use_pbc</em> when set to <code class="docutils literal"><span class="pre">True</span></code> allows the <em>pbc</em>
flag to be used by default.</p>
</div>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">[[xmin, ymin, zmin], [xmax, ymax, zmax]]</td>
</tr>
</tbody>
</table>
<div class="versionadded">
<p><span class="versionmodified">New in version 0.7.2.</span></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 0.8: </span>Added <em>pbc</em> keyword</p>
</div>
</dd></dl>

<dl class="attribute">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup.bfactors">
<code class="descname">bfactors</code><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup.bfactors" title="Permalink to this definition">¶</a></dt>
<dd><p>Crystallographic B-factors (from PDB) in A**2.</p>
</dd></dl>

<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup.bond">
<code class="descname">bond</code><span class="sig-paren">(</span><em>pbc=False</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#AtomGroup.bond"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup.bond" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the distance between atoms in a 2-atom group.</p>
<p>Distance between atoms 0 and 1:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="mi">0</span><span class="o">---</span><span class="mi">1</span>
</pre></div>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Only makes sense for a <a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup" title="MDAnalysis.core.AtomGroup.AtomGroup"><code class="xref py py-class docutils literal"><span class="pre">AtomGroup</span></code></a> with exactly 2
<a class="reference internal" href="#MDAnalysis.core.AtomGroup.Atom" title="MDAnalysis.core.AtomGroup.Atom"><code class="xref py py-class docutils literal"><span class="pre">Atom</span></code></a>; anything else will raise a
<code class="xref py py-exc docutils literal"><span class="pre">ValueError</span></code>.</p>
</div>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Keywords:</th><td class="field-body"><dl class="first last docutils">
<dt><em>pbc</em></dt>
<dd><p class="first last"><code class="docutils literal"><span class="pre">True</span></code>: Account for minimum image convention when calculating [<code class="docutils literal"><span class="pre">False</span></code>]</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
<div class="versionadded">
<p><span class="versionmodified">New in version 0.7.3.</span></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 0.8: </span>Added <em>pbc</em> keyword</p>
</div>
</dd></dl>

<dl class="attribute">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup.bonds">
<code class="descname">bonds</code><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup.bonds" title="Permalink to this definition">¶</a></dt>
<dd><p>All the bonds in this AtomGroup</p>
<p>Note that these bonds might extend out of the AtomGroup, to select
only bonds which are entirely contained by the AtomGroup use
u.bonds.atomgroup_intersection(ag, strict=True)</p>
<div class="versionadded">
<p><span class="versionmodified">New in version 0.9.0.</span></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 0.10.0: </span>Now always returns a (possibly empty) TopologyGroup</p>
</div>
</dd></dl>

<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup.bsphere">
<code class="descname">bsphere</code><span class="sig-paren">(</span><em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#AtomGroup.bsphere"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup.bsphere" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the bounding sphere of the selection.</p>
<p>The sphere is calculated relative to the centre of geometry.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Keywords:</th><td class="field-body"><dl class="first last docutils">
<dt><em>pbc</em></dt>
<dd><p class="first last"><code class="docutils literal"><span class="pre">True</span></code>: Move all atoms within primary unit cell before calculation [<code class="docutils literal"><span class="pre">False</span></code>]</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">The <code class="xref py py-class docutils literal"><span class="pre">MDAnalysis.core.flags</span></code> flag <em>use_pbc</em> when set to <code class="docutils literal"><span class="pre">True</span></code> allows the <em>pbc</em>
flag to be used by default.</p>
</div>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body"><cite>(R, [xcen,ycen,zcen])</cite></td>
</tr>
</tbody>
</table>
<div class="versionadded">
<p><span class="versionmodified">New in version 0.7.3.</span></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 0.8: </span>Added <em>pbc</em> keyword</p>
</div>
</dd></dl>

<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup.centerOfGeometry">
<code class="descname">centerOfGeometry</code><span class="sig-paren">(</span><em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#AtomGroup.centerOfGeometry"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup.centerOfGeometry" title="Permalink to this definition">¶</a></dt>
<dd><p>Center of geometry (also known as centroid) of the selection.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Keywords:</th><td class="field-body"><dl class="first last docutils">
<dt><em>pbc</em></dt>
<dd><p class="first last"><code class="docutils literal"><span class="pre">True</span></code>: Move all atoms within the primary unit cell before calculation [<code class="docutils literal"><span class="pre">False</span></code>]</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">The <code class="xref py py-class docutils literal"><span class="pre">MDAnalysis.core.flags</span></code> flag <em>use_pbc</em> when set to <code class="docutils literal"><span class="pre">True</span></code> allows the <em>pbc</em>
flag to be used by default.</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 0.8: </span>Added <em>pbc</em> keyword</p>
</div>
</dd></dl>

<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup.centerOfMass">
<code class="descname">centerOfMass</code><span class="sig-paren">(</span><em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#AtomGroup.centerOfMass"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup.centerOfMass" title="Permalink to this definition">¶</a></dt>
<dd><p>Center of mass of the selection.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Keywords:</th><td class="field-body"><dl class="first last docutils">
<dt><em>pbc</em></dt>
<dd><p class="first last"><code class="docutils literal"><span class="pre">True</span></code>: Move all atoms within the primary unit cell before calculation [<code class="docutils literal"><span class="pre">False</span></code>]</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">The <code class="xref py py-class docutils literal"><span class="pre">MDAnalysis.core.flags</span></code> flag <em>use_pbc</em> when set to <code class="docutils literal"><span class="pre">True</span></code> allows the <em>pbc</em>
flag to be used by default.</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 0.8: </span>Added <em>pbc</em> keyword</p>
</div>
</dd></dl>

<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup.centroid">
<code class="descname">centroid</code><span class="sig-paren">(</span><em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup.centroid" title="Permalink to this definition">¶</a></dt>
<dd><p>Center of geometry (also known as centroid) of the selection.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Keywords:</th><td class="field-body"><dl class="first last docutils">
<dt><em>pbc</em></dt>
<dd><p class="first last"><code class="docutils literal"><span class="pre">True</span></code>: Move all atoms within the primary unit cell before calculation [<code class="docutils literal"><span class="pre">False</span></code>]</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">The <code class="xref py py-class docutils literal"><span class="pre">MDAnalysis.core.flags</span></code> flag <em>use_pbc</em> when set to <code class="docutils literal"><span class="pre">True</span></code> allows the <em>pbc</em>
flag to be used by default.</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 0.8: </span>Added <em>pbc</em> keyword</p>
</div>
</dd></dl>

<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup.charges">
<code class="descname">charges</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#AtomGroup.charges"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup.charges" title="Permalink to this definition">¶</a></dt>
<dd><p>Array of partial charges of the atoms (as defined in the topology)</p>
</dd></dl>

<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup.coordinates">
<code class="descname">coordinates</code><span class="sig-paren">(</span><em>ts=None</em>, <em>copy=False</em>, <em>dtype=&lt;type 'numpy.float32'&gt;</em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup.coordinates" title="Permalink to this definition">¶</a></dt>
<dd><p>NumPy array of the coordinates.</p>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup.positions" title="MDAnalysis.core.AtomGroup.AtomGroup.positions"><code class="xref py py-attr docutils literal"><span class="pre">positions</span></code></a> and <a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup.get_positions" title="MDAnalysis.core.AtomGroup.AtomGroup.get_positions"><code class="xref py py-meth docutils literal"><span class="pre">get_positions()</span></code></a></p>
</div>
<div class="deprecated">
<p><span class="versionmodified">Deprecated since version 0.7.6: </span>In new scripts use <a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup.get_positions" title="MDAnalysis.core.AtomGroup.AtomGroup.get_positions"><code class="xref py py-meth docutils literal"><span class="pre">AtomGroup.get_positions()</span></code></a> preferrably.</p>
</div>
</dd></dl>

<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup.dihedral">
<code class="descname">dihedral</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#AtomGroup.dihedral"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup.dihedral" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate the dihedral angle in degrees.</p>
<p>Dihedral angle around axis connecting atoms 1 and 2 (i.e. the angle
between the planes spanned by atoms (0,1,2) and (1,2,3)):</p>
<div class="highlight-python"><div class="highlight"><pre>        3
        |
  1-----2
 /
0
</pre></div>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Only makes sense for a <a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup" title="MDAnalysis.core.AtomGroup.AtomGroup"><code class="xref py py-class docutils literal"><span class="pre">AtomGroup</span></code></a> with exactly 4
<a class="reference internal" href="#MDAnalysis.core.AtomGroup.Atom" title="MDAnalysis.core.AtomGroup.Atom"><code class="xref py py-class docutils literal"><span class="pre">Atom</span></code></a>; anything else will raise a <code class="xref py py-exc docutils literal"><span class="pre">ValueError</span></code>.</p>
</div>
<div class="versionadded">
<p><span class="versionmodified">New in version 0.7.0.</span></p>
</div>
</dd></dl>

<dl class="attribute">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup.dimensions">
<code class="descname">dimensions</code><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup.dimensions" title="Permalink to this definition">¶</a></dt>
<dd><p>Dimensions of the Universe to which the group belongs, at the current time step.</p>
</dd></dl>

<dl class="attribute">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup.forces">
<code class="descname">forces</code><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup.forces" title="Permalink to this definition">¶</a></dt>
<dd><p>Forces on the atoms in the AtomGroup.</p>
<p>The forces can be changed by assigning an array of the appropriate
shape, i.e. either Nx3 to assign individual force or 3, to assign
the <em>same</em> force to all atoms (e.g. <code class="docutils literal"><span class="pre">ag.forces</span> <span class="pre">=</span> <span class="pre">array([0,0,0])</span></code>
will set all forces to (0.,0.,0.)).</p>
<p>For more control use the <a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup.get_forces" title="MDAnalysis.core.AtomGroup.AtomGroup.get_forces"><code class="xref py py-meth docutils literal"><span class="pre">get_forces()</span></code></a> and
<a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup.set_forces" title="MDAnalysis.core.AtomGroup.AtomGroup.set_forces"><code class="xref py py-meth docutils literal"><span class="pre">set_forces()</span></code></a> methods.</p>
<div class="versionadded">
<p><span class="versionmodified">New in version 0.7.7.</span></p>
</div>
</dd></dl>

<dl class="attribute">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup.fragments">
<code class="descname">fragments</code><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup.fragments" title="Permalink to this definition">¶</a></dt>
<dd><p>Read-only list of fragments.</p>
<p>Contains all fragments that any Atom in this AtomGroup is part of, the contents of
the fragments may extend beyond the contents of this AtomGroup.</p>
</dd></dl>

<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup.get_forces">
<code class="descname">get_forces</code><span class="sig-paren">(</span><em>ts=None</em>, <em>copy=False</em>, <em>dtype=&lt;type 'numpy.float32'&gt;</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#AtomGroup.get_forces"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup.get_forces" title="Permalink to this definition">¶</a></dt>
<dd><p>Get a NumPy array of the atomic forces (if available).
Currently only supported for Gromacs .trr trajectories.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Keywords:</th><td class="field-body"><dl class="first last docutils">
<dt><em>ts</em></dt>
<dd><p class="first last">If <em>ts</em> is provided then positions are read from that
<a class="reference internal" href="../coordinates/base.html#MDAnalysis.coordinates.base.Timestep" title="MDAnalysis.coordinates.base.Timestep"><code class="xref py py-class docutils literal"><span class="pre">Timestep</span></code></a> instead of
the one from the current trajectory belonging to this universe.
The <em>ts</em> is indexed with the indices returned by
<a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup.indices" title="MDAnalysis.core.AtomGroup.AtomGroup.indices"><code class="xref py py-meth docutils literal"><span class="pre">indices()</span></code></a> and it is the user&#8217;s responsibility
to provide a time step that has the appropriate dimensions.</p>
</dd>
<dt><em>copy</em></dt>
<dd><p class="first last"><code class="docutils literal"><span class="pre">True</span></code>: always make a copy (slow), <code class="docutils literal"><span class="pre">False</span></code>: Try to
return a array view or reference (faster); note that for
passing coordinates to C-code it can be necessary to use
a copy [<code class="docutils literal"><span class="pre">False</span></code>]</p>
</dd>
<dt><em>dtype</em></dt>
<dd><p class="first last">NumPy Data type of the array; the default is usually
entirely appropriate. Most C-code actually requires the
default  [<code class="xref py py-class docutils literal"><span class="pre">numpy.float32</span></code>]</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
<p>Forces can also be directly obtained from the attribute
<a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup.forces" title="MDAnalysis.core.AtomGroup.AtomGroup.forces"><code class="xref py py-attr docutils literal"><span class="pre">forces</span></code></a>.</p>
<p>Forces can be directly set with <a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup.set_forces" title="MDAnalysis.core.AtomGroup.AtomGroup.set_forces"><code class="xref py py-meth docutils literal"><span class="pre">set_forces()</span></code></a> or
by assigning to <a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup.forces" title="MDAnalysis.core.AtomGroup.AtomGroup.forces"><code class="xref py py-attr docutils literal"><span class="pre">forces</span></code></a>.</p>
<div class="versionadded">
<p><span class="versionmodified">New in version 0.7.7.</span></p>
</div>
</dd></dl>

<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup.get_positions">
<code class="descname">get_positions</code><span class="sig-paren">(</span><em>ts=None</em>, <em>copy=False</em>, <em>dtype=&lt;type 'numpy.float32'&gt;</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#AtomGroup.get_positions"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup.get_positions" title="Permalink to this definition">¶</a></dt>
<dd><p>Get a NumPy array of the coordinates.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Keywords:</th><td class="field-body"><dl class="first last docutils">
<dt><em>ts</em></dt>
<dd><p class="first last">If <em>ts</em> is provided then positions are read from that
<a class="reference internal" href="../coordinates/base.html#MDAnalysis.coordinates.base.Timestep" title="MDAnalysis.coordinates.base.Timestep"><code class="xref py py-class docutils literal"><span class="pre">Timestep</span></code></a> instead of
the one from the current trajectory belonging to this universe.
The <em>ts</em> is indexed with the indices returned by
<a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup.indices" title="MDAnalysis.core.AtomGroup.AtomGroup.indices"><code class="xref py py-meth docutils literal"><span class="pre">indices()</span></code></a> and it is the user&#8217;s responsibility
to provide a time step that has the appropriate dimensions.</p>
</dd>
<dt><em>copy</em></dt>
<dd><p class="first last"><code class="docutils literal"><span class="pre">True</span></code>: always make a copy (slow), <code class="docutils literal"><span class="pre">False</span></code>: Try to
return a array view or reference (faster); note that for
passing coordinates to C-code it can be necessary to use
a copy [<code class="docutils literal"><span class="pre">False</span></code>]</p>
</dd>
<dt><em>dtype</em></dt>
<dd><p class="first last">NumPy Data type of the array; the default is usually
entirely appropriate. Most C-code actually requires the
default  [<code class="xref py py-class docutils literal"><span class="pre">numpy.float32</span></code>]</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
<p>Coordinates can also be directly obtained from the attribute
<a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup.positions" title="MDAnalysis.core.AtomGroup.AtomGroup.positions"><code class="xref py py-attr docutils literal"><span class="pre">positions</span></code></a>.</p>
<p>Coordinates can be directly set with <a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup.set_positions" title="MDAnalysis.core.AtomGroup.AtomGroup.set_positions"><code class="xref py py-meth docutils literal"><span class="pre">set_positions()</span></code></a> or
by assigning to <a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup.positions" title="MDAnalysis.core.AtomGroup.AtomGroup.positions"><code class="xref py py-attr docutils literal"><span class="pre">positions</span></code></a>.</p>
<p>This method is identical with <a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup.coordinates" title="MDAnalysis.core.AtomGroup.AtomGroup.coordinates"><code class="xref py py-meth docutils literal"><span class="pre">coordinates()</span></code></a> but named
differently for symmetry with with <a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup.set_positions" title="MDAnalysis.core.AtomGroup.AtomGroup.set_positions"><code class="xref py py-meth docutils literal"><span class="pre">set_positions()</span></code></a>.</p>
<div class="versionadded">
<p><span class="versionmodified">New in version 0.7.6.</span></p>
</div>
</dd></dl>

<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup.get_velocities">
<code class="descname">get_velocities</code><span class="sig-paren">(</span><em>ts=None</em>, <em>copy=False</em>, <em>dtype=&lt;type 'numpy.float32'&gt;</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#AtomGroup.get_velocities"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup.get_velocities" title="Permalink to this definition">¶</a></dt>
<dd><p>NumPy array of the velocities.</p>
<p>Raises a <a class="reference internal" href="#MDAnalysis.core.AtomGroup.NoDataError" title="MDAnalysis.core.AtomGroup.NoDataError"><code class="xref py py-exc docutils literal"><span class="pre">NoDataError</span></code></a> if the underlying
<a class="reference internal" href="../coordinates/base.html#MDAnalysis.coordinates.base.Timestep" title="MDAnalysis.coordinates.base.Timestep"><code class="xref py py-class docutils literal"><span class="pre">Timestep</span></code></a> does not contain
<a class="reference internal" href="../coordinates/base.html#MDAnalysis.coordinates.base.Timestep._velocities" title="MDAnalysis.coordinates.base.Timestep._velocities"><code class="xref py py-attr docutils literal"><span class="pre">_velocities</span></code></a>.</p>
<p>See also <a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup.set_velocities" title="MDAnalysis.core.AtomGroup.AtomGroup.set_velocities"><code class="xref py py-meth docutils literal"><span class="pre">AtomGroup.set_velocities()</span></code></a> and attribute access through
<a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup.velocities" title="MDAnalysis.core.AtomGroup.AtomGroup.velocities"><code class="xref py py-attr docutils literal"><span class="pre">AtomGroup.velocities</span></code></a>.</p>
<div class="versionadded">
<p><span class="versionmodified">New in version 0.7.6.</span></p>
</div>
</dd></dl>

<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup.guess_bonds">
<code class="descname">guess_bonds</code><span class="sig-paren">(</span><em>vdwradii=None</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#AtomGroup.guess_bonds"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup.guess_bonds" title="Permalink to this definition">¶</a></dt>
<dd><p>Guess all the bonds that exist within this AtomGroup and add to Universe.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Keywords:</th><td class="field-body"><dl class="first last docutils">
<dt><em>vdwradii</em></dt>
<dd><p class="first last">Pass a dict relating atom types to vdwradii.</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="../topology/core.html#MDAnalysis.topology.core.guess_bonds" title="MDAnalysis.topology.core.guess_bonds"><code class="xref py py-func docutils literal"><span class="pre">MDAnalysis.topology.core.guess_bonds()</span></code></a></p>
</div>
<div class="versionadded">
<p><span class="versionmodified">New in version 0.10.0.</span></p>
</div>
</dd></dl>

<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup.improper">
<code class="descname">improper</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#AtomGroup.improper"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup.improper" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the improper dihedral between 4 atoms.</p>
<p>The improper dihedral is calculated in the same way as the proper
<a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup.dihedral" title="MDAnalysis.core.AtomGroup.AtomGroup.dihedral"><code class="xref py py-meth docutils literal"><span class="pre">dihedral()</span></code></a>: The angle between the planes formed by atoms (0,1,2)
and (1,2,3).</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Only makes sense for a <a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup" title="MDAnalysis.core.AtomGroup.AtomGroup"><code class="xref py py-class docutils literal"><span class="pre">AtomGroup</span></code></a> with exactly 4
<a class="reference internal" href="#MDAnalysis.core.AtomGroup.Atom" title="MDAnalysis.core.AtomGroup.Atom"><code class="xref py py-class docutils literal"><span class="pre">Atom</span></code></a>; anything else will raise a <code class="xref py py-exc docutils literal"><span class="pre">ValueError</span></code>. The
interpretation of the angle as an &#8220;improper&#8221; solely depends on the
selection of atoms and thus the user input!</p>
</div>
<div class="versionadded">
<p><span class="versionmodified">New in version 0.7.3.</span></p>
</div>
</dd></dl>

<dl class="attribute">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup.impropers">
<code class="descname">impropers</code><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup.impropers" title="Permalink to this definition">¶</a></dt>
<dd><p>All the improper torsions in this AtomGroup</p>
<p>Note that these improper torsions might extend out of the AtomGroup,
to select only torsions which are entirely contained by the AtomGroup use
u.impropers.atomgroup_intersection(ag, strict=True)</p>
<div class="versionadded">
<p><span class="versionmodified">New in version 0.9.0.</span></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 0.10.0: </span>Now always returns a (possibly empty) TopologyGroup</p>
</div>
</dd></dl>

<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup.indices">
<code class="descname">indices</code><span class="sig-paren">(</span><em>*args</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#AtomGroup.indices"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup.indices" title="Permalink to this definition">¶</a></dt>
<dd><p>Array of all <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Atom.number" title="MDAnalysis.core.AtomGroup.Atom.number"><code class="xref py py-attr docutils literal"><span class="pre">Atom.number</span></code></a> in the group.</p>
<p>These indices are 0-based and can be used to directly index
<code class="xref py py-attr docutils literal"><span class="pre">Universe.atoms</span></code> or the coordinate array
<a class="reference internal" href="../coordinates/base.html#MDAnalysis.coordinates.base.Timestep._pos" title="MDAnalysis.coordinates.base.Timestep._pos"><code class="xref py py-attr docutils literal"><span class="pre">MDAnalysis.coordinates.base.Timestep._pos</span></code></a>.</p>
</dd></dl>

<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup.masses">
<code class="descname">masses</code><span class="sig-paren">(</span><em>*args</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#AtomGroup.masses"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup.masses" title="Permalink to this definition">¶</a></dt>
<dd><p>Array of atomic masses (as defined in the topology)</p>
</dd></dl>

<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup.momentOfInertia">
<code class="descname">momentOfInertia</code><span class="sig-paren">(</span><em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#AtomGroup.momentOfInertia"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup.momentOfInertia" title="Permalink to this definition">¶</a></dt>
<dd><p>Tensor of inertia as 3x3 NumPy array.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Keywords:</th><td class="field-body"><dl class="first last docutils">
<dt><em>pbc</em></dt>
<dd><p class="first last"><code class="docutils literal"><span class="pre">True</span></code>: Move all atoms within the primary unit cell before calculation [<code class="docutils literal"><span class="pre">False</span></code>]</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">The <code class="xref py py-class docutils literal"><span class="pre">MDAnalysis.core.flags</span></code> flag <em>use_pbc</em> when set to <code class="docutils literal"><span class="pre">True</span></code> allows the <em>pbc</em>
flag to be used by default.</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 0.8: </span>Added <em>pbc</em> keyword</p>
</div>
</dd></dl>

<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup.names">
<code class="descname">names</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#AtomGroup.names"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup.names" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a list of atom names.</p>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 0.8: </span>Returns a <a class="reference external" href="http://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.9)"><code class="xref py py-class docutils literal"><span class="pre">numpy.ndarray</span></code></a></p>
</div>
</dd></dl>

<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup.numberOfAtoms">
<code class="descname">numberOfAtoms</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#AtomGroup.numberOfAtoms"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup.numberOfAtoms" title="Permalink to this definition">¶</a></dt>
<dd><p>Total number of atoms in the group</p>
</dd></dl>

<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup.numberOfResidues">
<code class="descname">numberOfResidues</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#AtomGroup.numberOfResidues"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup.numberOfResidues" title="Permalink to this definition">¶</a></dt>
<dd><p>Total number of residues in the group</p>
</dd></dl>

<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup.numberOfSegments">
<code class="descname">numberOfSegments</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#AtomGroup.numberOfSegments"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup.numberOfSegments" title="Permalink to this definition">¶</a></dt>
<dd><p>Total number of segments in the group</p>
</dd></dl>

<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup.packIntoBox">
<code class="descname">packIntoBox</code><span class="sig-paren">(</span><em>box=None</em>, <em>inplace=True</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#AtomGroup.packIntoBox"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup.packIntoBox" title="Permalink to this definition">¶</a></dt>
<dd><p>Shift all atoms in this group to be within the primary unit cell.</p>
<p>AtomGroup.packintobox([box, [inplace=True]])</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Keywords:</th><td class="field-body"><dl class="first last docutils">
<dt><em>box</em></dt>
<dd><p class="first last">Unit cell to move atoms inside of.</p>
</dd>
<dt><em>inplace</em></dt>
<dd><p class="first last"><code class="docutils literal"><span class="pre">True</span></code>: Change coordinates in place and return
<code class="docutils literal"><span class="pre">False</span></code>: Only return the coordinates</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
<p>All atoms will be moved so that they lie between 0 and
boxlength <span class="math">\(L_i\)</span> in all dimensions, i.e. the lower left
corner of the simulation box is taken to be at (0,0,0):</p>
<div class="math">
\[x_i' = x_i - \left\lfloor\frac{x_i}{L_i}\right\rfloor\]</div>
<p>The default is to take unit cell information from the
underlying <a class="reference internal" href="../coordinates/base.html#MDAnalysis.coordinates.base.Timestep" title="MDAnalysis.coordinates.base.Timestep"><code class="xref py py-class docutils literal"><span class="pre">Timestep</span></code></a>
instance. The optional argument <em>box</em> can be used to provide
alternative unit cell information (in the MDAnalysis standard
format <code class="docutils literal"><span class="pre">[Lx,</span> <span class="pre">Ly,</span> <span class="pre">Lz,</span> <span class="pre">alpha,</span> <span class="pre">beta,</span> <span class="pre">gamma]</span></code>).</p>
<p>Works with either orthogonal or triclinic box types.</p>
<p>By default the coordinates are changed in place and returned</p>
<div class="versionadded">
<p><span class="versionmodified">New in version 0.8.</span></p>
</div>
</dd></dl>

<dl class="attribute">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup.positions">
<code class="descname">positions</code><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup.positions" title="Permalink to this definition">¶</a></dt>
<dd><p>Coordinates of the atoms in the AtomGroup.</p>
<p>The positions can be changed by assigning an array of the appropriate
shape, i.e. either Nx3 to assign individual coordinates or 3, to assign
the <em>same</em> coordinate to all atoms (e.g. <code class="docutils literal"><span class="pre">ag.positions</span> <span class="pre">=</span> <span class="pre">array([0,0,0])</span></code>
will move all particles to the origin).</p>
<p>For more control use the <a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup.get_positions" title="MDAnalysis.core.AtomGroup.AtomGroup.get_positions"><code class="xref py py-meth docutils literal"><span class="pre">get_positions()</span></code></a> and
<a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup.set_positions" title="MDAnalysis.core.AtomGroup.AtomGroup.set_positions"><code class="xref py py-meth docutils literal"><span class="pre">set_positions()</span></code></a> methods.</p>
<div class="versionadded">
<p><span class="versionmodified">New in version 0.7.6.</span></p>
</div>
</dd></dl>

<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup.principalAxes">
<code class="descname">principalAxes</code><span class="sig-paren">(</span><em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#AtomGroup.principalAxes"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup.principalAxes" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate the principal axes from the moment of inertia.</p>
<p>e1,e2,e3 = AtomGroup.principalAxes()</p>
<p>The eigenvectors are sorted by eigenvalue, i.e. the first one
corresponds to the highest eigenvalue and is thus the first principal axes.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Keywords:</th><td class="field-body"><dl class="first last docutils">
<dt><em>pbc</em></dt>
<dd><p class="first last"><code class="docutils literal"><span class="pre">True</span></code>: Move all atoms within primary unit cell before calculation</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">The <code class="xref py py-class docutils literal"><span class="pre">MDAnalysis.core.flags</span></code> flag <em>use_pbc</em> when set to <code class="docutils literal"><span class="pre">True</span></code> allows the <em>pbc</em>
flag to be used by default.</p>
</div>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">numpy.array <code class="docutils literal"><span class="pre">v</span></code> with <code class="docutils literal"><span class="pre">v[0]</span></code> as first, <code class="docutils literal"><span class="pre">v[1]</span></code> as second,
and <code class="docutils literal"><span class="pre">v[2]</span></code> as third eigenvector.</td>
</tr>
</tbody>
</table>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 0.8: </span>Added <em>pbc</em> keyword</p>
</div>
</dd></dl>

<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup.radii">
<code class="descname">radii</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#AtomGroup.radii"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup.radii" title="Permalink to this definition">¶</a></dt>
<dd><p>Array of atomic radii (as defined in the PQR file)</p>
</dd></dl>

<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup.radiusOfGyration">
<code class="descname">radiusOfGyration</code><span class="sig-paren">(</span><em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#AtomGroup.radiusOfGyration"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup.radiusOfGyration" title="Permalink to this definition">¶</a></dt>
<dd><p>Radius of gyration.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Keywords:</th><td class="field-body"><dl class="first last docutils">
<dt><em>pbc</em></dt>
<dd><p class="first last"><code class="docutils literal"><span class="pre">True</span></code>: Move all atoms within the primary unit cell before calculation [<code class="docutils literal"><span class="pre">False</span></code>]</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">The <code class="xref py py-class docutils literal"><span class="pre">MDAnalysis.core.flags</span></code> flag <em>use_pbc</em> when set to <code class="docutils literal"><span class="pre">True</span></code> allows the <em>pbc</em>
flag to be used by default.</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 0.8: </span>Added <em>pbc</em> keyword</p>
</div>
</dd></dl>

<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup.resids">
<code class="descname">resids</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#AtomGroup.resids"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup.resids" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a list of residue numbers.</p>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 0.8: </span>Returns a <a class="reference external" href="http://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.9)"><code class="xref py py-class docutils literal"><span class="pre">numpy.ndarray</span></code></a></p>
</div>
</dd></dl>

<dl class="attribute">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup.residues">
<code class="descname">residues</code><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup.residues" title="Permalink to this definition">¶</a></dt>
<dd><p>Read-only list of <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Residue" title="MDAnalysis.core.AtomGroup.Residue"><code class="xref py py-class docutils literal"><span class="pre">Residue</span></code></a> objects.</p>
<p>A <a class="reference internal" href="#MDAnalysis.core.AtomGroup.ResidueGroup" title="MDAnalysis.core.AtomGroup.ResidueGroup"><code class="xref py py-class docutils literal"><span class="pre">ResidueGroup</span></code></a> of all residues that contain atoms in
this group.</p>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 0.9.0: </span>Now returns strictly a ResidueGroup of the unique Residues that Atoms in this group
belong to.</p>
</div>
</dd></dl>

<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup.resnames">
<code class="descname">resnames</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#AtomGroup.resnames"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup.resnames" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a list of residue names.</p>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 0.8: </span>Returns a <a class="reference external" href="http://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.9)"><code class="xref py py-class docutils literal"><span class="pre">numpy.ndarray</span></code></a></p>
</div>
</dd></dl>

<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup.resnums">
<code class="descname">resnums</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#AtomGroup.resnums"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup.resnums" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a list of canonical residue numbers.</p>
<div class="versionadded">
<p><span class="versionmodified">New in version 0.7.4.</span></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 0.8: </span>Returns a <a class="reference external" href="http://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.9)"><code class="xref py py-class docutils literal"><span class="pre">numpy.ndarray</span></code></a></p>
</div>
</dd></dl>

<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup.rotate">
<code class="descname">rotate</code><span class="sig-paren">(</span><em>R</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#AtomGroup.rotate"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup.rotate" title="Permalink to this definition">¶</a></dt>
<dd><p>Apply a rotation matrix <em>R</em> to the selection&#8217;s coordinates.</p>
<p>AtomGroup.rotate(R)</p>
<p><span class="math">\(\mathsf{R}\)</span> is a 3x3 orthogonal matrix that transforms a vector
<span class="math">\(\mathbf{x} \rightarrow \mathbf{x}'\)</span>:</p>
<div class="math">
\[\mathbf{x}' = \mathsf{R}\mathbf{x}\]</div>
</dd></dl>

<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup.rotateby">
<code class="descname">rotateby</code><span class="sig-paren">(</span><em>angle</em>, <em>axis</em>, <em>point=None</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#AtomGroup.rotateby"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup.rotateby" title="Permalink to this definition">¶</a></dt>
<dd><p>Apply a rotation to the selection&#8217;s coordinates.</p>
<p>AtomGroup.rotateby(angle,axis[,point])</p>
<p>The transformation from current coordinates <span class="math">\(\mathbf{x}\)</span>
to new coordinates <span class="math">\(\mathbf{x}'\)</span> is</p>
<div class="math">
\[\mathbf{x}' = \mathsf{R}\,(\mathbf{x}-\mathbf{p}) + \mathbf{p}\]</div>
<p>where <span class="math">\(\mathsf{R}\)</span> is the rotation by <em>angle</em> around the
<em>axis</em> going through <em>point</em> <span class="math">\(\mathbf{p}\)</span>.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Arguments:</th><td class="field-body"><dl class="first docutils">
<dt><em>angle</em></dt>
<dd><p class="first last">rotation angle in degrees</p>
</dd>
<dt><em>axis</em></dt>
<dd><p class="first last">rotation axis vector, a 3-tuple, list, or array, or a 2-tuple of
two MDAnalysis objects from which the axis is calculated as the
vector from the first to the second center of geometry.</p>
</dd>
<dt><em>point</em></dt>
<dd><p class="first last">point on the rotation axis; by default (<code class="docutils literal"><span class="pre">None</span></code>) the center of
geometry of the selection is chosen, or, if <em>axis</em> is a tuple of
selections, it defaults to the first point of the axis. <em>point</em>
can be a 3-tuple, list, or array or a MDAnalysis object (in which
case its <a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup.centroid" title="MDAnalysis.core.AtomGroup.AtomGroup.centroid"><code class="xref py py-meth docutils literal"><span class="pre">centroid()</span></code></a> is used).</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">The 4x4 matrix which consists of the rotation matrix <code class="docutils literal"><span class="pre">M[:3,:3]</span></code>
and the translation vector <code class="docutils literal"><span class="pre">M[:3,3]</span></code>.</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup.segids">
<code class="descname">segids</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#AtomGroup.segids"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup.segids" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a list of segment ids (=segment names).</p>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 0.8: </span>Returns a <a class="reference external" href="http://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.9)"><code class="xref py py-class docutils literal"><span class="pre">numpy.ndarray</span></code></a></p>
</div>
</dd></dl>

<dl class="attribute">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup.segments">
<code class="descname">segments</code><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup.segments" title="Permalink to this definition">¶</a></dt>
<dd><p>Read-only list of <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Segment" title="MDAnalysis.core.AtomGroup.Segment"><code class="xref py py-class docutils literal"><span class="pre">Segment</span></code></a> objects.</p>
<p>A <a class="reference internal" href="#MDAnalysis.core.AtomGroup.SegmentGroup" title="MDAnalysis.core.AtomGroup.SegmentGroup"><code class="xref py py-class docutils literal"><span class="pre">SegmentGroup</span></code></a> of all segments that contain atoms in this group.</p>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 0.9.0: </span>Now strictly returns a <a class="reference internal" href="#MDAnalysis.core.AtomGroup.SegmentGroup" title="MDAnalysis.core.AtomGroup.SegmentGroup"><code class="xref py py-class docutils literal"><span class="pre">SegmentGroup</span></code></a> of a set of
the <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Segment" title="MDAnalysis.core.AtomGroup.Segment"><code class="xref py py-class docutils literal"><span class="pre">Segment</span></code></a> instances from this <a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup" title="MDAnalysis.core.AtomGroup.AtomGroup"><code class="xref py py-class docutils literal"><span class="pre">AtomGroup</span></code></a></p>
</div>
</dd></dl>

<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup.selectAtoms">
<code class="descname">selectAtoms</code><span class="sig-paren">(</span><em>sel</em>, <em>*othersel</em>, <em>**selgroups</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#AtomGroup.selectAtoms"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup.selectAtoms" title="Permalink to this definition">¶</a></dt>
<dd><p>Selection of atoms using the MDAnalysis selection syntax.</p>
<p>AtomGroup.selectAtoms(selection[,selection[,...]], [groupname=atomgroup[,groupname=atomgroup[,...]]])</p>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="#MDAnalysis.core.AtomGroup.Universe.selectAtoms" title="MDAnalysis.core.AtomGroup.Universe.selectAtoms"><code class="xref py py-meth docutils literal"><span class="pre">Universe.selectAtoms()</span></code></a></p>
</div>
</dd></dl>

<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup.sequence">
<code class="descname">sequence</code><span class="sig-paren">(</span><em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#AtomGroup.sequence"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup.sequence" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the amino acid sequence.</p>
<p>The format of the sequence is selected with the keyword <em>format</em>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="24%" />
<col width="76%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head"><em>format</em></th>
<th class="head">description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>&#8216;SeqRecord&#8217;</td>
<td><code class="xref py py-class docutils literal"><span class="pre">Bio.SeqRecord.SeqRecord</span></code> (default)</td>
</tr>
<tr class="row-odd"><td>&#8216;Seq&#8217;</td>
<td><code class="xref py py-class docutils literal"><span class="pre">Bio.Seq.Seq</span></code></td>
</tr>
<tr class="row-even"><td>&#8216;string&#8217;</td>
<td>string</td>
</tr>
</tbody>
</table>
<p>The sequence is returned by default (keyword <code class="docutils literal"><span class="pre">format</span> <span class="pre">=</span> <span class="pre">'SeqRecord'</span></code>)
as a <code class="xref py py-class docutils literal"><span class="pre">Bio.SeqRecord.SeqRecord</span></code> instance, which can then be
further processed. In this case, all keyword arguments (such as the
<em>id</em> string or the <em>name</em> or the <em>description</em>) are directly passed to
<code class="xref py py-class docutils literal"><span class="pre">Bio.SeqRecord.SeqRecord</span></code>.</p>
<p>If the keyword <em>format</em> is set to <code class="docutils literal"><span class="pre">'Seq'</span></code>, all <em>kwargs</em> are ignored and
a <code class="xref py py-class docutils literal"><span class="pre">Bio.Seq.Seq</span></code> instance is returned. The difference to the
record is that the record also contains metadata and can be directly
used as an input for other functions in <code class="xref py py-mod docutils literal"><span class="pre">Bio</span></code>.</p>
<p>If the keyword <em>format</em> is set to <code class="docutils literal"><span class="pre">'string'</span></code>, all <em>kwargs</em> are ignored
and a Python string is returned.</p>
<p class="rubric">Example: Write FASTA file</p>
<p>Use <code class="xref py py-func docutils literal"><span class="pre">Bio.SeqIO.write()</span></code>, which takes sequence records:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">import</span> <span class="nn">Bio.SeqIO</span>

<span class="c"># get the sequence record of a protein component of a Universe</span>
<span class="n">protein</span> <span class="o">=</span> <span class="n">u</span><span class="o">.</span><span class="n">selectAtoms</span><span class="p">(</span><span class="s">&quot;protein&quot;</span><span class="p">)</span>
<span class="n">record</span> <span class="o">=</span> <span class="n">protein</span><span class="o">.</span><span class="n">sequence</span><span class="p">(</span><span class="nb">id</span><span class="o">=</span><span class="s">&quot;myseq1&quot;</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s">&quot;myprotein&quot;</span><span class="p">)</span>

<span class="n">Bio</span><span class="o">.</span><span class="n">SeqIO</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">record</span><span class="p">,</span> <span class="s">&quot;single.fasta&quot;</span><span class="p">,</span> <span class="s">&quot;fasta&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>A FASTA file with multiple entries can be written with</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">Bio</span><span class="o">.</span><span class="n">SeqIO</span><span class="o">.</span><span class="n">write</span><span class="p">([</span><span class="n">record1</span><span class="p">,</span> <span class="n">record2</span><span class="p">,</span> <span class="o">...</span><span class="p">],</span> <span class="s">&quot;multi.fasta&quot;</span><span class="p">,</span> <span class="s">&quot;fasta&quot;</span><span class="p">)</span>
</pre></div>
</div>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Keywords:</th><td class="field-body"><dl class="first docutils">
<dt><em>format</em></dt>
<dd><blockquote class="first">
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">&quot;string&quot;</span></code>: return sequence as a string of 1-letter codes</li>
<li><code class="docutils literal"><span class="pre">&quot;Seq&quot;</span></code>: return a <code class="xref py py-class docutils literal"><span class="pre">Bio.Seq.Seq</span></code> instance</li>
<li><code class="docutils literal"><span class="pre">&quot;SeqRecord&quot;</span></code>: return a <code class="xref py py-class docutils literal"><span class="pre">Bio.SeqRecord.SeqRecord</span></code>
instance</li>
</ul>
<p>Default is <code class="docutils literal"><span class="pre">&quot;SeqRecord&quot;</span></code></p>
</div></blockquote>
<dl class="last docutils">
<dt><em>id</em></dt>
<dd><p class="first last">Sequence ID for SeqRecord (should be different for different
sequences)</p>
</dd>
<dt><em>name</em></dt>
<dd><p class="first last">Name of the protein.</p>
</dd>
<dt><em>description</em></dt>
<dd><p class="first last">Short description of the sequence.</p>
</dd>
<dt><em>kwargs</em></dt>
<dd><p class="first last">Any other keyword arguments that are understood by
<code class="xref py py-class docutils literal"><span class="pre">Bio.SeqRecord.SeqRecord</span></code>.</p>
</dd>
</dl>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Raises:</th><td class="field-body"><p class="first last"><code class="xref py py-exc docutils literal"><span class="pre">ValueError</span></code> if a residue name cannot be converted to a
1-letter IUPAC protein amino acid code; make sure to only
select protein residues. Raises <code class="xref py py-exc docutils literal"><span class="pre">TypeError</span></code> if an unknown
<em>format</em> is selected.</p>
</td>
</tr>
</tbody>
</table>
<div class="versionadded">
<p><span class="versionmodified">New in version 0.9.0.</span></p>
</div>
</dd></dl>

<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup.set">
<code class="descname">set</code><span class="sig-paren">(</span><em>name</em>, <em>value</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup.set" title="Permalink to this definition">¶</a></dt>
<dd><p>Set attribute <em>name</em> to <em>value</em> for all atoms in the <a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup" title="MDAnalysis.core.AtomGroup.AtomGroup"><code class="xref py py-class docutils literal"><span class="pre">AtomGroup</span></code></a>.</p>
<p>If <em>value</em> is a sequence of the same length as the <a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup" title="MDAnalysis.core.AtomGroup.AtomGroup"><code class="xref py py-class docutils literal"><span class="pre">AtomGroup</span></code></a>
then each <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Atom" title="MDAnalysis.core.AtomGroup.Atom"><code class="xref py py-class docutils literal"><span class="pre">Atom</span></code></a>&#8216;s property <em>name</em> is set to the corresponding value. If
<em>value</em> is neither of length 1 (or a scalar) nor of the length of the
<a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup" title="MDAnalysis.core.AtomGroup.AtomGroup"><code class="xref py py-class docutils literal"><span class="pre">AtomGroup</span></code></a> then a <code class="xref py py-exc docutils literal"><span class="pre">ValueError</span></code> is raised.</p>
<div class="versionadded">
<p><span class="versionmodified">New in version 0.7.4.</span></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 0.8: </span>Can set atoms to distinct values by providing a sequence or iterable.</p>
</div>
</dd></dl>

<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup.set_bfactor">
<code class="descname">set_bfactor</code><span class="sig-paren">(</span><em>bfactor</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#AtomGroup.set_bfactor"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup.set_bfactor" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the atom bfactor to float <em>bfactor</em> for <strong>all atoms</strong> in the AtomGroup.</p>
<div class="versionadded">
<p><span class="versionmodified">New in version 0.7.4.</span></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 0.8: </span>Can set atoms and residues to distinct values by providing a sequence or iterable.</p>
</div>
</dd></dl>

<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup.set_charge">
<code class="descname">set_charge</code><span class="sig-paren">(</span><em>charge</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#AtomGroup.set_charge"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup.set_charge" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the partial charge to float <em>charge</em> for <strong>all atoms</strong> in the AtomGroup.</p>
<div class="versionadded">
<p><span class="versionmodified">New in version 0.7.4.</span></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 0.8: </span>Can set atoms and residues to distinct values by providing a sequence or iterable.</p>
</div>
</dd></dl>

<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup.set_forces">
<code class="descname">set_forces</code><span class="sig-paren">(</span><em>forces</em>, <em>ts=None</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#AtomGroup.set_forces"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup.set_forces" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the forces for all atoms in the group.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Arguments:</th><td class="field-body"><dl class="first docutils">
<dt><em>forces</em></dt>
<dd><p class="first last">a Nx3 NumPy <a class="reference external" href="http://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.9)"><code class="xref py py-class docutils literal"><span class="pre">numpy.ndarray</span></code></a> where N is the number of
atoms in this atom group.</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Keywords:</th><td class="field-body"><dl class="first last docutils">
<dt><em>ts</em></dt>
<dd><p class="first last"><a class="reference internal" href="../coordinates/base.html#MDAnalysis.coordinates.base.Timestep" title="MDAnalysis.coordinates.base.Timestep"><code class="xref py py-class docutils literal"><span class="pre">Timestep</span></code></a>, defaults
to <code class="docutils literal"><span class="pre">None</span></code> and then the current time step is used.</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">If the group contains N atoms and <em>force</em> is a single vector (i.e. an
array of length 3) then all N atom positions are set to <em>force</em> (due
to NumPy&#8217;s broadcasting rules), as described for
<a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup.forces" title="MDAnalysis.core.AtomGroup.AtomGroup.forces"><code class="xref py py-attr docutils literal"><span class="pre">forces</span></code></a>.</p>
</div>
<p>See also <a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup.get_forces" title="MDAnalysis.core.AtomGroup.AtomGroup.get_forces"><code class="xref py py-meth docutils literal"><span class="pre">get_forces()</span></code></a> and attribute access through
<a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup.forces" title="MDAnalysis.core.AtomGroup.AtomGroup.forces"><code class="xref py py-attr docutils literal"><span class="pre">forces</span></code></a>.</p>
<div class="versionadded">
<p><span class="versionmodified">New in version 0.7.7.</span></p>
</div>
</dd></dl>

<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup.set_mass">
<code class="descname">set_mass</code><span class="sig-paren">(</span><em>mass</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#AtomGroup.set_mass"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup.set_mass" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the atom mass to float <em>mass</em> for <strong>all atoms</strong> in the AtomGroup.</p>
<div class="versionadded">
<p><span class="versionmodified">New in version 0.7.4.</span></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 0.8: </span>Can set atoms and residues to distinct values by providing a sequence or iterable.</p>
</div>
</dd></dl>

<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup.set_name">
<code class="descname">set_name</code><span class="sig-paren">(</span><em>name</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#AtomGroup.set_name"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup.set_name" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the atom name to string for <em>all atoms</em> in the AtomGroup.</p>
<p>If <em>value</em> is a sequence of the same length as the <a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup" title="MDAnalysis.core.AtomGroup.AtomGroup"><code class="xref py py-class docutils literal"><span class="pre">AtomGroup</span></code></a>
then each <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Atom.name" title="MDAnalysis.core.AtomGroup.Atom.name"><code class="xref py py-attr docutils literal"><span class="pre">Atom.name</span></code></a> is set to the corresponding value. If
<em>value</em> is neither of length 1 (or a scalar) nor of the length of the
<a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup" title="MDAnalysis.core.AtomGroup.AtomGroup"><code class="xref py py-class docutils literal"><span class="pre">AtomGroup</span></code></a> then a <code class="xref py py-exc docutils literal"><span class="pre">ValueError</span></code> is raised.</p>
<div class="versionadded">
<p><span class="versionmodified">New in version 0.7.4.</span></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 0.8: </span>Can set atoms to distinct values by providing a sequence or iterable.</p>
</div>
</dd></dl>

<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup.set_positions">
<code class="descname">set_positions</code><span class="sig-paren">(</span><em>coords</em>, <em>ts=None</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#AtomGroup.set_positions"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup.set_positions" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the positions for all atoms in the group.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Arguments:</th><td class="field-body"><dl class="first docutils">
<dt><em>coords</em></dt>
<dd><p class="first last">a Nx3 NumPy <a class="reference external" href="http://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.9)"><code class="xref py py-class docutils literal"><span class="pre">numpy.ndarray</span></code></a> where N is the number of
atoms in this atom group.</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Keywords:</th><td class="field-body"><dl class="first last docutils">
<dt><em>ts</em></dt>
<dd><p class="first last"><a class="reference internal" href="../coordinates/base.html#MDAnalysis.coordinates.base.Timestep" title="MDAnalysis.coordinates.base.Timestep"><code class="xref py py-class docutils literal"><span class="pre">Timestep</span></code></a>, defaults
to <code class="docutils literal"><span class="pre">None</span></code> and then the current time step is used.</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">If the group contains N atoms and <em>coord</em> is a single point (i.e. an
array of length 3) then all N atom positions are set to <em>coord</em> (due
to NumPy&#8217;s broadcasting rules), as described for
<a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup.positions" title="MDAnalysis.core.AtomGroup.AtomGroup.positions"><code class="xref py py-attr docutils literal"><span class="pre">positions</span></code></a>.</p>
</div>
<p>See also <a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup.get_positions" title="MDAnalysis.core.AtomGroup.AtomGroup.get_positions"><code class="xref py py-meth docutils literal"><span class="pre">get_positions()</span></code></a> and attribute access through
<a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup.positions" title="MDAnalysis.core.AtomGroup.AtomGroup.positions"><code class="xref py py-attr docutils literal"><span class="pre">positions</span></code></a>.</p>
<div class="versionadded">
<p><span class="versionmodified">New in version 0.7.6.</span></p>
</div>
</dd></dl>

<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup.set_radius">
<code class="descname">set_radius</code><span class="sig-paren">(</span><em>radius</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#AtomGroup.set_radius"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup.set_radius" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the atom radius to float <em>radius</em> for <strong>all atoms</strong> in the AtomGroup.</p>
<div class="versionadded">
<p><span class="versionmodified">New in version 0.7.4.</span></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 0.8: </span>Can set atoms and residues to distinct values by providing a sequence or iterable.</p>
</div>
</dd></dl>

<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup.set_resid">
<code class="descname">set_resid</code><span class="sig-paren">(</span><em>resid</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#AtomGroup.set_resid"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup.set_resid" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the resid to integer <em>resid</em> for <strong>all atoms</strong> in the <a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup" title="MDAnalysis.core.AtomGroup.AtomGroup"><code class="xref py py-class docutils literal"><span class="pre">AtomGroup</span></code></a>.</p>
<p>If <em>resid</em> is a sequence of the same length as the <a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup" title="MDAnalysis.core.AtomGroup.AtomGroup"><code class="xref py py-class docutils literal"><span class="pre">AtomGroup</span></code></a>
then each <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Atom.resid" title="MDAnalysis.core.AtomGroup.Atom.resid"><code class="xref py py-attr docutils literal"><span class="pre">Atom.resid</span></code></a> is set to the corresponding value together
with the <code class="xref py py-attr docutils literal"><span class="pre">Residue.id</span></code> of the residue the atom belongs to. If
<em>value</em> is neither of length 1 (or a scalar) nor of the length of the
<a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup" title="MDAnalysis.core.AtomGroup.AtomGroup"><code class="xref py py-class docutils literal"><span class="pre">AtomGroup</span></code></a> then a <code class="xref py py-exc docutils literal"><span class="pre">ValueError</span></code> is raised.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p>Changing resids can change the topology.</p>
<p class="last">Assigning the same <em>resid</em> to multiple residues will
<strong>merge</strong> these residues. Assigning different <em>resid</em> to
atoms in the same residue will <strong>split</strong> a residue (and
potentially merge with another one).</p>
</div>
<div class="versionadded">
<p><span class="versionmodified">New in version 0.7.4.</span></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 0.7.5: </span>Also changes the residues.</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 0.8: </span>Can set atoms and residues to distinct values by providing a
sequence or iterable and can change the topology via
<a class="reference internal" href="../topology/core.html#MDAnalysis.topology.core.build_residues" title="MDAnalysis.topology.core.build_residues"><code class="xref py py-func docutils literal"><span class="pre">MDAnalysis.topology.core.build_residues()</span></code></a>.</p>
</div>
</dd></dl>

<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup.set_resname">
<code class="descname">set_resname</code><span class="sig-paren">(</span><em>resname</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#AtomGroup.set_resname"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup.set_resname" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the resname to string <em>resname</em> for <strong>all atoms</strong> in the <a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup" title="MDAnalysis.core.AtomGroup.AtomGroup"><code class="xref py py-class docutils literal"><span class="pre">AtomGroup</span></code></a>.</p>
<p>If <em>resname</em> is a sequence of the same length as the <a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup" title="MDAnalysis.core.AtomGroup.AtomGroup"><code class="xref py py-class docutils literal"><span class="pre">AtomGroup</span></code></a>
then each <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Atom.resname" title="MDAnalysis.core.AtomGroup.Atom.resname"><code class="xref py py-attr docutils literal"><span class="pre">Atom.resname</span></code></a> is set to the corresponding value together
with the <code class="xref py py-attr docutils literal"><span class="pre">Residue.name</span></code> of the residue the atom belongs to. If
<em>value</em> is neither of length 1 (or a scalar) nor of the length of the
<a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup" title="MDAnalysis.core.AtomGroup.AtomGroup"><code class="xref py py-class docutils literal"><span class="pre">AtomGroup</span></code></a> then a <code class="xref py py-exc docutils literal"><span class="pre">ValueError</span></code> is raised.</p>
<div class="versionadded">
<p><span class="versionmodified">New in version 0.7.4.</span></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 0.7.5: </span>Also changes the residues.</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 0.8: </span>Can set atoms and residues to distinct values by providing a sequence or iterable.</p>
</div>
</dd></dl>

<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup.set_resnum">
<code class="descname">set_resnum</code><span class="sig-paren">(</span><em>resnum</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#AtomGroup.set_resnum"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup.set_resnum" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the resnum to <em>resnum</em> for <strong>all atoms</strong> in the <a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup" title="MDAnalysis.core.AtomGroup.AtomGroup"><code class="xref py py-class docutils literal"><span class="pre">AtomGroup</span></code></a>.</p>
<p>If <em>resnum</em> is a sequence of the same length as the <a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup" title="MDAnalysis.core.AtomGroup.AtomGroup"><code class="xref py py-class docutils literal"><span class="pre">AtomGroup</span></code></a>
then each <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Atom.resnum" title="MDAnalysis.core.AtomGroup.Atom.resnum"><code class="xref py py-attr docutils literal"><span class="pre">Atom.resnum</span></code></a> is set to the corresponding value together
with the <code class="xref py py-attr docutils literal"><span class="pre">Residue.resnum</span></code> of the residue the atom belongs to. If
<em>value</em> is neither of length 1 (or a scalar) nor of the length of the
<a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup" title="MDAnalysis.core.AtomGroup.AtomGroup"><code class="xref py py-class docutils literal"><span class="pre">AtomGroup</span></code></a> then a <code class="xref py py-exc docutils literal"><span class="pre">ValueError</span></code> is raised.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Changing <em>resnum</em> will not affect the topology: you can have
multiple residues with the same <em>resnum</em>.</p>
</div>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup.set_resid" title="MDAnalysis.core.AtomGroup.AtomGroup.set_resid"><code class="xref py py-meth docutils literal"><span class="pre">set_resid()</span></code></a></p>
</div>
<div class="versionadded">
<p><span class="versionmodified">New in version 0.7.4.</span></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 0.7.5: </span>Also changes the residues.</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 0.8: </span>Can set atoms and residues to distinct values by providing a sequence or iterable.</p>
</div>
</dd></dl>

<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup.set_segid">
<code class="descname">set_segid</code><span class="sig-paren">(</span><em>segid</em>, <em>buildsegments=True</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#AtomGroup.set_segid"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup.set_segid" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the segid to <em>segid</em> for all atoms in the <a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup" title="MDAnalysis.core.AtomGroup.AtomGroup"><code class="xref py py-class docutils literal"><span class="pre">AtomGroup</span></code></a>.</p>
<p>If <em>segid</em> is a sequence of the same length as the <a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup" title="MDAnalysis.core.AtomGroup.AtomGroup"><code class="xref py py-class docutils literal"><span class="pre">AtomGroup</span></code></a>
then each <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Atom.segid" title="MDAnalysis.core.AtomGroup.Atom.segid"><code class="xref py py-attr docutils literal"><span class="pre">Atom.segid</span></code></a> is set to the corresponding value together
with the <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Segment.id" title="MDAnalysis.core.AtomGroup.Segment.id"><code class="xref py py-attr docutils literal"><span class="pre">Segment.id</span></code></a> of the residue the atom belongs to. If
<em>value</em> is neither of length 1 (or a scalar) nor of the length of the
<a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup" title="MDAnalysis.core.AtomGroup.AtomGroup"><code class="xref py py-class docutils literal"><span class="pre">AtomGroup</span></code></a> then a <code class="xref py py-exc docutils literal"><span class="pre">ValueError</span></code> is raised.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p><a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup.set_segid" title="MDAnalysis.core.AtomGroup.AtomGroup.set_segid"><code class="xref py py-meth docutils literal"><span class="pre">set_segid()</span></code></a> can change the topology.</p>
<p class="last">With the default <em>buildsegments</em> = <code class="docutils literal"><span class="pre">True</span></code> it can be used to join
segments or to break groups into multiple disjoint segments. Note
that each <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Atom" title="MDAnalysis.core.AtomGroup.Atom"><code class="xref py py-class docutils literal"><span class="pre">Atom</span></code></a> can only belong to a single
<a class="reference internal" href="#MDAnalysis.core.AtomGroup.Segment" title="MDAnalysis.core.AtomGroup.Segment"><code class="xref py py-class docutils literal"><span class="pre">Segment</span></code></a>.</p>
</div>
<p>For performance reasons, <em>buildsegments</em> can be set to <code class="docutils literal"><span class="pre">False</span></code>. Then
one needs to run <code class="xref py py-meth docutils literal"><span class="pre">Universe._build_segments()</span></code> manually later in
order to update the list of <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Segment" title="MDAnalysis.core.AtomGroup.Segment"><code class="xref py py-class docutils literal"><span class="pre">Segment</span></code></a> instances and regenerate
the segid instant selectors.</p>
<div class="versionadded">
<p><span class="versionmodified">New in version 0.7.4.</span></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 0.8: </span>Can set atoms and residues to distinct values by providing a sequence or iterable.</p>
</div>
</dd></dl>

<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup.set_type">
<code class="descname">set_type</code><span class="sig-paren">(</span><em>atype</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#AtomGroup.set_type"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup.set_type" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the atom type to <em>atype</em> for <strong>all atoms</strong> in the AtomGroup.</p>
<div class="versionadded">
<p><span class="versionmodified">New in version 0.7.4.</span></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 0.8: </span>Can set atoms and residues to distinct values by providing a sequence or iterable.</p>
</div>
</dd></dl>

<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup.set_velocities">
<code class="descname">set_velocities</code><span class="sig-paren">(</span><em>v</em>, <em>ts=None</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#AtomGroup.set_velocities"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup.set_velocities" title="Permalink to this definition">¶</a></dt>
<dd><p>Assign the velocities <em>v</em> to the timestep.</p>
<p>Raises a <a class="reference internal" href="#MDAnalysis.core.AtomGroup.NoDataError" title="MDAnalysis.core.AtomGroup.NoDataError"><code class="xref py py-exc docutils literal"><span class="pre">NoDataError</span></code></a> if the underlying
<a class="reference internal" href="../coordinates/base.html#MDAnalysis.coordinates.base.Timestep" title="MDAnalysis.coordinates.base.Timestep"><code class="xref py py-class docutils literal"><span class="pre">Timestep</span></code></a> does not contain
<a class="reference internal" href="../coordinates/base.html#MDAnalysis.coordinates.base.Timestep._velocities" title="MDAnalysis.coordinates.base.Timestep._velocities"><code class="xref py py-attr docutils literal"><span class="pre">_velocities</span></code></a>.</p>
<p>See also <a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup.get_velocities" title="MDAnalysis.core.AtomGroup.AtomGroup.get_velocities"><code class="xref py py-meth docutils literal"><span class="pre">AtomGroup.get_velocities()</span></code></a> and <a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup.velocities" title="MDAnalysis.core.AtomGroup.AtomGroup.velocities"><code class="xref py py-attr docutils literal"><span class="pre">AtomGroup.velocities</span></code></a> for
attribute access.</p>
<div class="versionadded">
<p><span class="versionmodified">New in version 0.7.6.</span></p>
</div>
</dd></dl>

<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup.shapeParameter">
<code class="descname">shapeParameter</code><span class="sig-paren">(</span><em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#AtomGroup.shapeParameter"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup.shapeParameter" title="Permalink to this definition">¶</a></dt>
<dd><p>Shape parameter.</p>
<p>See <a class="reference internal" href="#dima2004" id="id2">[Dima2004]</a> for background information.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Keywords:</th><td class="field-body"><dl class="first last docutils">
<dt><em>pbc</em></dt>
<dd><p class="first last"><code class="docutils literal"><span class="pre">True</span></code>: Move all atoms within the primary unit cell before calculation [<code class="docutils literal"><span class="pre">False</span></code>]</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">The <code class="xref py py-class docutils literal"><span class="pre">MDAnalysis.core.flags</span></code> flag <em>use_pbc</em> when set to <code class="docutils literal"><span class="pre">True</span></code> allows the <em>pbc</em>
flag to be used by default.</p>
</div>
<div class="versionadded">
<p><span class="versionmodified">New in version 0.7.7.</span></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 0.8: </span>Added <em>pbc</em> keyword</p>
</div>
</dd></dl>

<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup.split">
<code class="descname">split</code><span class="sig-paren">(</span><em>level</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#AtomGroup.split"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup.split" title="Permalink to this definition">¶</a></dt>
<dd><p>Split atomgroup into a list of atomgroups by <em>level</em>.</p>
<p><em>level</em> can be &#8220;atom&#8221;, &#8220;residue&#8221;, &#8220;segment&#8221;.</p>
<div class="versionadded">
<p><span class="versionmodified">New in version 0.9.0.</span></p>
</div>
</dd></dl>

<dl class="attribute">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup.torsions">
<code class="descname">torsions</code><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup.torsions" title="Permalink to this definition">¶</a></dt>
<dd><p>All the torsions in this AtomGroup</p>
<p>Note that these torsions might extend out of the AtomGroup, to select
only torsions which are entirely contained by the AtomGroup use
u.torsions.atomgroup_intersection(ag, strict=True)</p>
<div class="versionadded">
<p><span class="versionmodified">New in version 0.9.0.</span></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 0.10.0: </span>Now always returns a (possibly empty) TopologyGroup</p>
</div>
</dd></dl>

<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup.totalCharge">
<code class="descname">totalCharge</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#AtomGroup.totalCharge"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup.totalCharge" title="Permalink to this definition">¶</a></dt>
<dd><p>Sum of all partial charges (must be defined in topology).</p>
</dd></dl>

<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup.totalMass">
<code class="descname">totalMass</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#AtomGroup.totalMass"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup.totalMass" title="Permalink to this definition">¶</a></dt>
<dd><p>Total mass of the selection (masses are taken from the topology or guessed).</p>
</dd></dl>

<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup.transform">
<code class="descname">transform</code><span class="sig-paren">(</span><em>M</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#AtomGroup.transform"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup.transform" title="Permalink to this definition">¶</a></dt>
<dd><p>Apply homogenous transformation matrix <em>M</em> to the coordinates.</p>
<p>The matrix <em>M</em> must be a 4x4 matrix, with the rotation in
<cite>R = `M[:3,:3]`</cite> and the translation in <code class="docutils literal"><span class="pre">t</span> <span class="pre">=</span> <span class="pre">M[:3,3]</span></code>.</p>
<p>The rotation <span class="math">\(\mathsf{R}\)</span> is applied before the
translation <span class="math">\(\mathbf{t}\)</span>:</p>
<div class="math">
\[\mathbf{x}' = \mathsf{R}\mathbf{x} + \mathbf{t}\]</div>
</dd></dl>

<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup.translate">
<code class="descname">translate</code><span class="sig-paren">(</span><em>t</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#AtomGroup.translate"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup.translate" title="Permalink to this definition">¶</a></dt>
<dd><p>Apply translation vector <em>t</em> to the selection&#8217;s coordinates.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">AtomGroup</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">t</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">AtomGroup</span><span class="o">.</span><span class="n">translate</span><span class="p">((</span><span class="n">A</span><span class="p">,</span> <span class="n">B</span><span class="p">))</span>
</pre></div>
</div>
<p>The method applies a translation to the AtomGroup from current
coordinates <span class="math">\(\mathbf{x}\)</span> to new coordinates <span class="math">\(\mathbf{x}'\)</span>:</p>
<div class="math">
\[\mathbf{x}' = \mathbf{x} + \mathbf{t}\]</div>
<p>The translation can also be given as a tuple of two MDAnalysis objects
such as two selections <cite>(selA, selB)</cite>, i.e. two <a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup" title="MDAnalysis.core.AtomGroup.AtomGroup"><code class="xref py py-class docutils literal"><span class="pre">AtomGroup</span></code></a>, or
two <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Atom" title="MDAnalysis.core.AtomGroup.Atom"><code class="xref py py-class docutils literal"><span class="pre">Atom</span></code></a>. The translation vector is computed as the
difference between the centers of geometry (centroid) of B and A:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">t</span> <span class="o">=</span> <span class="n">B</span><span class="o">.</span><span class="n">centroid</span><span class="p">()</span> <span class="o">-</span> <span class="n">A</span><span class="o">.</span><span class="n">centroid</span><span class="p">()</span>
</pre></div>
</div>
</dd></dl>

<dl class="attribute">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup.ts">
<code class="descname">ts</code><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup.ts" title="Permalink to this definition">¶</a></dt>
<dd><p>Temporary Timestep that contains the selection coordinates.</p>
<p>A <a class="reference internal" href="../coordinates/base.html#MDAnalysis.coordinates.base.Timestep" title="MDAnalysis.coordinates.base.Timestep"><code class="xref py py-class docutils literal"><span class="pre">Timestep</span></code></a> instance,
which can be passed to a trajectory writer.</p>
<p>If <a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup.ts" title="MDAnalysis.core.AtomGroup.AtomGroup.ts"><code class="xref py py-attr docutils literal"><span class="pre">ts</span></code></a> is modified then these modifications
will be present until the frame number changes (which
typically happens when the underlying trajectory frame
changes).</p>
<p>It is not possible to assign a new
<a class="reference internal" href="../coordinates/base.html#MDAnalysis.coordinates.base.Timestep" title="MDAnalysis.coordinates.base.Timestep"><code class="xref py py-class docutils literal"><span class="pre">Timestep</span></code></a> to the
<a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup.ts" title="MDAnalysis.core.AtomGroup.AtomGroup.ts"><code class="xref py py-attr docutils literal"><span class="pre">AtomGroup.ts</span></code></a> attribute; change attributes of the object.</p>
</dd></dl>

<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup.types">
<code class="descname">types</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#AtomGroup.types"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup.types" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns an array of atom types.</p>
</dd></dl>

<dl class="attribute">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup.universe">
<code class="descname">universe</code><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup.universe" title="Permalink to this definition">¶</a></dt>
<dd><p>The universe to which the atoms belong (read-only).</p>
</dd></dl>

<dl class="attribute">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup.velocities">
<code class="descname">velocities</code><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup.velocities" title="Permalink to this definition">¶</a></dt>
<dd><p>NumPy array of the velocities of the atoms in the group.</p>
<p>If the trajectory does not contain velocity information then a
<code class="xref py py-exc docutils literal"><span class="pre">NoDataError</span></code> is raised.</p>
<div class="versionadded">
<p><span class="versionmodified">New in version 0.7.5.</span></p>
</div>
<div class="deprecated">
<p><span class="versionmodified">Deprecated since version 0.7.6: </span>In 0.8 this will become an attribute! You can already use <a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup.get_velocities" title="MDAnalysis.core.AtomGroup.AtomGroup.get_velocities"><code class="xref py py-meth docutils literal"><span class="pre">get_velocities()</span></code></a>
and <a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup.set_velocities" title="MDAnalysis.core.AtomGroup.AtomGroup.set_velocities"><code class="xref py py-meth docutils literal"><span class="pre">set_velocities()</span></code></a>.</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 0.8: </span>Became an attribute.</p>
</div>
</dd></dl>

<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup.wrap">
<code class="descname">wrap</code><span class="sig-paren">(</span><em>compound='atoms'</em>, <em>center='com'</em>, <em>box=None</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#AtomGroup.wrap"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup.wrap" title="Permalink to this definition">¶</a></dt>
<dd><p>Shift the contents of this AtomGroup back into the unit cell.</p>
<p>This is a more powerful version of <a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup.packIntoBox" title="MDAnalysis.core.AtomGroup.AtomGroup.packIntoBox"><code class="xref py py-meth docutils literal"><span class="pre">packIntoBox()</span></code></a>, allowing
groups of atoms to be kept together through the process.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Keywords:</th><td class="field-body"><dl class="first last docutils">
<dt><em>compound</em></dt>
<dd><p class="first">The group which will be kept together through the shifting
process. [<code class="docutils literal"><span class="pre">atoms</span></code>]
Possible options:</p>
<blockquote class="last">
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">atoms</span></code></li>
<li><code class="docutils literal"><span class="pre">group</span></code> - This AtomGroup</li>
<li><code class="docutils literal"><span class="pre">residues</span></code></li>
<li><code class="docutils literal"><span class="pre">segments</span></code></li>
<li><code class="docutils literal"><span class="pre">fragments</span></code></li>
</ul>
</div></blockquote>
</dd>
<dt><em>center</em></dt>
<dd><p class="first last">How to define the center of a given group of atoms [<code class="docutils literal"><span class="pre">com</span></code>]</p>
</dd>
<dt><em>box</em></dt>
<dd><p class="first last">Unit cell information.  If not provided, the values from
Timestep will be used.</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
<p>When specifying a <em>compound</em>, the translation is calculated based on
each compound. The same translation is applied to all atoms
within this compound, meaning it will not be broken by the shift.
This might however mean that all atoms from the compound are not
inside the unit cell, but rather the center of the compound is.
Compounds available for use are <em>atoms</em>, <em>residues</em>,
<em>segments</em> and <em>fragments</em></p>
<p><em>center</em> allows the definition of the center of each group to be
specified.  This can be either &#8216;com&#8217; for center of mass, or &#8216;cog&#8217;
for center of geometry.</p>
<p><em>box</em> allows a unit cell to be given for the transformation.  If not
specified, an the dimensions information from the current Timestep
will be used.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">wrap with all default keywords is identical to <a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup.packIntoBox" title="MDAnalysis.core.AtomGroup.AtomGroup.packIntoBox"><code class="xref py py-meth docutils literal"><span class="pre">packIntoBox()</span></code></a></p>
</div>
<div class="versionadded">
<p><span class="versionmodified">New in version 0.9.2.</span></p>
</div>
</dd></dl>

<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup.write">
<code class="descname">write</code><span class="sig-paren">(</span><em>filename=None</em>, <em>format='PDB'</em>, <em>filenamefmt='%(trjname)s_%(frame)d'</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#AtomGroup.write"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup.write" title="Permalink to this definition">¶</a></dt>
<dd><p>Write AtomGroup to a file.</p>
<p>AtomGroup.write(filename[,format])</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Keywords:</th><td class="field-body"><dl class="first last docutils">
<dt><em>filename</em></dt>
<dd><p class="first last"><code class="docutils literal"><span class="pre">None</span></code>: create TRJNAME_FRAME.FORMAT from filenamefmt [<code class="docutils literal"><span class="pre">None</span></code>]</p>
</dd>
<dt><em>format</em></dt>
<dd><p class="first last">PDB, CRD, GRO, VMD (tcl), PyMol (pml), Gromacs (ndx) CHARMM (str);
case-insensitive and can also be supplied as the filename
extension [PDB]</p>
</dd>
<dt><em>filenamefmt</em></dt>
<dd><p class="first last">format string for default filename; use substitution tokens
&#8216;trjname&#8217; and &#8216;frame&#8217; [&#8220;%(trjname)s_%(frame)d&#8221;]</p>
</dd>
<dt><em>bonds</em></dt>
<dd><p class="first">how to handle bond information, especially relevant for PDBs;
default is <code class="docutils literal"><span class="pre">&quot;conect&quot;</span></code>.</p>
<ul class="last simple">
<li><code class="docutils literal"><span class="pre">&quot;conect&quot;</span></code>: write only the CONECT records defined in the original
file</li>
<li><code class="docutils literal"><span class="pre">&quot;all&quot;</span></code>: write out all bonds, both the original defined and those
guessed by MDAnalysis</li>
<li><code class="docutils literal"><span class="pre">None</span></code>: do not write out bonds</li>
</ul>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 0.9.0: </span>Merged with write_selection.  This method can now write both
selections out.</p>
</div>
</dd></dl>

<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.AtomGroup.write_selection">
<code class="descname">write_selection</code><span class="sig-paren">(</span><em>filename=None</em>, <em>format='vmd'</em>, <em>filenamefmt='%(trjname)s_%(frame)d'</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#AtomGroup.write_selection"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.AtomGroup.write_selection" title="Permalink to this definition">¶</a></dt>
<dd><p>Write AtomGroup selection to a file to be used in another programme.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Keywords:</th><td class="field-body"><dl class="first last docutils">
<dt><em>filename</em></dt>
<dd><p class="first last"><code class="docutils literal"><span class="pre">None</span></code>: create TRJNAME_FRAME.FORMAT from <em>filenamefmt</em></p>
</dd>
<dt><em>format</em></dt>
<dd><p class="first last">output file format: VMD (tcl), PyMol (pml), Gromacs (ndx), CHARMM (str);
can also be supplied as the filename extension. Case insensitive. [vmd]</p>
</dd>
<dt><em>filenamefmt</em></dt>
<dd><p class="first last">format string for default filename; use &#8216;%(trjname)s&#8217; and &#8216;%(frame)s&#8217;
placeholders; the extension is set according to the <em>format</em>
[&#8220;%(trjname)s_%(frame)d&#8221;]</p>
</dd>
<dt><em>kwargs</em></dt>
<dd><p class="first last">additional keywords are passed on to the appropriate
<a class="reference internal" href="../selections/base.html#MDAnalysis.selections.base.SelectionWriter" title="MDAnalysis.selections.base.SelectionWriter"><code class="xref py py-class docutils literal"><span class="pre">SelectionWriter</span></code></a></p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
<div class="deprecated">
<p><span class="versionmodified">Deprecated since version 0.9.0: </span>Use <a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup.write" title="MDAnalysis.core.AtomGroup.AtomGroup.write"><code class="xref py py-meth docutils literal"><span class="pre">write()</span></code></a></p>
</div>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="MDAnalysis.core.AtomGroup.Atom">
<em class="property">class </em><code class="descclassname">MDAnalysis.core.AtomGroup.</code><code class="descname">Atom</code><span class="sig-paren">(</span><em>number</em>, <em>name</em>, <em>type</em>, <em>resname</em>, <em>resid</em>, <em>segid</em>, <em>mass</em>, <em>charge</em>, <em>residue=None</em>, <em>segment=None</em>, <em>radius=None</em>, <em>bfactor=None</em>, <em>resnum=None</em>, <em>serial=None</em>, <em>altLoc=None</em>, <em>universe=None</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#Atom"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.Atom" title="Permalink to this definition">¶</a></dt>
<dd><p>A class representing a single atom.</p>
<p><a class="reference internal" href="#MDAnalysis.core.AtomGroup.Atom" title="MDAnalysis.core.AtomGroup.Atom"><code class="xref py py-class docutils literal"><span class="pre">Atom</span></code></a> is the basic building block of all larger data
structures in MDAnalysis, in particular of the
<a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup" title="MDAnalysis.core.AtomGroup.AtomGroup"><code class="xref py py-class docutils literal"><span class="pre">AtomGroup</span></code></a>.</p>
<p>An <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Atom" title="MDAnalysis.core.AtomGroup.Atom"><code class="xref py py-class docutils literal"><span class="pre">Atom</span></code></a> is typically generated by a <a class="reference internal" href="../topology/init.html#supported-topology-formats"><span>topology reader</span></a> from <code class="xref py py-mod docutils literal"><span class="pre">MDAnalysis.topology</span></code>.</p>
<p>For performance reasons, only a predefined number of attributes
are included (and thus it is not possible to add attributes &#8220;on
the fly&#8221;; they have to be included in the class definition).</p>
<dl class="attribute">
<dt id="MDAnalysis.core.AtomGroup.Atom.number">
<code class="descname">number</code><a class="headerlink" href="#MDAnalysis.core.AtomGroup.Atom.number" title="Permalink to this definition">¶</a></dt>
<dd><p>atom number</p>
</dd></dl>

<dl class="attribute">
<dt id="MDAnalysis.core.AtomGroup.Atom.segid">
<code class="descname">segid</code><a class="headerlink" href="#MDAnalysis.core.AtomGroup.Atom.segid" title="Permalink to this definition">¶</a></dt>
<dd><p>name of the segment</p>
</dd></dl>

<dl class="attribute">
<dt id="MDAnalysis.core.AtomGroup.Atom.resid">
<code class="descname">resid</code><a class="headerlink" href="#MDAnalysis.core.AtomGroup.Atom.resid" title="Permalink to this definition">¶</a></dt>
<dd><p>residue number</p>
</dd></dl>

<dl class="attribute">
<dt id="MDAnalysis.core.AtomGroup.Atom.resnum">
<code class="descname">resnum</code><a class="headerlink" href="#MDAnalysis.core.AtomGroup.Atom.resnum" title="Permalink to this definition">¶</a></dt>
<dd><p>canonical residue number as, for instance, used in the original
PDB file</p>
<div class="versionadded">
<p><span class="versionmodified">New in version 0.7.4.</span></p>
</div>
</dd></dl>

<dl class="attribute">
<dt id="MDAnalysis.core.AtomGroup.Atom.resname">
<code class="descname">resname</code><a class="headerlink" href="#MDAnalysis.core.AtomGroup.Atom.resname" title="Permalink to this definition">¶</a></dt>
<dd><p>residue name</p>
</dd></dl>

<dl class="attribute">
<dt id="MDAnalysis.core.AtomGroup.Atom.residue">
<code class="descname">residue</code><a class="headerlink" href="#MDAnalysis.core.AtomGroup.Atom.residue" title="Permalink to this definition">¶</a></dt>
<dd><p><a class="reference internal" href="#MDAnalysis.core.AtomGroup.Residue" title="MDAnalysis.core.AtomGroup.Residue"><code class="xref py py-class docutils literal"><span class="pre">Residue</span></code></a> object containing the atoms</p>
</dd></dl>

<dl class="attribute">
<dt id="MDAnalysis.core.AtomGroup.Atom.id">
<code class="descname">id</code><a class="headerlink" href="#MDAnalysis.core.AtomGroup.Atom.id" title="Permalink to this definition">¶</a></dt>
<dd><p>atom number inside the residue</p>
</dd></dl>

<dl class="attribute">
<dt id="MDAnalysis.core.AtomGroup.Atom.name">
<code class="descname">name</code><a class="headerlink" href="#MDAnalysis.core.AtomGroup.Atom.name" title="Permalink to this definition">¶</a></dt>
<dd><p>string, short name</p>
</dd></dl>

<dl class="attribute">
<dt id="MDAnalysis.core.AtomGroup.Atom.type">
<code class="descname">type</code><a class="headerlink" href="#MDAnalysis.core.AtomGroup.Atom.type" title="Permalink to this definition">¶</a></dt>
<dd><p>string or number (from force field), describing the atom type;
stored as a string.</p>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 0.7.6: </span>The <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Atom.type" title="MDAnalysis.core.AtomGroup.Atom.type"><code class="xref py py-attr docutils literal"><span class="pre">Atom.type</span></code></a> attribute is always stored as a string.</p>
</div>
</dd></dl>

<dl class="attribute">
<dt id="MDAnalysis.core.AtomGroup.Atom.mass">
<code class="descname">mass</code><a class="headerlink" href="#MDAnalysis.core.AtomGroup.Atom.mass" title="Permalink to this definition">¶</a></dt>
<dd><p>float, in <a class="reference external" href="http://physics.nist.gov/cgi-bin/cuu/Value?u">atomic mass units</a> (u)</p>
</dd></dl>

<dl class="attribute">
<dt id="MDAnalysis.core.AtomGroup.Atom.charge">
<code class="descname">charge</code><a class="headerlink" href="#MDAnalysis.core.AtomGroup.Atom.charge" title="Permalink to this definition">¶</a></dt>
<dd><p>float, in <a class="reference external" href="http://physics.nist.gov/cgi-bin/cuu/Value?e">electron charges</a> (<em>e</em>)</p>
</dd></dl>

<dl class="attribute">
<dt id="MDAnalysis.core.AtomGroup.Atom.radius">
<code class="descname">radius</code><a class="headerlink" href="#MDAnalysis.core.AtomGroup.Atom.radius" title="Permalink to this definition">¶</a></dt>
<dd><p>Born-radius for electrostatic calculations. (Only if read from a
PQR file with the
<a class="reference internal" href="../coordinates/PQR.html#MDAnalysis.coordinates.PQR.PQRReader" title="MDAnalysis.coordinates.PQR.PQRReader"><code class="xref py py-class docutils literal"><span class="pre">PQRReader</span></code></a>.)</p>
</dd></dl>

<dl class="attribute">
<dt id="MDAnalysis.core.AtomGroup.Atom.altLoc">
<code class="descname">altLoc</code><a class="headerlink" href="#MDAnalysis.core.AtomGroup.Atom.altLoc" title="Permalink to this definition">¶</a></dt>
<dd><p>Alternate location indicator (as used in <cite>ATOM_</cite> records in PDB
files)</p>
<div class="versionadded">
<p><span class="versionmodified">New in version 0.9.0.</span></p>
</div>
</dd></dl>

<dl class="attribute">
<dt id="MDAnalysis.core.AtomGroup.Atom.bonds">
<code class="descname">bonds</code><a class="headerlink" href="#MDAnalysis.core.AtomGroup.Atom.bonds" title="Permalink to this definition">¶</a></dt>
<dd><p>List of <code class="xref py py-class docutils literal"><span class="pre">Bond</span></code> instances that
contains all the bonds that this atom participates in.</p>
<div class="versionadded">
<p><span class="versionmodified">New in version 0.8.1.</span></p>
</div>
</dd></dl>

<dl class="attribute">
<dt id="MDAnalysis.core.AtomGroup.Atom.angles">
<code class="descname">angles</code><a class="headerlink" href="#MDAnalysis.core.AtomGroup.Atom.angles" title="Permalink to this definition">¶</a></dt>
<dd><p>List of <code class="xref py py-class docutils literal"><span class="pre">Angle</span></code> instances
that contains all the angles that this atom participates in.</p>
<div class="versionadded">
<p><span class="versionmodified">New in version 0.8.1.</span></p>
</div>
</dd></dl>

<dl class="attribute">
<dt id="MDAnalysis.core.AtomGroup.Atom.torsions">
<code class="descname">torsions</code><a class="headerlink" href="#MDAnalysis.core.AtomGroup.Atom.torsions" title="Permalink to this definition">¶</a></dt>
<dd><p>List of <code class="xref py py-class docutils literal"><span class="pre">Torsion</span></code> instances
that contains all the dihedral torsions that this atom
participates in.</p>
<div class="versionadded">
<p><span class="versionmodified">New in version 0.8.1.</span></p>
</div>
</dd></dl>

<dl class="attribute">
<dt>
<code class="descname">angles</code></dt>
<dd><p>A list of the angles that this Atom is in</p>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 0.9.0: </span>Changed to managed property</p>
</div>
</dd></dl>

<dl class="attribute">
<dt id="MDAnalysis.core.AtomGroup.Atom.bonded_atoms">
<code class="descname">bonded_atoms</code><a class="headerlink" href="#MDAnalysis.core.AtomGroup.Atom.bonded_atoms" title="Permalink to this definition">¶</a></dt>
<dd><p>An AtomGroup of the Atoms that this Atom is bonded to.</p>
<div class="versionadded">
<p><span class="versionmodified">New in version 0.9.</span></p>
</div>
</dd></dl>

<dl class="attribute">
<dt>
<code class="descname">bonds</code></dt>
<dd><p>A list of the bonds that this Atom is in</p>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 0.9.0: </span>Changed to managed property</p>
</div>
</dd></dl>

<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.Atom.centroid">
<code class="descname">centroid</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#Atom.centroid"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.Atom.centroid" title="Permalink to this definition">¶</a></dt>
<dd><p>The centroid of an atom is its position, <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Atom.pos" title="MDAnalysis.core.AtomGroup.Atom.pos"><code class="xref py py-attr docutils literal"><span class="pre">Atom.pos</span></code></a>.</p>
</dd></dl>

<dl class="attribute">
<dt id="MDAnalysis.core.AtomGroup.Atom.dihedrals">
<code class="descname">dihedrals</code><a class="headerlink" href="#MDAnalysis.core.AtomGroup.Atom.dihedrals" title="Permalink to this definition">¶</a></dt>
<dd><p>A list of the torsions/dihedrals that this Atom is in</p>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 0.9.0: </span>Changed to managed property</p>
</div>
</dd></dl>

<dl class="attribute">
<dt id="MDAnalysis.core.AtomGroup.Atom.force">
<code class="descname">force</code><a class="headerlink" href="#MDAnalysis.core.AtomGroup.Atom.force" title="Permalink to this definition">¶</a></dt>
<dd><p>Current force of the atom.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">a (3,) shape numpy array</td>
</tr>
</tbody>
</table>
<p>A <code class="xref py py-exc docutils literal"><span class="pre">NoDataError</span></code> is raised if the trajectory
does not contain velocities.</p>
<div class="versionadded">
<p><span class="versionmodified">New in version 0.9.2.</span></p>
</div>
</dd></dl>

<dl class="attribute">
<dt id="MDAnalysis.core.AtomGroup.Atom.fragment">
<code class="descname">fragment</code><a class="headerlink" href="#MDAnalysis.core.AtomGroup.Atom.fragment" title="Permalink to this definition">¶</a></dt>
<dd><p>The fragment that this Atom is part of</p>
<div class="versionadded">
<p><span class="versionmodified">New in version 0.9.0.</span></p>
</div>
</dd></dl>

<dl class="attribute">
<dt id="MDAnalysis.core.AtomGroup.Atom.impropers">
<code class="descname">impropers</code><a class="headerlink" href="#MDAnalysis.core.AtomGroup.Atom.impropers" title="Permalink to this definition">¶</a></dt>
<dd><p>A list of the improper torsions that this Atom is in</p>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 0.9.0: </span>Changed to managed property</p>
</div>
</dd></dl>

<dl class="attribute">
<dt id="MDAnalysis.core.AtomGroup.Atom.pos">
<code class="descname">pos</code><a class="headerlink" href="#MDAnalysis.core.AtomGroup.Atom.pos" title="Permalink to this definition">¶</a></dt>
<dd><p>coordinates of the atom</p>
<p>Get the current cartesian coordinates of the atom (read-only).</p>
<div class="deprecated">
<p><span class="versionmodified">Deprecated since version 0.8: </span>use <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Atom.position" title="MDAnalysis.core.AtomGroup.Atom.position"><code class="xref py py-attr docutils literal"><span class="pre">position</span></code></a></p>
</div>
</dd></dl>

<dl class="attribute">
<dt id="MDAnalysis.core.AtomGroup.Atom.position">
<code class="descname">position</code><a class="headerlink" href="#MDAnalysis.core.AtomGroup.Atom.position" title="Permalink to this definition">¶</a></dt>
<dd><p>coordinates of the atom</p>
<p>Get the current cartesian coordinates of the atom.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">a (3,) shape numpy array</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt>
<code class="descname">torsions</code></dt>
<dd><p>A list of the torsions/dihedrals that this Atom is in</p>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 0.9.0: </span>Changed to managed property</p>
</div>
</dd></dl>

<dl class="attribute">
<dt id="MDAnalysis.core.AtomGroup.Atom.universe">
<code class="descname">universe</code><a class="headerlink" href="#MDAnalysis.core.AtomGroup.Atom.universe" title="Permalink to this definition">¶</a></dt>
<dd><p>a pointer back to the Universe</p>
</dd></dl>

<dl class="attribute">
<dt id="MDAnalysis.core.AtomGroup.Atom.velocity">
<code class="descname">velocity</code><a class="headerlink" href="#MDAnalysis.core.AtomGroup.Atom.velocity" title="Permalink to this definition">¶</a></dt>
<dd><p>Current velocity of the atom.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">a (3,) shape numpy array</td>
</tr>
</tbody>
</table>
<p>A <code class="xref py py-exc docutils literal"><span class="pre">NoDataError</span></code> is raised if the trajectory
does not contain velocities.</p>
<div class="versionadded">
<p><span class="versionmodified">New in version 0.7.5.</span></p>
</div>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="MDAnalysis.core.AtomGroup.Residue">
<em class="property">class </em><code class="descclassname">MDAnalysis.core.AtomGroup.</code><code class="descname">Residue</code><span class="sig-paren">(</span><em>name</em>, <em>id</em>, <em>atoms</em>, <em>resnum=None</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#Residue"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.Residue" title="Permalink to this definition">¶</a></dt>
<dd><p>A group of atoms corresponding to a residue.</p>
<dl class="docutils">
<dt>Pythonic access to atoms:</dt>
<dd><ul class="first last">
<li><p class="first">Using a atom name as attribute returns the matching atom (a
<a class="reference internal" href="#MDAnalysis.core.AtomGroup.Atom" title="MDAnalysis.core.AtomGroup.Atom"><code class="xref py py-class docutils literal"><span class="pre">Atom</span></code></a> instance), i.e. <code class="docutils literal"><span class="pre">r.name</span></code>. Example:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">MDAnalysis.tests.datafiles</span> <span class="kn">import</span> <span class="n">PSF</span><span class="p">,</span><span class="n">DCD</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">u</span> <span class="o">=</span> <span class="n">Universe</span><span class="p">(</span><span class="n">PSF</span><span class="p">,</span><span class="n">DCD</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span><span class="p">(</span><span class="n">u</span><span class="o">.</span><span class="n">s4AKE</span><span class="o">.</span><span class="n">r1</span><span class="o">.</span><span class="n">CA</span><span class="p">)</span>  <span class="c"># C-alpha of M1</span>
<span class="go">&lt; Atom 5: name &#39;CA&#39; of type &#39;22&#39; of resname &#39;MET&#39;, resid 1 and segid &#39;4AKE&#39;&gt;</span>
</pre></div>
</div>
</li>
<li><p class="first"><code class="docutils literal"><span class="pre">r['name']</span></code> or <code class="docutils literal"><span class="pre">r[id]</span></code> - returns the atom corresponding to that name</p>
</li>
</ul>
</dd>
</dl>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Data:</th><td class="field-body"><dl class="first last docutils">
<dt><code class="xref py py-attr docutils literal"><span class="pre">Residue.name</span></code></dt>
<dd><p class="first last">Three letter residue name.</p>
</dd>
<dt><code class="xref py py-attr docutils literal"><span class="pre">Residue.id</span></code></dt>
<dd><p class="first last">Numeric (integer) resid, taken from the topology.</p>
</dd>
<dt><code class="xref py py-attr docutils literal"><span class="pre">Residue.resnum</span></code></dt>
<dd><p class="first last">Numeric canonical residue id (e.g. as used in the PDB structure).</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Creating a <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Residue" title="MDAnalysis.core.AtomGroup.Residue"><code class="xref py py-class docutils literal"><span class="pre">Residue</span></code></a> modifies the underlying <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Atom" title="MDAnalysis.core.AtomGroup.Atom"><code class="xref py py-class docutils literal"><span class="pre">Atom</span></code></a>
instances. Each <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Atom" title="MDAnalysis.core.AtomGroup.Atom"><code class="xref py py-class docutils literal"><span class="pre">Atom</span></code></a> can only belong to a single
<a class="reference internal" href="#MDAnalysis.core.AtomGroup.Residue" title="MDAnalysis.core.AtomGroup.Residue"><code class="xref py py-class docutils literal"><span class="pre">Residue</span></code></a>.</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 0.7.4: </span>Added <code class="xref py py-attr docutils literal"><span class="pre">Residue.resnum</span></code> attribute and <em>resnum</em> keyword argument.</p>
</div>
<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.Residue.chi1_selection">
<code class="descname">chi1_selection</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#Residue.chi1_selection"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.Residue.chi1_selection" title="Permalink to this definition">¶</a></dt>
<dd><p>AtomGroup corresponding to the chi1 sidechain dihedral N-CA-CB-CG.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">4-atom selection in the correct order.  If no CB and/or CG is
found then this method returns <code class="docutils literal"><span class="pre">None</span></code>.</td>
</tr>
</tbody>
</table>
<div class="versionadded">
<p><span class="versionmodified">New in version 0.7.5.</span></p>
</div>
</dd></dl>

<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.Residue.omega_selection">
<code class="descname">omega_selection</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#Residue.omega_selection"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.Residue.omega_selection" title="Permalink to this definition">¶</a></dt>
<dd><p>AtomGroup corresponding to the omega protein backbone dihedral CA-C-N&#8217;-CA&#8217;.</p>
<p>omega describes the -C-N- peptide bond. Typically, it is trans
(180 degrees) although cis-bonds (0 degrees) are also
occasionally observed (especially near Proline).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">4-atom selection in the correct order.  If no C&#8217;
found in the previous residue (by resid) then this
method returns <code class="docutils literal"><span class="pre">None</span></code>.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.Residue.phi_selection">
<code class="descname">phi_selection</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#Residue.phi_selection"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.Residue.phi_selection" title="Permalink to this definition">¶</a></dt>
<dd><p>AtomGroup corresponding to the phi protein backbone dihedral C&#8217;-N-CA-C.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">4-atom selection in the correct order.  If no C&#8217;
found in the previous residue (by resid) then this
method returns <code class="docutils literal"><span class="pre">None</span></code>.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.Residue.psi_selection">
<code class="descname">psi_selection</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#Residue.psi_selection"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.Residue.psi_selection" title="Permalink to this definition">¶</a></dt>
<dd><p>AtomGroup corresponding to the psi protein backbone dihedral N-CA-C-N&#8217;.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">4-atom selection in the correct order.  If no N&#8217;
found in the following residue (by resid) then this
method returns <code class="docutils literal"><span class="pre">None</span></code>.</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="MDAnalysis.core.AtomGroup.ResidueGroup">
<em class="property">class </em><code class="descclassname">MDAnalysis.core.AtomGroup.</code><code class="descname">ResidueGroup</code><span class="sig-paren">(</span><em>residues</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#ResidueGroup"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.ResidueGroup" title="Permalink to this definition">¶</a></dt>
<dd><p>A group of residues.</p>
<dl class="docutils">
<dt>Pythonic access to atoms:</dt>
<dd><ul class="first last">
<li><p class="first">Using a atom name as attribute returns a list of all atoms (a
<a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup" title="MDAnalysis.core.AtomGroup.AtomGroup"><code class="xref py py-class docutils literal"><span class="pre">AtomGroup</span></code></a>) of the same name. Example:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">MDAnalysis.tests.datafiles</span> <span class="kn">import</span> <span class="n">PSF</span><span class="p">,</span><span class="n">DCD</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">u</span> <span class="o">=</span> <span class="n">Universe</span><span class="p">(</span><span class="n">PSF</span><span class="p">,</span><span class="n">DCD</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span><span class="p">(</span><span class="n">u</span><span class="o">.</span><span class="n">s4AKE</span><span class="o">.</span><span class="n">MET</span><span class="o">.</span><span class="n">CA</span><span class="p">)</span>  <span class="c"># C-alpha of all Met</span>
<span class="go">&lt;AtomGroup with 6 atoms&gt;</span>
</pre></div>
</div>
</li>
</ul>
</dd>
</dl>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Data:</th><td class="field-body"><code class="xref py py-attr docutils literal"><span class="pre">ResidueGroup._residues</span></code></td>
</tr>
</tbody>
</table>
<p>Initialize the ResidueGroup with a list of <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Residue" title="MDAnalysis.core.AtomGroup.Residue"><code class="xref py py-class docutils literal"><span class="pre">Residue</span></code></a> instances.</p>
<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.ResidueGroup.set">
<code class="descname">set</code><span class="sig-paren">(</span><em>name</em>, <em>value</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.AtomGroup.ResidueGroup.set" title="Permalink to this definition">¶</a></dt>
<dd><p>Set attribute <em>name</em> to <em>value</em> for all residues in the <a class="reference internal" href="#MDAnalysis.core.AtomGroup.ResidueGroup" title="MDAnalysis.core.AtomGroup.ResidueGroup"><code class="xref py py-class docutils literal"><span class="pre">ResidueGroup</span></code></a>.</p>
<p>If <em>value</em> is a sequence of the same length as the
<a class="reference internal" href="#MDAnalysis.core.AtomGroup.ResidueGroup" title="MDAnalysis.core.AtomGroup.ResidueGroup"><code class="xref py py-class docutils literal"><span class="pre">ResidueGroup</span></code></a> (<a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup.residues" title="MDAnalysis.core.AtomGroup.AtomGroup.residues"><code class="xref py py-attr docutils literal"><span class="pre">AtomGroup.residues</span></code></a>) then each
<a class="reference internal" href="#MDAnalysis.core.AtomGroup.Residue" title="MDAnalysis.core.AtomGroup.Residue"><code class="xref py py-class docutils literal"><span class="pre">Residue</span></code></a>&#8216;s property <em>name</em> is set to the corresponding
value. If <em>value</em> is neither of length 1 (or a scalar) nor of the
length of the <a class="reference internal" href="#MDAnalysis.core.AtomGroup.ResidueGroup" title="MDAnalysis.core.AtomGroup.ResidueGroup"><code class="xref py py-class docutils literal"><span class="pre">ResidueGroup</span></code></a> then a <code class="xref py py-exc docutils literal"><span class="pre">ValueError</span></code> is raised.</p>
<div class="versionadded">
<p><span class="versionmodified">New in version 0.7.5.</span></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 0.8: </span>Can set residues to distinct values by providing a sequence or iterable.</p>
</div>
</dd></dl>

<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.ResidueGroup.set_resid">
<code class="descname">set_resid</code><span class="sig-paren">(</span><em>resid</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#ResidueGroup.set_resid"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.ResidueGroup.set_resid" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the resid to integer <em>resid</em> for <strong>all residues</strong> in the <a class="reference internal" href="#MDAnalysis.core.AtomGroup.ResidueGroup" title="MDAnalysis.core.AtomGroup.ResidueGroup"><code class="xref py py-class docutils literal"><span class="pre">ResidueGroup</span></code></a>.</p>
<p>If <em>resid</em> is a sequence of the same length as the <a class="reference internal" href="#MDAnalysis.core.AtomGroup.ResidueGroup" title="MDAnalysis.core.AtomGroup.ResidueGroup"><code class="xref py py-class docutils literal"><span class="pre">ResidueGroup</span></code></a>
then each <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Atom.resid" title="MDAnalysis.core.AtomGroup.Atom.resid"><code class="xref py py-attr docutils literal"><span class="pre">Atom.resid</span></code></a> is set to the corresponding value together
with the <code class="xref py py-attr docutils literal"><span class="pre">Residue.id</span></code> of the residue the atom belongs to. If
<em>value</em> is neither of length 1 (or a scalar) nor of the length of the
<a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup" title="MDAnalysis.core.AtomGroup.AtomGroup"><code class="xref py py-class docutils literal"><span class="pre">AtomGroup</span></code></a> then a <code class="xref py py-exc docutils literal"><span class="pre">ValueError</span></code> is raised.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p>Changing resids can change the topology.</p>
<p class="last">Assigning the same <em>resid</em> to multiple residues will <strong>merge</strong> these
residues. The new residue name will be the name of the first old
residue in the merged residue.</p>
</div>
<div class="admonition warning">
<p class="first admonition-title">Warning</p>
<p class="last">The values of <em>this</em> <a class="reference internal" href="#MDAnalysis.core.AtomGroup.ResidueGroup" title="MDAnalysis.core.AtomGroup.ResidueGroup"><code class="xref py py-class docutils literal"><span class="pre">ResidueGroup</span></code></a> are not being
changed. You <strong>must create a new</strong> <a class="reference internal" href="#MDAnalysis.core.AtomGroup.ResidueGroup" title="MDAnalysis.core.AtomGroup.ResidueGroup"><code class="xref py py-class docutils literal"><span class="pre">ResidueGroup</span></code></a> <strong>from the</strong>
<a class="reference internal" href="#MDAnalysis.core.AtomGroup.Universe" title="MDAnalysis.core.AtomGroup.Universe"><code class="xref py py-class docutils literal"><span class="pre">Universe</span></code></a> &#8212; only <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Atom" title="MDAnalysis.core.AtomGroup.Atom"><code class="xref py py-class docutils literal"><span class="pre">Atom</span></code></a> instances are changed,
everything else is derived from these atoms.</p>
</div>
<div class="versionadded">
<p><span class="versionmodified">New in version 0.8.</span></p>
</div>
</dd></dl>

<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.ResidueGroup.set_resname">
<code class="descname">set_resname</code><span class="sig-paren">(</span><em>resname</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#ResidueGroup.set_resname"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.ResidueGroup.set_resname" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the resname to string <em>resname</em> for <strong>all residues</strong> in the <a class="reference internal" href="#MDAnalysis.core.AtomGroup.ResidueGroup" title="MDAnalysis.core.AtomGroup.ResidueGroup"><code class="xref py py-class docutils literal"><span class="pre">ResidueGroup</span></code></a>.</p>
<p>If <em>resname</em> is a sequence of the same length as the <a class="reference internal" href="#MDAnalysis.core.AtomGroup.ResidueGroup" title="MDAnalysis.core.AtomGroup.ResidueGroup"><code class="xref py py-class docutils literal"><span class="pre">ResidueGroup</span></code></a>
then each <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Atom.resname" title="MDAnalysis.core.AtomGroup.Atom.resname"><code class="xref py py-attr docutils literal"><span class="pre">Atom.resname</span></code></a> is set to the corresponding value together
with the <code class="xref py py-attr docutils literal"><span class="pre">Residue.name</span></code> of the residue the atom belongs to. If
<em>value</em> is neither of length 1 (or a scalar) nor of the length of the
<a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup" title="MDAnalysis.core.AtomGroup.AtomGroup"><code class="xref py py-class docutils literal"><span class="pre">AtomGroup</span></code></a> then a <code class="xref py py-exc docutils literal"><span class="pre">ValueError</span></code> is raised.</p>
<div class="versionadded">
<p><span class="versionmodified">New in version 0.7.4.</span></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 0.7.5: </span>Also changes the residues.</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 0.8: </span>Can set atoms and residues to distinct values by providing a sequence or iterable.</p>
</div>
</dd></dl>

<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.ResidueGroup.set_resnum">
<code class="descname">set_resnum</code><span class="sig-paren">(</span><em>resnum</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#ResidueGroup.set_resnum"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.ResidueGroup.set_resnum" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the resnum to <em>resnum</em> for <strong>all residues</strong> in the <a class="reference internal" href="#MDAnalysis.core.AtomGroup.ResidueGroup" title="MDAnalysis.core.AtomGroup.ResidueGroup"><code class="xref py py-class docutils literal"><span class="pre">ResidueGroup</span></code></a>.</p>
<p>If <em>resnum</em> is a sequence of the same length as the <a class="reference internal" href="#MDAnalysis.core.AtomGroup.ResidueGroup" title="MDAnalysis.core.AtomGroup.ResidueGroup"><code class="xref py py-class docutils literal"><span class="pre">ResidueGroup</span></code></a>
then each <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Atom.resnum" title="MDAnalysis.core.AtomGroup.Atom.resnum"><code class="xref py py-attr docutils literal"><span class="pre">Atom.resnum</span></code></a> is set to the corresponding value together
with the <code class="xref py py-attr docutils literal"><span class="pre">Residue.resnum</span></code> of the residue the atom belongs to. If
<em>value</em> is neither of length 1 (or a scalar) nor of the length of the
<a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup" title="MDAnalysis.core.AtomGroup.AtomGroup"><code class="xref py py-class docutils literal"><span class="pre">AtomGroup</span></code></a> then a <code class="xref py py-exc docutils literal"><span class="pre">ValueError</span></code> is raised.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Changing <em>resnum</em> will not affect the topology: you can have
multiple residues with the same <em>resnum</em>.</p>
</div>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="#MDAnalysis.core.AtomGroup.ResidueGroup.set_resid" title="MDAnalysis.core.AtomGroup.ResidueGroup.set_resid"><code class="xref py py-meth docutils literal"><span class="pre">set_resid()</span></code></a></p>
</div>
<div class="versionadded">
<p><span class="versionmodified">New in version 0.7.4.</span></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 0.7.5: </span>Also changes the residues.</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 0.8: </span>Can set atoms and residues to distinct values by providing a sequence or iterable.</p>
</div>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="MDAnalysis.core.AtomGroup.Segment">
<em class="property">class </em><code class="descclassname">MDAnalysis.core.AtomGroup.</code><code class="descname">Segment</code><span class="sig-paren">(</span><em>name</em>, <em>residues</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#Segment"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.Segment" title="Permalink to this definition">¶</a></dt>
<dd><p>A group of residues corresponding to one segment of the topology.</p>
<p>Pythonic access to residues:</p>
<blockquote>
<div><ul>
<li><p class="first">The attribute rN returns the N-th residue <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Residue" title="MDAnalysis.core.AtomGroup.Residue"><code class="xref py py-class docutils literal"><span class="pre">Residue</span></code></a> of the
segment (numbering starts at N=1). Example:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">MDAnalysis.tests.datafiles</span> <span class="kn">import</span> <span class="n">PSF</span><span class="p">,</span><span class="n">DCD</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">u</span> <span class="o">=</span> <span class="n">Universe</span><span class="p">(</span><span class="n">PSF</span><span class="p">,</span><span class="n">DCD</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span><span class="p">(</span><span class="n">u</span><span class="o">.</span><span class="n">s4AKE</span><span class="o">.</span><span class="n">r1</span><span class="p">)</span>
<span class="go">&lt;Residue &#39;MET&#39;, 1&gt;</span>
</pre></div>
</div>
</li>
<li><p class="first">Using a residue name as attribute returns a list of all residues (a
<a class="reference internal" href="#MDAnalysis.core.AtomGroup.ResidueGroup" title="MDAnalysis.core.AtomGroup.ResidueGroup"><code class="xref py py-class docutils literal"><span class="pre">ResidueGroup</span></code></a>) of the same name. Example:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">MDAnalysis.tests.datafiles</span> <span class="kn">import</span> <span class="n">PSF</span><span class="p">,</span><span class="n">DCD</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">u</span> <span class="o">=</span> <span class="n">Universe</span><span class="p">(</span><span class="n">PSF</span><span class="p">,</span><span class="n">DCD</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span><span class="p">(</span><span class="n">u</span><span class="o">.</span><span class="n">s4AKE</span><span class="o">.</span><span class="n">CYS</span><span class="p">)</span>
<span class="go">&lt;ResidueGroup [&lt;Residue &#39;CYS&#39;, 77&gt;]&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span><span class="p">(</span><span class="n">u</span><span class="o">.</span><span class="n">s4AKE</span><span class="o">.</span><span class="n">MET</span><span class="p">)</span>
<span class="go">&lt;ResidueGroup [&lt;Residue &#39;MET&#39;, 1&gt;, &lt;Residue &#39;MET&#39;, 21&gt;, &lt;Residue &#39;MET&#39;, 34&gt;, &lt;Residue &#39;MET&#39;, 53&gt;,</span>
<span class="go">&lt;Residue &#39;MET&#39;, 96&gt;, &lt;Residue &#39;MET&#39;, 174&gt;]&gt;</span>
</pre></div>
</div>
</li>
</ul>
</div></blockquote>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Data:</th><td class="field-body"><code class="xref py py-attr docutils literal"><span class="pre">Segment.name</span></code> is the segid from the topology or the
chain identifier when loaded from a PDB</td>
</tr>
</tbody>
</table>
<p>Initialize a Segment with segid <em>name</em> from a list of <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Residue" title="MDAnalysis.core.AtomGroup.Residue"><code class="xref py py-class docutils literal"><span class="pre">Residue</span></code></a> instances.</p>
<dl class="attribute">
<dt id="MDAnalysis.core.AtomGroup.Segment.id">
<code class="descname">id</code><a class="headerlink" href="#MDAnalysis.core.AtomGroup.Segment.id" title="Permalink to this definition">¶</a></dt>
<dd><p>Segment id (alias for <code class="xref py py-attr docutils literal"><span class="pre">Segment.name</span></code>)</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="MDAnalysis.core.AtomGroup.SegmentGroup">
<em class="property">class </em><code class="descclassname">MDAnalysis.core.AtomGroup.</code><code class="descname">SegmentGroup</code><span class="sig-paren">(</span><em>segments</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#SegmentGroup"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.SegmentGroup" title="Permalink to this definition">¶</a></dt>
<dd><p>A group of segments.</p>
<dl class="docutils">
<dt>Pythonic access to segments:</dt>
<dd><ul class="first last">
<li><p class="first">Using a segid as attribute returns the segment. Because
of python language rule, any segid starting with a non-letter
character is prefixed with &#8216;s&#8217;, thus &#8216;4AKE&#8217; &#8211;&gt; &#8216;s4AKE&#8217;.</p>
<p>Example:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">MDAnalysis.tests.datafiles</span> <span class="kn">import</span> <span class="n">PSF</span><span class="p">,</span><span class="n">DCD</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">u</span> <span class="o">=</span> <span class="n">Universe</span><span class="p">(</span><span class="n">PSF</span><span class="p">,</span><span class="n">DCD</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span><span class="p">(</span><span class="n">u</span><span class="o">.</span><span class="n">atoms</span><span class="o">.</span><span class="n">segments</span><span class="o">.</span><span class="n">s4AKE</span><span class="p">)</span>  <span class="c"># segment 4AKE</span>
<span class="go">&lt;AtomGroup with 3314 atoms&gt;</span>
</pre></div>
</div>
</li>
<li><p class="first">Indexing the group returns the appropriate segment.</p>
</li>
</ul>
</dd>
</dl>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Data:</th><td class="field-body"><code class="xref py py-attr docutils literal"><span class="pre">SegmentGroup._segments</span></code></td>
</tr>
</tbody>
</table>
<p>Initialize the SegmentGroup with a list of <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Segment" title="MDAnalysis.core.AtomGroup.Segment"><code class="xref py py-class docutils literal"><span class="pre">Segment</span></code></a> instances.</p>
<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.SegmentGroup.set">
<code class="descname">set</code><span class="sig-paren">(</span><em>name</em>, <em>value</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.AtomGroup.SegmentGroup.set" title="Permalink to this definition">¶</a></dt>
<dd><p>Set attribute <em>name</em> to <em>value</em> for all <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Segment" title="MDAnalysis.core.AtomGroup.Segment"><code class="xref py py-class docutils literal"><span class="pre">Segment</span></code></a> in this <a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup" title="MDAnalysis.core.AtomGroup.AtomGroup"><code class="xref py py-class docutils literal"><span class="pre">AtomGroup</span></code></a>.</p>
<p>If <em>value</em> is a sequence of the same length as the
<a class="reference internal" href="#MDAnalysis.core.AtomGroup.SegmentGroup" title="MDAnalysis.core.AtomGroup.SegmentGroup"><code class="xref py py-class docutils literal"><span class="pre">SegmentGroup</span></code></a> (<a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup.residues" title="MDAnalysis.core.AtomGroup.AtomGroup.residues"><code class="xref py py-attr docutils literal"><span class="pre">AtomGroup.residues</span></code></a>) then each
<a class="reference internal" href="#MDAnalysis.core.AtomGroup.Segment" title="MDAnalysis.core.AtomGroup.Segment"><code class="xref py py-class docutils literal"><span class="pre">Segment</span></code></a>&#8216;s property <em>name</em> is set to the corresponding
value. If <em>value</em> is neither of length 1 (or a scalar) nor of the
length of the <a class="reference internal" href="#MDAnalysis.core.AtomGroup.SegmentGroup" title="MDAnalysis.core.AtomGroup.SegmentGroup"><code class="xref py py-class docutils literal"><span class="pre">SegmentGroup</span></code></a> then a <code class="xref py py-exc docutils literal"><span class="pre">ValueError</span></code> is raised.</p>
<div class="versionadded">
<p><span class="versionmodified">New in version 0.8.</span></p>
</div>
</dd></dl>

<dl class="method">
<dt id="MDAnalysis.core.AtomGroup.SegmentGroup.set_segid">
<code class="descname">set_segid</code><span class="sig-paren">(</span><em>segid</em>, <em>buildsegments=True</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#SegmentGroup.set_segid"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.SegmentGroup.set_segid" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the segid to <em>segid</em> for all atoms in the <a class="reference internal" href="#MDAnalysis.core.AtomGroup.SegmentGroup" title="MDAnalysis.core.AtomGroup.SegmentGroup"><code class="xref py py-class docutils literal"><span class="pre">SegmentGroup</span></code></a>.</p>
<p>If <em>segid</em> is a sequence of the same length as the <a class="reference internal" href="#MDAnalysis.core.AtomGroup.SegmentGroup" title="MDAnalysis.core.AtomGroup.SegmentGroup"><code class="xref py py-class docutils literal"><span class="pre">SegmentGroup</span></code></a>
then each <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Atom.segid" title="MDAnalysis.core.AtomGroup.Atom.segid"><code class="xref py py-attr docutils literal"><span class="pre">Atom.segid</span></code></a> is set to the corresponding value together
with the <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Segment.id" title="MDAnalysis.core.AtomGroup.Segment.id"><code class="xref py py-attr docutils literal"><span class="pre">Segment.id</span></code></a> of the segment the atom belongs to. If
<em>value</em> is neither of length 1 (or a scalar) nor of the length of the
<a class="reference internal" href="#MDAnalysis.core.AtomGroup.AtomGroup" title="MDAnalysis.core.AtomGroup.AtomGroup"><code class="xref py py-class docutils literal"><span class="pre">AtomGroup</span></code></a> then a <code class="xref py py-exc docutils literal"><span class="pre">ValueError</span></code> is raised.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p><a class="reference internal" href="#MDAnalysis.core.AtomGroup.SegmentGroup.set_segid" title="MDAnalysis.core.AtomGroup.SegmentGroup.set_segid"><code class="xref py py-meth docutils literal"><span class="pre">set_segid()</span></code></a> can change the topology.</p>
<p class="last">With the default <em>buildsegments</em> = <code class="docutils literal"><span class="pre">True</span></code> it can be used to join
segments or to break groups into multiple disjoint segments. Note
that each <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Atom" title="MDAnalysis.core.AtomGroup.Atom"><code class="xref py py-class docutils literal"><span class="pre">Atom</span></code></a> can only belong to a single
<a class="reference internal" href="#MDAnalysis.core.AtomGroup.Segment" title="MDAnalysis.core.AtomGroup.Segment"><code class="xref py py-class docutils literal"><span class="pre">Segment</span></code></a>.</p>
</div>
<p>For performance reasons, <em>buildsegments</em> can be set to <code class="docutils literal"><span class="pre">False</span></code>. Then
one needs to run <code class="xref py py-meth docutils literal"><span class="pre">Universe._build_segments()</span></code> manually later in
order to update the list of <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Segment" title="MDAnalysis.core.AtomGroup.Segment"><code class="xref py py-class docutils literal"><span class="pre">Segment</span></code></a> instances and regenerate
the segid instant selectors.</p>
<div class="admonition warning">
<p class="first admonition-title">Warning</p>
<p class="last">The values of <em>this</em> <a class="reference internal" href="#MDAnalysis.core.AtomGroup.SegmentGroup" title="MDAnalysis.core.AtomGroup.SegmentGroup"><code class="xref py py-class docutils literal"><span class="pre">SegmentGroup</span></code></a> are not being changed
(i.e. if you assign multiple segids this instance will not be broken
in multiple segments, rather you will have one <a class="reference internal" href="#MDAnalysis.core.AtomGroup.SegmentGroup" title="MDAnalysis.core.AtomGroup.SegmentGroup"><code class="xref py py-class docutils literal"><span class="pre">SegmentGroup</span></code></a>
that groups multiple segments together). You <strong>must create a new</strong>
<a class="reference internal" href="#MDAnalysis.core.AtomGroup.SegmentGroup" title="MDAnalysis.core.AtomGroup.SegmentGroup"><code class="xref py py-class docutils literal"><span class="pre">SegmentGroup</span></code></a> <strong>from the</strong> <a class="reference internal" href="#MDAnalysis.core.AtomGroup.Universe" title="MDAnalysis.core.AtomGroup.Universe"><code class="xref py py-class docutils literal"><span class="pre">Universe</span></code></a> &#8212; only
<a class="reference internal" href="#MDAnalysis.core.AtomGroup.Atom" title="MDAnalysis.core.AtomGroup.Atom"><code class="xref py py-class docutils literal"><span class="pre">Atom</span></code></a> instances are changed, everything else is derived from
these atoms.</p>
</div>
<div class="versionadded">
<p><span class="versionmodified">New in version 0.7.4.</span></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 0.8: </span>Can set atoms and residues to distinct values by providing a sequence or iterable.</p>
</div>
</dd></dl>

</dd></dl>

<dl class="function">
<dt id="MDAnalysis.core.AtomGroup.asUniverse">
<code class="descclassname">MDAnalysis.core.AtomGroup.</code><code class="descname">asUniverse</code><span class="sig-paren">(</span><em>*args</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/AtomGroup.html#asUniverse"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.asUniverse" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a universe from the input arguments.</p>
<ol class="arabic">
<li><p class="first">If the first argument is a universe, just return it:</p>
<div class="highlight-python"><div class="highlight"><pre>as_universe(universe) --&gt; universe
</pre></div>
</div>
</li>
<li><p class="first">Otherwise try to build a universe from the first or the first
and second argument:</p>
<div class="highlight-python"><div class="highlight"><pre>asUniverse(PDB, **kwargs) --&gt; Universe(PDB, **kwargs)
asUniverse(PSF, DCD, **kwargs) --&gt; Universe(PSF, DCD, **kwargs)
asUniverse(*args, **kwargs) --&gt; Universe(*args, **kwargs)
</pre></div>
</div>
</li>
</ol>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">an instance of <code class="xref py py-class docutils literal"><span class="pre">Universe</span></code></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="exception">
<dt id="MDAnalysis.core.AtomGroup.SelectionError">
<em class="property">exception </em><code class="descclassname">MDAnalysis.core.AtomGroup.</code><code class="descname">SelectionError</code><a class="reference internal" href="../../_modules/MDAnalysis.html#SelectionError"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.SelectionError" title="Permalink to this definition">¶</a></dt>
<dd><p>Raised when a atom selection failed.</p>
</dd></dl>

<dl class="exception">
<dt id="MDAnalysis.core.AtomGroup.SelectionWarning">
<em class="property">exception </em><code class="descclassname">MDAnalysis.core.AtomGroup.</code><code class="descname">SelectionWarning</code><a class="reference internal" href="../../_modules/MDAnalysis.html#SelectionWarning"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.SelectionWarning" title="Permalink to this definition">¶</a></dt>
<dd><p>Warning indicating a possible problem with a selection.</p>
</dd></dl>

<dl class="exception">
<dt id="MDAnalysis.core.AtomGroup.NoDataError">
<em class="property">exception </em><code class="descclassname">MDAnalysis.core.AtomGroup.</code><code class="descname">NoDataError</code><a class="reference internal" href="../../_modules/MDAnalysis.html#NoDataError"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.core.AtomGroup.NoDataError" title="Permalink to this definition">¶</a></dt>
<dd><p>Raised when empty input is not allowed or required data are missing.</p>
</dd></dl>

</div>
</div>


          </div>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="Selection.html" title="7.3. Atom selection Hierarchy — MDAnalysis.core.Selection"
             >next</a> |</li>
        <li class="right" >
          <a href="init.html" title="7.1. Core functions of MDAnalysis"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../../index.html">MDAnalysis 0.11.0-dev documentation</a> &raquo;</li>
          <li class="nav-item nav-item-1"><a href="../core_modules.html" >7. Core modules</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2005-2015, Naveen Michaud-Agrawal, Elizabeth J. Denning, Joshua Adelman,
    Christian Beckstein (logo), Alejandro Bernardin, Sébastien Buchoux,
    David Caplan, Matthieu Chavent, Xavier Deupi, Jan Domański, David L. Dotson
    Lennard van der Feltz, Philip Fowler, Joseph Goose, Richard J. Gowers, Lukas Grossar,
    Benjamin Hall, Joe Jordan, Jinju Lu, Robert McGibbon, Alex Nesterenko,
    Manuel Nuno Melo, Caio S. Souza, Danny Parton, Joshua L. Phillips, Tyler Reddy,
    Paul Rigor, Sean L. Seyler, Andy Somogyi, Lukas Stelzl, Zhuyi Xue,
    and Oliver Beckstein.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3.1.
    </div>
  </body>
</html>